// twitch/lib/API/Helix/Subscriptions/HelixSubscriptionApi.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.HelixSubscriptionApi = void 0;
var tslib_1 = require("tslib");
var twitch_api_call_1 = require("twitch-api-call");
var twitch_common_1 = require("twitch-common");
var BaseApi_1 = require("../../BaseApi");
var HelixPaginatedRequest_1 = require("../HelixPaginatedRequest");
var HelixPaginatedRequestWithTotal_1 = require("../HelixPaginatedRequestWithTotal");
var HelixPaginatedResult_1 = require("../HelixPaginatedResult");
var HelixSubscription_1 = require("./HelixSubscription");
var HelixSubscriptionEvent_1 = require("./HelixSubscriptionEvent");
var HelixUserSubscription_1 = require("./HelixUserSubscription");
/**
 * The Helix API methods that deal with subscriptions.
 *
 * Can be accessed using `client.helix.subscriptions` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const subscription = await api.helix.subscriptions.getSubscriptionForUser('61369223', '125328655');
 * ```
 */
var HelixSubscriptionApi = /** @class */ (function (_super) ***REMOVED***
    tslib_1.__extends(HelixSubscriptionApi, _super);
    function HelixSubscriptionApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves a list of all subscriptions to a given broadcaster.
     *
     * @param broadcaster The broadcaster to list subscriptions to.
     */
    HelixSubscriptionApi.prototype.getSubscriptions = function (broadcaster) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            url: 'subscriptions',
                            scope: 'channel:read:subscriptions',
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            query: ***REMOVED***
                                broadcaster_id: twitch_common_1.extractUserId(broadcaster)
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, HelixPaginatedResult_1.createPaginatedResultWithTotal(result, HelixSubscription_1.HelixSubscription, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for all subscriptions to a given broadcaster.
     *
     * @param broadcaster The broadcaster to list subscriptions to.
     */
    HelixSubscriptionApi.prototype.getSubscriptionsPaginated = function (broadcaster) ***REMOVED***
        var _this = this;
        return new HelixPaginatedRequestWithTotal_1.HelixPaginatedRequestWithTotal(***REMOVED***
            url: 'subscriptions',
            scope: 'channel:read:subscriptions',
            query: ***REMOVED***
                broadcaster_id: twitch_common_1.extractUserId(broadcaster)
            ***REMOVED***
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixSubscription_1.HelixSubscription(data, _this._client); ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the subset of the given user list that is subscribed to the given broadcaster.
     *
     * @param broadcaster The broadcaster to find subscriptions to.
     * @param users The users that should be checked for subscriptions.
     */
    HelixSubscriptionApi.prototype.getSubscriptionsForUsers = function (broadcaster, users) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            var _this = this;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            url: 'subscriptions',
                            scope: 'channel:read:subscriptions',
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            query: ***REMOVED***
                                broadcaster_id: twitch_common_1.extractUserId(broadcaster),
                                user_id: users.map(twitch_common_1.extractUserId)
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.map(function (data) ***REMOVED*** return new HelixSubscription_1.HelixSubscription(data, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the subscription data for a given user to a given broadcaster.
     *
     * This checks with the authorization of a broadcaster.
     * If you only have the authorization of a user, check ***REMOVED***@HelixSubscriptionApi#checkUserSubscription***REMOVED***.
     *
     * @param broadcaster The broadcaster to check.
     * @param user The user to check.
     */
    HelixSubscriptionApi.prototype.getSubscriptionForUser = function (broadcaster, user) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var list;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.getSubscriptionsForUsers(broadcaster, [user])];
                    case 1:
                        list = _a.sent();
                        return [2 /*return*/, list.length ? list[0] : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the most recent subscription events for a given broadcaster.
     *
     * @param broadcaster The broadcaster to retrieve subscription events for.
     */
    HelixSubscriptionApi.prototype.getSubscriptionEventsForBroadcaster = function (broadcaster) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                return [2 /*return*/, this._getSubscriptionEvents('broadcaster_id', twitch_common_1.extractUserId(broadcaster))];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for the recent subscription events for a given broadcaster.
     *
     * @param broadcaster The broadcaster to retrieve subscription events for.
     */
    HelixSubscriptionApi.prototype.getSubscriptionEventsForBroadcasterPaginated = function (broadcaster) ***REMOVED***
        var _this = this;
        return new HelixPaginatedRequest_1.HelixPaginatedRequest(***REMOVED***
            url: 'subscriptions/events',
            scope: 'channel:read:subscriptions',
            query: ***REMOVED***
                broadcaster_id: twitch_common_1.extractUserId(broadcaster)
            ***REMOVED***
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixSubscriptionEvent_1.HelixSubscriptionEvent(data, _this._client); ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a single subscription event by ID.
     *
     * @param id The event ID.
     */
    HelixSubscriptionApi.prototype.getSubscriptionEventById = function (id) ***REMOVED***
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var events;
            return tslib_1.__generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._getSubscriptionEvents('id', id)];
                    case 1:
                        events = _b.sent();
                        return [2 /*return*/, (_a = events.data[0]) !== null && _a !== void 0 ? _a : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Checks if a given user is subscribed to a given broadcaster. Returns null if not subscribed.
     *
     * This checks with the authorization of a user.
     * If you only have the authorization of a broadcaster, check ***REMOVED***@HelixSubscriptionApi#getSubscriptionForUser***REMOVED***.
     *
     * @param user The broadcaster to check the user's subscription for.
     * @param broadcaster The user to check.
     */
    HelixSubscriptionApi.prototype.checkUserSubscription = function (user, broadcaster) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result, e_1;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: twitch_api_call_1.TwitchApiCallType.Helix,
                                url: 'subscriptions/user',
                                scope: 'user:read:subscriptions',
                                query: ***REMOVED***
                                    broadcaster_id: twitch_common_1.extractUserId(broadcaster),
                                    user_id: twitch_common_1.extractUserId(user)
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixUserSubscription_1.HelixUserSubscription(result.data[0], this._client)];
                    case 2:
                        e_1 = _a.sent();
                        if (e_1 instanceof twitch_api_call_1.HttpStatusCodeError && e_1.statusCode === 404) ***REMOVED***
                            return [2 /*return*/, null];
                        ***REMOVED***
                        throw e_1;
                    case 3: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixSubscriptionApi.prototype._getSubscriptionEvents = function (by, id) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            var _a;
            return tslib_1.__generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'subscriptions/events',
                            scope: 'channel:read:subscriptions',
                            query: (_a = ***REMOVED******REMOVED***,
                                _a[by] = id,
                                _a)
                        ***REMOVED***)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, HelixPaginatedResult_1.createPaginatedResult(result, HelixSubscriptionEvent_1.HelixSubscriptionEvent, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixSubscriptionApi = tslib_1.__decorate([
        twitch_common_1.rtfm('twitch', 'HelixSubscriptionApi')
    ], HelixSubscriptionApi);
    return HelixSubscriptionApi;
***REMOVED***(BaseApi_1.BaseApi));
exports.HelixSubscriptionApi = HelixSubscriptionApi;
