// twitch/lib/API/Helix/Poll/HelixPollApi.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.HelixPollApi = void 0;
var tslib_1 = require("tslib");
var twitch_api_call_1 = require("twitch-api-call");
var twitch_common_1 = require("twitch-common");
var BaseApi_1 = require("../../BaseApi");
var HelixPaginatedRequest_1 = require("../HelixPaginatedRequest");
var HelixPaginatedResult_1 = require("../HelixPaginatedResult");
var HelixPagination_1 = require("../HelixPagination");
var HelixPoll_1 = require("./HelixPoll");
/**
 * The Helix API methods that deal with polls.
 *
 * Can be accessed using `client.helix.polls` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const ***REMOVED*** data: polls ***REMOVED*** = await api.helix.polls.getPolls('61369223');
 * ```
 */
var HelixPollApi = /** @class */ (function (_super) ***REMOVED***
    tslib_1.__extends(HelixPollApi, _super);
    function HelixPollApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves a list of polls for the given broadcaster.
     *
     * @param broadcaster The broadcaster to retrieve polls for.
     * @param pagination
     *
     * @expandParams
     */
    HelixPollApi.prototype.getPolls = function (broadcaster, pagination) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'polls',
                            scope: 'channel:read:polls',
                            query: tslib_1.__assign(***REMOVED*** broadcaster_id: twitch_common_1.extractUserId(broadcaster) ***REMOVED***, HelixPagination_1.makePaginationQuery(pagination))
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, HelixPaginatedResult_1.createPaginatedResult(result, HelixPoll_1.HelixPoll, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for polls for the given broadcaster.
     *
     * @param broadcaster The broadcaster to retrieve polls for.
     */
    HelixPollApi.prototype.getPollsPaginated = function (broadcaster) ***REMOVED***
        var _this = this;
        return new HelixPaginatedRequest_1.HelixPaginatedRequest(***REMOVED***
            url: 'polls',
            scope: 'channel:read:polls',
            query: ***REMOVED***
                broadcaster_id: twitch_common_1.extractUserId(broadcaster)
            ***REMOVED***
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixPoll_1.HelixPoll(data, _this._client); ***REMOVED***, 20);
    ***REMOVED***;
    /**
     * Retrieves polls by IDs.
     *
     * @param broadcaster The broadcaster to retrieve the polls for.
     * @param ids The IDs of the polls.
     */
    HelixPollApi.prototype.getPollsByIds = function (broadcaster, ids) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            var _this = this;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        if (!ids.length) ***REMOVED***
                            return [2 /*return*/, []];
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: twitch_api_call_1.TwitchApiCallType.Helix,
                                url: 'polls',
                                scope: 'channel:read:polls',
                                query: ***REMOVED***
                                    broadcaster_id: twitch_common_1.extractUserId(broadcaster),
                                    id: ids
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.map(function (data) ***REMOVED*** return new HelixPoll_1.HelixPoll(data, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a poll by ID.
     *
     * @param broadcaster The broadcaster to retrieve the poll for.
     * @param id The ID of the poll.
     */
    HelixPollApi.prototype.getPollById = function (broadcaster, id) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var polls;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.getPollsByIds(broadcaster, [id])];
                    case 1:
                        polls = _a.sent();
                        return [2 /*return*/, polls.length ? polls[0] : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a new poll.
     *
     * @param broadcaster The broadcaster to create the poll for.
     * @param data
     *
     * @expandParams
     */
    HelixPollApi.prototype.createPoll = function (broadcaster, data) ***REMOVED***
        var _a, _b;
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            return tslib_1.__generator(this, function (_c) ***REMOVED***
                switch (_c.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'polls',
                            method: 'POST',
                            scope: 'channel:manage:polls',
                            jsonBody: ***REMOVED***
                                broadcaster_id: twitch_common_1.extractUserId(broadcaster),
                                title: data.title,
                                choices: data.choices.map(function (title) ***REMOVED*** return (***REMOVED*** title: title ***REMOVED***); ***REMOVED***),
                                duration: data.duration,
                                bits_voting_enabled: data.bitsPerVote != null,
                                bits_per_vote: (_a = data.bitsPerVote) !== null && _a !== void 0 ? _a : 0,
                                channel_points_voting_enabled: data.channelPointsPerVote != null,
                                channel_points_per_vote: (_b = data.channelPointsPerVote) !== null && _b !== void 0 ? _b : 0
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, new HelixPoll_1.HelixPoll(result.data[0], this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Ends a poll.
     *
     * @param broadcaster The broadcaster to end the poll for.
     * @param id The ID of the poll to end.
     * @param showResult Whether to allow the result to be viewed publicly.
     */
    HelixPollApi.prototype.endPoll = function (broadcaster, id, showResult) ***REMOVED***
        if (showResult === void 0) ***REMOVED*** showResult = true; ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'polls',
                            method: 'PATCH',
                            scope: 'channel:manage:polls',
                            jsonBody: ***REMOVED***
                                broadcaster_id: twitch_common_1.extractUserId(broadcaster),
                                id: id,
                                status: showResult ? 'TERMINATED' : 'ARCHIVED'
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixPoll_1.HelixPoll(result.data[0], this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixPollApi = tslib_1.__decorate([
        twitch_common_1.rtfm('twitch', 'HelixPollApi')
    ], HelixPollApi);
    return HelixPollApi;
***REMOVED***(BaseApi_1.BaseApi));
exports.HelixPollApi = HelixPollApi;
