// @d-fischer/qs/dist/qs.js
(function(f)***REMOVED***if(typeof exports==="object"&&typeof module!=="undefined")***REMOVED***module.exports=f()***REMOVED***else if(typeof define==="function"&&define.amd)***REMOVED***define([],f)***REMOVED***else***REMOVED***var g;if(typeof window!=="undefined")***REMOVED***g=window***REMOVED***else if(typeof global!=="undefined")***REMOVED***g=global***REMOVED***else if(typeof self!=="undefined")***REMOVED***g=self***REMOVED***else***REMOVED***g=this***REMOVED***g.Qs = f()***REMOVED******REMOVED***)(function()***REMOVED***var define,module,exports;return (function()***REMOVED***function r(e,n,t)***REMOVED***function o(i,f)***REMOVED***if(!n[i])***REMOVED***if(!e[i])***REMOVED***var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a***REMOVED***var p=n[i]=***REMOVED***exports:***REMOVED******REMOVED******REMOVED***;e[i][0].call(p.exports,function(r)***REMOVED***var n=e[i][1][r];return o(n||r)***REMOVED***,p,p.exports,r,e,n,t)***REMOVED***return n[i].exports***REMOVED***for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o***REMOVED***return r***REMOVED***)()(***REMOVED***1:[function(require,module,exports)***REMOVED***
'use strict';

var replace = String.prototype.replace;
var percentTwenties = /%20/g;

var util = require('./utils');

var Format = ***REMOVED***
    RFC1738: 'RFC1738',
    RFC3986: 'RFC3986'
***REMOVED***;

module.exports = util.assign(
    ***REMOVED***
        'default': Format.RFC3986,
        formatters: ***REMOVED***
            RFC1738: function (value) ***REMOVED***
                return replace.call(value, percentTwenties, '+');
        ***REMOVED***
            RFC3986: function (value) ***REMOVED***
                return String(value);
            ***REMOVED***
        ***REMOVED***
***REMOVED***
    Format
);

***REMOVED***,***REMOVED***"./utils":5***REMOVED***],2:[function(require,module,exports)***REMOVED***
'use strict';

var stringify = require('./stringify');
var parse = require('./parse');
var formats = require('./formats');

module.exports = ***REMOVED***
    formats: formats,
    parse: parse,
    stringify: stringify
***REMOVED***;

***REMOVED***,***REMOVED***"./formats":1,"./parse":3,"./stringify":4***REMOVED***],3:[function(require,module,exports)***REMOVED***
'use strict';

var utils = require('./utils');

var has = Object.prototype.hasOwnProperty;
var isArray = Array.isArray;

var defaults = ***REMOVED***
    allowDots: false,
    allowPrototypes: false,
    arrayLimit: 20,
    charset: 'utf-8',
    charsetSentinel: false,
    comma: false,
    decoder: utils.decode,
    delimiter: '&',
    depth: 5,
    ignoreQueryPrefix: false,
    interpretNumericEntities: false,
    parameterLimit: 1000,
    parseArrays: true,
    plainObjects: false,
    strictNullHandling: false
***REMOVED***;

var interpretNumericEntities = function (str) ***REMOVED***
    return str.replace(/&#(\d+);/g, function ($0, numberStr) ***REMOVED***
        return String.fromCharCode(parseInt(numberStr, 10));
    ***REMOVED***);
***REMOVED***;

var parseArrayValue = function (val, options) ***REMOVED***
    if (val && typeof val === 'string' && options.comma && val.indexOf(',') > -1) ***REMOVED***
        return val.split(',');
    ***REMOVED***

    return val;
***REMOVED***;

var maybeMap = function maybeMap(val, fn) ***REMOVED***
    if (isArray(val)) ***REMOVED***
        var mapped = [];
        for (var i = 0; i < val.length; i += 1) ***REMOVED***
            mapped.push(fn(val[i]));
        ***REMOVED***
        return mapped;
    ***REMOVED***
    return fn(val);
***REMOVED***;

// This is what browsers will submit when the ✓ character occurs in an
// application/x-www-form-urlencoded body and the encoding of the page containing
// the form is iso-8859-1, or when the submitted form has an accept-charset
// attribute of iso-8859-1. Presumably also with other charsets that do not contain
// the ✓ character, such as us-ascii.
var isoSentinel = 'utf8=%26%2310003%3B'; // encodeURIComponent('&#10003;')

// These are the percent-encoded utf-8 octets representing a checkmark, indicating that the request actually is utf-8 encoded.
var charsetSentinel = 'utf8=%E2%9C%93'; // encodeURIComponent('✓')

var parseValues = function parseQueryStringValues(str, options) ***REMOVED***
    var obj = ***REMOVED******REMOVED***;
    var cleanStr = options.ignoreQueryPrefix ? str.replace(/^\?/, '') : str;
    var limit = options.parameterLimit === Infinity ? undefined : options.parameterLimit;
    var parts = cleanStr.split(options.delimiter, limit);
    var skipIndex = -1; // Keep track of where the utf8 sentinel was found
    var i;

    var charset = options.charset;
    if (options.charsetSentinel) ***REMOVED***
        for (i = 0; i < parts.length; ++i) ***REMOVED***
            if (parts[i].indexOf('utf8=') === 0) ***REMOVED***
                if (parts[i] === charsetSentinel) ***REMOVED***
                    charset = 'utf-8';
                ***REMOVED*** else if (parts[i] === isoSentinel) ***REMOVED***
                    charset = 'iso-8859-1';
                ***REMOVED***
                skipIndex = i;
                i = parts.length; // The eslint settings do not allow break;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    for (i = 0; i < parts.length; ++i) ***REMOVED***
        if (i === skipIndex) ***REMOVED***
            continue;
        ***REMOVED***
        var part = parts[i];

        var bracketEqualsPos = part.indexOf(']=');
        var pos = bracketEqualsPos === -1 ? part.indexOf('=') : bracketEqualsPos + 1;

        var key, val;
        if (pos === -1) ***REMOVED***
            key = options.decoder(part, defaults.decoder, charset, 'key');
            val = options.strictNullHandling ? null : '';
        ***REMOVED*** else ***REMOVED***
            key = options.decoder(part.slice(0, pos), defaults.decoder, charset, 'key');
            val = maybeMap(
                parseArrayValue(part.slice(pos + 1), options),
                function (encodedVal) ***REMOVED***
                    return options.decoder(encodedVal, defaults.decoder, charset, 'value');
                ***REMOVED***
            );
        ***REMOVED***

        if (val && options.interpretNumericEntities && charset === 'iso-8859-1') ***REMOVED***
            val = interpretNumericEntities(val);
        ***REMOVED***

        if (part.indexOf('[]=') > -1) ***REMOVED***
            val = isArray(val) ? [val] : val;
        ***REMOVED***

        if (has.call(obj, key)) ***REMOVED***
            obj[key] = utils.combine(obj[key], val);
        ***REMOVED*** else ***REMOVED***
            obj[key] = val;
        ***REMOVED***
    ***REMOVED***

    return obj;
***REMOVED***;

var parseObject = function (chain, val, options, valuesParsed) ***REMOVED***
    var leaf = valuesParsed ? val : parseArrayValue(val, options);

    for (var i = chain.length - 1; i >= 0; --i) ***REMOVED***
        var obj;
        var root = chain[i];

        if (root === '[]' && options.parseArrays) ***REMOVED***
            obj = [].concat(leaf);
        ***REMOVED*** else ***REMOVED***
            obj = options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;
            var cleanRoot = root.charAt(0) === '[' && root.charAt(root.length - 1) === ']' ? root.slice(1, -1) : root;
            var index = parseInt(cleanRoot, 10);
            if (!options.parseArrays && cleanRoot === '') ***REMOVED***
                obj = ***REMOVED*** 0: leaf ***REMOVED***;
            ***REMOVED*** else if (
                !isNaN(index)
                && root !== cleanRoot
                && String(index) === cleanRoot
                && index >= 0
                && (options.parseArrays && index <= options.arrayLimit)
            ) ***REMOVED***
                obj = [];
                obj[index] = leaf;
            ***REMOVED*** else ***REMOVED***
                obj[cleanRoot] = leaf;
            ***REMOVED***
        ***REMOVED***

        leaf = obj; // eslint-disable-line no-param-reassign
    ***REMOVED***

    return leaf;
***REMOVED***;

var parseKeys = function parseQueryStringKeys(givenKey, val, options, valuesParsed) ***REMOVED***
    if (!givenKey) ***REMOVED***
        return;
    ***REMOVED***

    // Transform dot notation to bracket notation
    var key = options.allowDots ? givenKey.replace(/\.([^.[]+)/g, '[$1]') : givenKey;

    // The regex chunks

    var brackets = /(\[[^[\]]*])/;
    var child = /(\[[^[\]]*])/g;

    // Get the parent

    var segment = options.depth > 0 && brackets.exec(key);
    var parent = segment ? key.slice(0, segment.index) : key;

    // Stash the parent if it exists

    var keys = [];
    if (parent) ***REMOVED***
        // If we aren't using plain objects, optionally prefix keys that would overwrite object prototype properties
        if (!options.plainObjects && has.call(Object.prototype, parent)) ***REMOVED***
            if (!options.allowPrototypes) ***REMOVED***
                return;
            ***REMOVED***
        ***REMOVED***

        keys.push(parent);
    ***REMOVED***

    // Loop through children appending to the array until we hit depth

    var i = 0;
    while (options.depth > 0 && (segment = child.exec(key)) !== null && i < options.depth) ***REMOVED***
        i += 1;
        if (!options.plainObjects && has.call(Object.prototype, segment[1].slice(1, -1))) ***REMOVED***
            if (!options.allowPrototypes) ***REMOVED***
                return;
            ***REMOVED***
        ***REMOVED***
        keys.push(segment[1]);
    ***REMOVED***

    // If there's a remainder, just add whatever is left

    if (segment) ***REMOVED***
        keys.push('[' + key.slice(segment.index) + ']');
    ***REMOVED***

    return parseObject(keys, val, options, valuesParsed);
***REMOVED***;

var normalizeParseOptions = function normalizeParseOptions(opts) ***REMOVED***
    if (!opts) ***REMOVED***
        return defaults;
    ***REMOVED***

    if (opts.decoder !== null && opts.decoder !== undefined && typeof opts.decoder !== 'function') ***REMOVED***
        throw new TypeError('Decoder has to be a function.');
    ***REMOVED***

    if (typeof opts.charset !== 'undefined' && opts.charset !== 'utf-8' && opts.charset !== 'iso-8859-1') ***REMOVED***
        throw new TypeError('The charset option must be either utf-8, iso-8859-1, or undefined');
    ***REMOVED***
    var charset = typeof opts.charset === 'undefined' ? defaults.charset : opts.charset;

    return ***REMOVED***
        allowDots: typeof opts.allowDots === 'undefined' ? defaults.allowDots : !!opts.allowDots,
        allowPrototypes: typeof opts.allowPrototypes === 'boolean' ? opts.allowPrototypes : defaults.allowPrototypes,
        arrayLimit: typeof opts.arrayLimit === 'number' ? opts.arrayLimit : defaults.arrayLimit,
        charset: charset,
        charsetSentinel: typeof opts.charsetSentinel === 'boolean' ? opts.charsetSentinel : defaults.charsetSentinel,
        comma: typeof opts.comma === 'boolean' ? opts.comma : defaults.comma,
        decoder: typeof opts.decoder === 'function' ? opts.decoder : defaults.decoder,
        delimiter: typeof opts.delimiter === 'string' || utils.isRegExp(opts.delimiter) ? opts.delimiter : defaults.delimiter,
        // eslint-disable-next-line no-implicit-coercion, no-extra-parens
        depth: (typeof opts.depth === 'number' || opts.depth === false) ? +opts.depth : defaults.depth,
        ignoreQueryPrefix: opts.ignoreQueryPrefix === true,
        interpretNumericEntities: typeof opts.interpretNumericEntities === 'boolean' ? opts.interpretNumericEntities : defaults.interpretNumericEntities,
        parameterLimit: typeof opts.parameterLimit === 'number' ? opts.parameterLimit : defaults.parameterLimit,
        parseArrays: opts.parseArrays !== false,
        plainObjects: typeof opts.plainObjects === 'boolean' ? opts.plainObjects : defaults.plainObjects,
        strictNullHandling: typeof opts.strictNullHandling === 'boolean' ? opts.strictNullHandling : defaults.strictNullHandling
    ***REMOVED***;
***REMOVED***;

module.exports = function (str, opts) ***REMOVED***
    var options = normalizeParseOptions(opts);

    if (str === '' || str === null || typeof str === 'undefined') ***REMOVED***
        return options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;
    ***REMOVED***

    var tempObj = typeof str === 'string' ? parseValues(str, options) : str;
    var obj = options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;

    // Iterate over the keys and setup the new object

    var keys = Object.keys(tempObj);
    for (var i = 0; i < keys.length; ++i) ***REMOVED***
        var key = keys[i];
        var newObj = parseKeys(key, tempObj[key], options, typeof str === 'string');
        obj = utils.merge(obj, newObj, options);
    ***REMOVED***

    return utils.compact(obj);
***REMOVED***;

***REMOVED***,***REMOVED***"./utils":5***REMOVED***],4:[function(require,module,exports)***REMOVED***
'use strict';

var utils = require('./utils');
var formats = require('./formats');
var has = Object.prototype.hasOwnProperty;

var arrayPrefixGenerators = ***REMOVED***
    brackets: function brackets(prefix) ***REMOVED***
        return prefix + '[]';
***REMOVED***
    comma: 'comma',
    indices: function indices(prefix, key) ***REMOVED***
        return prefix + '[' + key + ']';
***REMOVED***
    repeat: function repeat(prefix) ***REMOVED***
        return prefix;
    ***REMOVED***
***REMOVED***;

var isArray = Array.isArray;
var push = Array.prototype.push;
var pushToArray = function (arr, valueOrArray) ***REMOVED***
    push.apply(arr, isArray(valueOrArray) ? valueOrArray : [valueOrArray]);
***REMOVED***;

var toISO = Date.prototype.toISOString;

var defaultFormat = formats['default'];
var defaults = ***REMOVED***
    addQueryPrefix: false,
    allowDots: false,
    charset: 'utf-8',
    charsetSentinel: false,
    delimiter: '&',
    encode: true,
    encoder: utils.encode,
    encodeValuesOnly: false,
    format: defaultFormat,
    formatter: formats.formatters[defaultFormat],
    // deprecated
    indices: false,
    serializeDate: function serializeDate(date) ***REMOVED***
        return toISO.call(date);
***REMOVED***
    skipNulls: false,
    strictNullHandling: false
***REMOVED***;

var isNonNullishPrimitive = function isNonNullishPrimitive(v) ***REMOVED***
    return typeof v === 'string'
        || typeof v === 'number'
        || typeof v === 'boolean'
        || typeof v === 'symbol'
        || typeof v === 'bigint';
***REMOVED***;

var stringify = function stringify(
    object,
    prefix,
    generateArrayPrefix,
    strictNullHandling,
    skipNulls,
    encoder,
    filter,
    sort,
    allowDots,
    serializeDate,
    formatter,
    encodeValuesOnly,
    charset
) ***REMOVED***
    var obj = object;
    if (typeof filter === 'function') ***REMOVED***
        obj = filter(prefix, obj);
    ***REMOVED*** else if (obj instanceof Date) ***REMOVED***
        obj = serializeDate(obj);
    ***REMOVED*** else if (generateArrayPrefix === 'comma' && isArray(obj)) ***REMOVED***
        obj = obj.join(',');
    ***REMOVED***

    if (obj === null) ***REMOVED***
        if (strictNullHandling) ***REMOVED***
            return encoder && !encodeValuesOnly ? encoder(prefix, defaults.encoder, charset, 'key') : prefix;
        ***REMOVED***

        obj = '';
    ***REMOVED***

    if (isNonNullishPrimitive(obj) || utils.isBuffer(obj)) ***REMOVED***
        if (encoder) ***REMOVED***
            var keyValue = encodeValuesOnly ? prefix : encoder(prefix, defaults.encoder, charset, 'key');
            return [formatter(keyValue) + '=' + formatter(encoder(obj, defaults.encoder, charset, 'value'))];
        ***REMOVED***
        return [formatter(prefix) + '=' + formatter(String(obj))];
    ***REMOVED***

    var values = [];

    if (typeof obj === 'undefined') ***REMOVED***
        return values;
    ***REMOVED***

    var objKeys;
    if (isArray(filter)) ***REMOVED***
        objKeys = filter;
    ***REMOVED*** else ***REMOVED***
        var keys = Object.keys(obj);
        objKeys = sort ? keys.sort(sort) : keys;
    ***REMOVED***

    for (var i = 0; i < objKeys.length; ++i) ***REMOVED***
        var key = objKeys[i];

        if (skipNulls && obj[key] === null) ***REMOVED***
            continue;
        ***REMOVED***

        if (isArray(obj)) ***REMOVED***
            pushToArray(values, stringify(
                obj[key],
                typeof generateArrayPrefix === 'function' ? generateArrayPrefix(prefix, key) : prefix,
                generateArrayPrefix,
                strictNullHandling,
                skipNulls,
                encoder,
                filter,
                sort,
                allowDots,
                serializeDate,
                formatter,
                encodeValuesOnly,
                charset
            ));
        ***REMOVED*** else ***REMOVED***
            pushToArray(values, stringify(
                obj[key],
                prefix + (allowDots ? '.' + key : '[' + key + ']'),
                generateArrayPrefix,
                strictNullHandling,
                skipNulls,
                encoder,
                filter,
                sort,
                allowDots,
                serializeDate,
                formatter,
                encodeValuesOnly,
                charset
            ));
        ***REMOVED***
    ***REMOVED***

    return values;
***REMOVED***;

var normalizeStringifyOptions = function normalizeStringifyOptions(opts) ***REMOVED***
    if (!opts) ***REMOVED***
        return defaults;
    ***REMOVED***

    if (opts.encoder !== null && opts.encoder !== undefined && typeof opts.encoder !== 'function') ***REMOVED***
        throw new TypeError('Encoder has to be a function.');
    ***REMOVED***

    var charset = opts.charset || defaults.charset;
    if (typeof opts.charset !== 'undefined' && opts.charset !== 'utf-8' && opts.charset !== 'iso-8859-1') ***REMOVED***
        throw new TypeError('The charset option must be either utf-8, iso-8859-1, or undefined');
    ***REMOVED***

    var format = formats['default'];
    if (typeof opts.format !== 'undefined') ***REMOVED***
        if (!has.call(formats.formatters, opts.format)) ***REMOVED***
            throw new TypeError('Unknown format option provided.');
        ***REMOVED***
        format = opts.format;
    ***REMOVED***
    var formatter = formats.formatters[format];

    var filter = defaults.filter;
    if (typeof opts.filter === 'function' || isArray(opts.filter)) ***REMOVED***
        filter = opts.filter;
    ***REMOVED***

    return ***REMOVED***
        addQueryPrefix: typeof opts.addQueryPrefix === 'boolean' ? opts.addQueryPrefix : defaults.addQueryPrefix,
        allowDots: typeof opts.allowDots === 'undefined' ? defaults.allowDots : !!opts.allowDots,
        charset: charset,
        charsetSentinel: typeof opts.charsetSentinel === 'boolean' ? opts.charsetSentinel : defaults.charsetSentinel,
        delimiter: typeof opts.delimiter === 'undefined' ? defaults.delimiter : opts.delimiter,
        encode: typeof opts.encode === 'boolean' ? opts.encode : defaults.encode,
        encoder: typeof opts.encoder === 'function' ? opts.encoder : defaults.encoder,
        encodeValuesOnly: typeof opts.encodeValuesOnly === 'boolean' ? opts.encodeValuesOnly : defaults.encodeValuesOnly,
        filter: filter,
        formatter: formatter,
        serializeDate: typeof opts.serializeDate === 'function' ? opts.serializeDate : defaults.serializeDate,
        skipNulls: typeof opts.skipNulls === 'boolean' ? opts.skipNulls : defaults.skipNulls,
        sort: typeof opts.sort === 'function' ? opts.sort : null,
        strictNullHandling: typeof opts.strictNullHandling === 'boolean' ? opts.strictNullHandling : defaults.strictNullHandling
    ***REMOVED***;
***REMOVED***;

module.exports = function (object, opts) ***REMOVED***
    var obj = object;
    var options = normalizeStringifyOptions(opts);

    var objKeys;
    var filter;

    if (typeof options.filter === 'function') ***REMOVED***
        filter = options.filter;
        obj = filter('', obj);
    ***REMOVED*** else if (isArray(options.filter)) ***REMOVED***
        filter = options.filter;
        objKeys = filter;
    ***REMOVED***

    var keys = [];

    if (typeof obj !== 'object' || obj === null) ***REMOVED***
        return '';
    ***REMOVED***

    var arrayFormat;
    if (opts && opts.arrayFormat in arrayPrefixGenerators) ***REMOVED***
        arrayFormat = opts.arrayFormat;
    ***REMOVED*** else if (opts && 'indices' in opts) ***REMOVED***
        arrayFormat = opts.indices ? 'indices' : 'repeat';
    ***REMOVED*** else ***REMOVED***
        arrayFormat = 'indices';
    ***REMOVED***

    var generateArrayPrefix = arrayPrefixGenerators[arrayFormat];

    if (!objKeys) ***REMOVED***
        objKeys = Object.keys(obj);
    ***REMOVED***

    if (options.sort) ***REMOVED***
        objKeys.sort(options.sort);
    ***REMOVED***

    for (var i = 0; i < objKeys.length; ++i) ***REMOVED***
        var key = objKeys[i];

        if (options.skipNulls && obj[key] === null) ***REMOVED***
            continue;
        ***REMOVED***
        pushToArray(keys, stringify(
            obj[key],
            key,
            generateArrayPrefix,
            options.strictNullHandling,
            options.skipNulls,
            options.encode ? options.encoder : null,
            options.filter,
            options.sort,
            options.allowDots,
            options.serializeDate,
            options.formatter,
            options.encodeValuesOnly,
            options.charset
        ));
    ***REMOVED***

    var joined = keys.join(options.delimiter);
    var prefix = options.addQueryPrefix === true ? '?' : '';

    if (options.charsetSentinel) ***REMOVED***
        if (options.charset === 'iso-8859-1') ***REMOVED***
            // encodeURIComponent('&#10003;'), the "numeric entity" representation of a checkmark
            prefix += 'utf8=%26%2310003%3B&';
        ***REMOVED*** else ***REMOVED***
            // encodeURIComponent('✓')
            prefix += 'utf8=%E2%9C%93&';
        ***REMOVED***
    ***REMOVED***

    return joined.length > 0 ? prefix + joined : '';
***REMOVED***;

***REMOVED***,***REMOVED***"./formats":1,"./utils":5***REMOVED***],5:[function(require,module,exports)***REMOVED***
'use strict';

var has = Object.prototype.hasOwnProperty;
var isArray = Array.isArray;

var hexTable = (function () ***REMOVED***
    var array = [];
    for (var i = 0; i < 256; ++i) ***REMOVED***
        array.push('%' + ((i < 16 ? '0' : '') + i.toString(16)).toUpperCase());
    ***REMOVED***

    return array;
***REMOVED***());

var compactQueue = function compactQueue(queue) ***REMOVED***
    while (queue.length > 1) ***REMOVED***
        var item = queue.pop();
        var obj = item.obj[item.prop];

        if (isArray(obj)) ***REMOVED***
            var compacted = [];

            for (var j = 0; j < obj.length; ++j) ***REMOVED***
                if (typeof obj[j] !== 'undefined') ***REMOVED***
                    compacted.push(obj[j]);
                ***REMOVED***
            ***REMOVED***

            item.obj[item.prop] = compacted;
        ***REMOVED***
    ***REMOVED***
***REMOVED***;

var arrayToObject = function arrayToObject(source, options) ***REMOVED***
    var obj = options && options.plainObjects ? Object.create(null) : ***REMOVED******REMOVED***;
    for (var i = 0; i < source.length; ++i) ***REMOVED***
        if (typeof source[i] !== 'undefined') ***REMOVED***
            obj[i] = source[i];
        ***REMOVED***
    ***REMOVED***

    return obj;
***REMOVED***;

var merge = function merge(target, source, options) ***REMOVED***
    /* eslint no-param-reassign: 0 */
    if (!source) ***REMOVED***
        return target;
    ***REMOVED***

    if (typeof source !== 'object') ***REMOVED***
        if (isArray(target)) ***REMOVED***
            target.push(source);
        ***REMOVED*** else if (target && typeof target === 'object') ***REMOVED***
            if ((options && (options.plainObjects || options.allowPrototypes)) || !has.call(Object.prototype, source)) ***REMOVED***
                target[source] = true;
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            return [target, source];
        ***REMOVED***

        return target;
    ***REMOVED***

    if (!target || typeof target !== 'object') ***REMOVED***
        return [target].concat(source);
    ***REMOVED***

    var mergeTarget = target;
    if (isArray(target) && !isArray(source)) ***REMOVED***
        mergeTarget = arrayToObject(target, options);
    ***REMOVED***

    if (isArray(target) && isArray(source)) ***REMOVED***
        source.forEach(function (item, i) ***REMOVED***
            if (has.call(target, i)) ***REMOVED***
                var targetItem = target[i];
                if (targetItem && typeof targetItem === 'object' && item && typeof item === 'object') ***REMOVED***
                    target[i] = merge(targetItem, item, options);
                ***REMOVED*** else ***REMOVED***
                    target.push(item);
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                target[i] = item;
            ***REMOVED***
        ***REMOVED***);
        return target;
    ***REMOVED***

    return Object.keys(source).reduce(function (acc, key) ***REMOVED***
        var value = source[key];

        if (has.call(acc, key)) ***REMOVED***
            acc[key] = merge(acc[key], value, options);
        ***REMOVED*** else ***REMOVED***
            acc[key] = value;
        ***REMOVED***
        return acc;
***REMOVED*** mergeTarget);
***REMOVED***;

var assign = function assignSingleSource(target, source) ***REMOVED***
    return Object.keys(source).reduce(function (acc, key) ***REMOVED***
        acc[key] = source[key];
        return acc;
***REMOVED*** target);
***REMOVED***;

var decode = function (str, decoder, charset) ***REMOVED***
    var strWithoutPlus = str.replace(/\+/g, ' ');
    if (charset === 'iso-8859-1') ***REMOVED***
        // unescape never throws, no try...catch needed:
        return strWithoutPlus.replace(/%[0-9a-f]***REMOVED***2***REMOVED***/gi, unescape);
    ***REMOVED***
    // utf-8
    try ***REMOVED***
        return decodeURIComponent(strWithoutPlus);
    ***REMOVED*** catch (e) ***REMOVED***
        return strWithoutPlus;
    ***REMOVED***
***REMOVED***;

var encode = function encode(str, defaultEncoder, charset) ***REMOVED***
    // This code was originally written by Brian White (mscdex) for the io.js core querystring library.
    // It has been adapted here for stricter adherence to RFC 3986
    if (str.length === 0) ***REMOVED***
        return str;
    ***REMOVED***

    var string = str;
    if (typeof str === 'symbol') ***REMOVED***
        string = Symbol.prototype.toString.call(str);
    ***REMOVED*** else if (typeof str !== 'string') ***REMOVED***
        string = String(str);
    ***REMOVED***

    if (charset === 'iso-8859-1') ***REMOVED***
        return escape(string).replace(/%u[0-9a-f]***REMOVED***4***REMOVED***/gi, function ($0) ***REMOVED***
            return '%26%23' + parseInt($0.slice(2), 16) + '%3B';
        ***REMOVED***);
    ***REMOVED***

    var out = '';
    for (var i = 0; i < string.length; ++i) ***REMOVED***
        var c = string.charCodeAt(i);

        if (
            c === 0x2D // -
            || c === 0x2E // .
            || c === 0x5F // _
            || c === 0x7E // ~
            || (c >= 0x30 && c <= 0x39) // 0-9
            || (c >= 0x41 && c <= 0x5A) // a-z
            || (c >= 0x61 && c <= 0x7A) // A-Z
        ) ***REMOVED***
            out += string.charAt(i);
            continue;
        ***REMOVED***

        if (c < 0x80) ***REMOVED***
            out = out + hexTable[c];
            continue;
        ***REMOVED***

        if (c < 0x800) ***REMOVED***
            out = out + (hexTable[0xC0 | (c >> 6)] + hexTable[0x80 | (c & 0x3F)]);
            continue;
        ***REMOVED***

        if (c < 0xD800 || c >= 0xE000) ***REMOVED***
            out = out + (hexTable[0xE0 | (c >> 12)] + hexTable[0x80 | ((c >> 6) & 0x3F)] + hexTable[0x80 | (c & 0x3F)]);
            continue;
        ***REMOVED***

        i += 1;
        c = 0x10000 + (((c & 0x3FF) << 10) | (string.charCodeAt(i) & 0x3FF));
        out += hexTable[0xF0 | (c >> 18)]
            + hexTable[0x80 | ((c >> 12) & 0x3F)]
            + hexTable[0x80 | ((c >> 6) & 0x3F)]
            + hexTable[0x80 | (c & 0x3F)];
    ***REMOVED***

    return out;
***REMOVED***;

var compact = function compact(value) ***REMOVED***
    var queue = [***REMOVED*** obj: ***REMOVED*** o: value ***REMOVED***, prop: 'o' ***REMOVED***];
    var refs = [];

    for (var i = 0; i < queue.length; ++i) ***REMOVED***
        var item = queue[i];
        var obj = item.obj[item.prop];

        var keys = Object.keys(obj);
        for (var j = 0; j < keys.length; ++j) ***REMOVED***
            var key = keys[j];
            var val = obj[key];
            if (typeof val === 'object' && val !== null && refs.indexOf(val) === -1) ***REMOVED***
                queue.push(***REMOVED*** obj: obj, prop: key ***REMOVED***);
                refs.push(val);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    compactQueue(queue);

    return value;
***REMOVED***;

var isRegExp = function isRegExp(obj) ***REMOVED***
    return Object.prototype.toString.call(obj) === '[object RegExp]';
***REMOVED***;

var isBuffer = function isBuffer(obj) ***REMOVED***
    if (!obj || typeof obj !== 'object') ***REMOVED***
        return false;
    ***REMOVED***

    return !!(obj.constructor && obj.constructor.isBuffer && obj.constructor.isBuffer(obj));
***REMOVED***;

var combine = function combine(a, b) ***REMOVED***
    return [].concat(a, b);
***REMOVED***;

module.exports = ***REMOVED***
    arrayToObject: arrayToObject,
    assign: assign,
    combine: combine,
    compact: compact,
    decode: decode,
    encode: encode,
    isBuffer: isBuffer,
    isRegExp: isRegExp,
    merge: merge
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***]***REMOVED***,***REMOVED******REMOVED***,[2])(2)
***REMOVED***);
