// twitch/lib/API/Helix/Game/HelixGameApi.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.HelixGameApi = void 0;
var tslib_1 = require("tslib");
var twitch_api_call_1 = require("twitch-api-call");
var twitch_common_1 = require("twitch-common");
var BaseApi_1 = require("../../BaseApi");
var HelixPaginatedRequest_1 = require("../HelixPaginatedRequest");
var HelixPaginatedResult_1 = require("../HelixPaginatedResult");
var HelixPagination_1 = require("../HelixPagination");
var HelixGame_1 = require("./HelixGame");
/**
 * The Helix API methods that deal with games.
 *
 * Can be accessed using `client.helix.games` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const game = await api.helix.games.getGameByName('Hearthstone');
 * ```
 */
var HelixGameApi = /** @class */ (function (_super) ***REMOVED***
    tslib_1.__extends(HelixGameApi, _super);
    function HelixGameApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves the game data for the given list of game IDs.
     *
     * @param ids The game IDs you want to look up.
     */
    HelixGameApi.prototype.getGamesByIds = function (ids) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                return [2 /*return*/, this._getGames('id', ids)];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the game data for the given list of game names.
     *
     * @param names The game names you want to look up.
     */
    HelixGameApi.prototype.getGamesByNames = function (names) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                return [2 /*return*/, this._getGames('name', names)];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the game data for the given game ID.
     *
     * @param id The game ID you want to look up.
     */
    HelixGameApi.prototype.getGameById = function (id) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var games;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._getGames('id', [id])];
                    case 1:
                        games = _a.sent();
                        return [2 /*return*/, games.length ? games[0] : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the game data for the given game name.
     *
     * @param name The game name you want to look up.
     */
    HelixGameApi.prototype.getGameByName = function (name) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var games;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._getGames('name', [name])];
                    case 1:
                        games = _a.sent();
                        return [2 /*return*/, games.length ? games[0] : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a list of the most viewed games at the moment.
     *
     * @param pagination
     *
     * @expandParams
     */
    HelixGameApi.prototype.getTopGames = function (pagination) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'games/top',
                            query: HelixPagination_1.makePaginationQuery(pagination)
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, HelixPaginatedResult_1.createPaginatedResult(result, HelixGame_1.HelixGame, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for the most viewed games at the moment.
     */
    HelixGameApi.prototype.getTopGamesPaginated = function () ***REMOVED***
        var _this = this;
        return new HelixPaginatedRequest_1.HelixPaginatedRequest(***REMOVED***
            url: 'games/top'
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixGame_1.HelixGame(data, _this._client); ***REMOVED***);
    ***REMOVED***;
    HelixGameApi.prototype._getGames = function (filterType, filterValues) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            var _a;
            var _this = this;
            return tslib_1.__generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        if (!filterValues.length) ***REMOVED***
                            return [2 /*return*/, []];
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: twitch_api_call_1.TwitchApiCallType.Helix,
                                url: 'games',
                                query: (_a = ***REMOVED******REMOVED***,
                                    _a[filterType] = filterValues,
                                    _a)
                            ***REMOVED***)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, result.data.map(function (entry) ***REMOVED*** return new HelixGame_1.HelixGame(entry, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixGameApi = tslib_1.__decorate([
        twitch_common_1.rtfm('twitch', 'HelixGameApi')
    ], HelixGameApi);
    return HelixGameApi;
***REMOVED***(BaseApi_1.BaseApi));
exports.HelixGameApi = HelixGameApi;
