// @d-fischer/logger/lib/CustomLoggerWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.CustomLoggerWrapper = void 0;
var shared_utils_1 = require("@d-fischer/shared-utils");
var getMinLogLevelFromEnv_1 = require("./getMinLogLevelFromEnv");
var LogLevel_1 = require("./LogLevel");
var CustomLoggerWrapper = /** @class */ (function () ***REMOVED***
    function CustomLoggerWrapper(_a) ***REMOVED***
        var name = _a.name, minLevel = _a.minLevel, custom = _a.custom;
        var _b;
        this._minLevel = (_b = shared_utils_1.mapOptional(minLevel, function (lv) ***REMOVED*** return LogLevel_1.resolveLogLevel(lv); ***REMOVED***)) !== null && _b !== void 0 ? _b : getMinLogLevelFromEnv_1.getMinLogLevelFromEnv(name);
        this._override = typeof custom === 'function' ? ***REMOVED*** log: custom ***REMOVED*** : custom;
    ***REMOVED***
    CustomLoggerWrapper.prototype.log = function (level, message) ***REMOVED***
        if (this._shouldLog(level)) ***REMOVED***
            this._override.log(level, message);
        ***REMOVED***
    ***REMOVED***;
    CustomLoggerWrapper.prototype.crit = function (message) ***REMOVED***
        if (!this._override.crit) ***REMOVED***
            this.log(LogLevel_1.LogLevel.CRITICAL, message);
        ***REMOVED***
        else if (this._shouldLog(LogLevel_1.LogLevel.CRITICAL)) ***REMOVED***
            this._override.crit(message);
        ***REMOVED***
    ***REMOVED***;
    CustomLoggerWrapper.prototype.error = function (message) ***REMOVED***
        if (!this._override.error) ***REMOVED***
            this.log(LogLevel_1.LogLevel.ERROR, message);
        ***REMOVED***
        else if (this._shouldLog(LogLevel_1.LogLevel.ERROR)) ***REMOVED***
            this._override.error(message);
        ***REMOVED***
    ***REMOVED***;
    CustomLoggerWrapper.prototype.warn = function (message) ***REMOVED***
        if (!this._override.warn) ***REMOVED***
            this.log(LogLevel_1.LogLevel.WARNING, message);
        ***REMOVED***
        else if (this._shouldLog(LogLevel_1.LogLevel.WARNING)) ***REMOVED***
            this._override.warn(message);
        ***REMOVED***
    ***REMOVED***;
    CustomLoggerWrapper.prototype.info = function (message) ***REMOVED***
        if (!this._override.info) ***REMOVED***
            this.log(LogLevel_1.LogLevel.INFO, message);
        ***REMOVED***
        else if (this._shouldLog(LogLevel_1.LogLevel.INFO)) ***REMOVED***
            this._override.info(message);
        ***REMOVED***
    ***REMOVED***;
    CustomLoggerWrapper.prototype.debug = function (message) ***REMOVED***
        if (!this._override.debug) ***REMOVED***
            this.log(LogLevel_1.LogLevel.DEBUG, message);
        ***REMOVED***
        else if (this._shouldLog(LogLevel_1.LogLevel.DEBUG)) ***REMOVED***
            this._override.debug(message);
        ***REMOVED***
    ***REMOVED***;
    CustomLoggerWrapper.prototype.trace = function (message) ***REMOVED***
        if (!this._override.trace) ***REMOVED***
            this.log(LogLevel_1.LogLevel.TRACE, message);
        ***REMOVED***
        else if (this._shouldLog(LogLevel_1.LogLevel.TRACE)) ***REMOVED***
            this._override.trace(message);
        ***REMOVED***
    ***REMOVED***;
    CustomLoggerWrapper.prototype._shouldLog = function (level) ***REMOVED***
        return this._minLevel === undefined || this._minLevel >= level;
    ***REMOVED***;
    return CustomLoggerWrapper;
***REMOVED***());
exports.CustomLoggerWrapper = CustomLoggerWrapper;
