// twitch/es/API/Helix/Subscriptions/HelixSubscriptionApi.mjs
import ***REMOVED*** __awaiter, __decorate, __extends, __generator ***REMOVED*** from "tslib";
import ***REMOVED*** HttpStatusCodeError, TwitchApiCallType ***REMOVED*** from 'twitch-api-call';
import ***REMOVED*** extractUserId, rtfm ***REMOVED*** from 'twitch-common';
import ***REMOVED*** BaseApi ***REMOVED*** from "../../BaseApi.mjs";
import ***REMOVED*** HelixPaginatedRequest ***REMOVED*** from "../HelixPaginatedRequest.mjs";
import ***REMOVED*** HelixPaginatedRequestWithTotal ***REMOVED*** from "../HelixPaginatedRequestWithTotal.mjs";
import ***REMOVED*** createPaginatedResult, createPaginatedResultWithTotal ***REMOVED*** from "../HelixPaginatedResult.mjs";
import ***REMOVED*** HelixSubscription ***REMOVED*** from "./HelixSubscription.mjs";
import ***REMOVED*** HelixSubscriptionEvent ***REMOVED*** from "./HelixSubscriptionEvent.mjs";
import ***REMOVED*** HelixUserSubscription ***REMOVED*** from "./HelixUserSubscription.mjs";
/**
 * The Helix API methods that deal with subscriptions.
 *
 * Can be accessed using `client.helix.subscriptions` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const subscription = await api.helix.subscriptions.getSubscriptionForUser('61369223', '125328655');
 * ```
 */
var HelixSubscriptionApi = /** @class */ (function (_super) ***REMOVED***
    __extends(HelixSubscriptionApi, _super);
    function HelixSubscriptionApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves a list of all subscriptions to a given broadcaster.
     *
     * @param broadcaster The broadcaster to list subscriptions to.
     */
    HelixSubscriptionApi.prototype.getSubscriptions = function (broadcaster) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            url: 'subscriptions',
                            scope: 'channel:read:subscriptions',
                            type: TwitchApiCallType.Helix,
                            query: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster)
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, createPaginatedResultWithTotal(result, HelixSubscription, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for all subscriptions to a given broadcaster.
     *
     * @param broadcaster The broadcaster to list subscriptions to.
     */
    HelixSubscriptionApi.prototype.getSubscriptionsPaginated = function (broadcaster) ***REMOVED***
        var _this = this;
        return new HelixPaginatedRequestWithTotal(***REMOVED***
            url: 'subscriptions',
            scope: 'channel:read:subscriptions',
            query: ***REMOVED***
                broadcaster_id: extractUserId(broadcaster)
            ***REMOVED***
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixSubscription(data, _this._client); ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the subset of the given user list that is subscribed to the given broadcaster.
     *
     * @param broadcaster The broadcaster to find subscriptions to.
     * @param users The users that should be checked for subscriptions.
     */
    HelixSubscriptionApi.prototype.getSubscriptionsForUsers = function (broadcaster, users) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            url: 'subscriptions',
                            scope: 'channel:read:subscriptions',
                            type: TwitchApiCallType.Helix,
                            query: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster),
                                user_id: users.map(extractUserId)
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.map(function (data) ***REMOVED*** return new HelixSubscription(data, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the subscription data for a given user to a given broadcaster.
     *
     * This checks with the authorization of a broadcaster.
     * If you only have the authorization of a user, check ***REMOVED***@HelixSubscriptionApi#checkUserSubscription***REMOVED***.
     *
     * @param broadcaster The broadcaster to check.
     * @param user The user to check.
     */
    HelixSubscriptionApi.prototype.getSubscriptionForUser = function (broadcaster, user) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var list;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.getSubscriptionsForUsers(broadcaster, [user])];
                    case 1:
                        list = _a.sent();
                        return [2 /*return*/, list.length ? list[0] : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the most recent subscription events for a given broadcaster.
     *
     * @param broadcaster The broadcaster to retrieve subscription events for.
     */
    HelixSubscriptionApi.prototype.getSubscriptionEventsForBroadcaster = function (broadcaster) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                return [2 /*return*/, this._getSubscriptionEvents('broadcaster_id', extractUserId(broadcaster))];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for the recent subscription events for a given broadcaster.
     *
     * @param broadcaster The broadcaster to retrieve subscription events for.
     */
    HelixSubscriptionApi.prototype.getSubscriptionEventsForBroadcasterPaginated = function (broadcaster) ***REMOVED***
        var _this = this;
        return new HelixPaginatedRequest(***REMOVED***
            url: 'subscriptions/events',
            scope: 'channel:read:subscriptions',
            query: ***REMOVED***
                broadcaster_id: extractUserId(broadcaster)
            ***REMOVED***
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixSubscriptionEvent(data, _this._client); ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a single subscription event by ID.
     *
     * @param id The event ID.
     */
    HelixSubscriptionApi.prototype.getSubscriptionEventById = function (id) ***REMOVED***
        var _a;
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var events;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._getSubscriptionEvents('id', id)];
                    case 1:
                        events = _b.sent();
                        return [2 /*return*/, (_a = events.data[0]) !== null && _a !== void 0 ? _a : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Checks if a given user is subscribed to a given broadcaster. Returns null if not subscribed.
     *
     * This checks with the authorization of a user.
     * If you only have the authorization of a broadcaster, check ***REMOVED***@HelixSubscriptionApi#getSubscriptionForUser***REMOVED***.
     *
     * @param user The broadcaster to check the user's subscription for.
     * @param broadcaster The user to check.
     */
    HelixSubscriptionApi.prototype.checkUserSubscription = function (user, broadcaster) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result, e_1;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'subscriptions/user',
                                scope: 'user:read:subscriptions',
                                query: ***REMOVED***
                                    broadcaster_id: extractUserId(broadcaster),
                                    user_id: extractUserId(user)
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixUserSubscription(result.data[0], this._client)];
                    case 2:
                        e_1 = _a.sent();
                        if (e_1 instanceof HttpStatusCodeError && e_1.statusCode === 404) ***REMOVED***
                            return [2 /*return*/, null];
                        ***REMOVED***
                        throw e_1;
                    case 3: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixSubscriptionApi.prototype._getSubscriptionEvents = function (by, id) ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            var _a;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'subscriptions/events',
                            scope: 'channel:read:subscriptions',
                            query: (_a = ***REMOVED******REMOVED***,
                                _a[by] = id,
                                _a)
                        ***REMOVED***)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, createPaginatedResult(result, HelixSubscriptionEvent, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixSubscriptionApi = __decorate([
        rtfm('twitch', 'HelixSubscriptionApi')
    ], HelixSubscriptionApi);
    return HelixSubscriptionApi;
***REMOVED***(BaseApi));
export ***REMOVED*** HelixSubscriptionApi ***REMOVED***;
