// twitch/es/API/Helix/Bits/HelixBitsApi.mjs
import ***REMOVED*** __awaiter, __decorate, __extends, __generator ***REMOVED*** from "tslib";
import ***REMOVED*** TwitchApiCallType ***REMOVED*** from 'twitch-api-call';
import ***REMOVED*** extractUserId, rtfm ***REMOVED*** from 'twitch-common';
import ***REMOVED*** BaseApi ***REMOVED*** from "../../BaseApi.mjs";
import ***REMOVED*** HelixBitsLeaderboard ***REMOVED*** from "./HelixBitsLeaderboard.mjs";
import ***REMOVED*** HelixCheermoteList ***REMOVED*** from "./HelixCheermoteList.mjs";
/**
 * The Helix API methods that deal with bits.
 *
 * Can be accessed using `client.helix.bits` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const leaderboard = await api.helix.bits.getLeaderboard(***REMOVED*** period: 'day' ***REMOVED***);
 * ```
 */
var HelixBitsApi = /** @class */ (function (_super) ***REMOVED***
    __extends(HelixBitsApi, _super);
    function HelixBitsApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves a bits leaderboard of your channel.
     *
     * @expandParams
     */
    HelixBitsApi.prototype.getLeaderboard = function (params) ***REMOVED***
        if (params === void 0) ***REMOVED*** params = ***REMOVED******REMOVED***; ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var _a, count, _b, period, startDate, contextUserId, result;
            return __generator(this, function (_c) ***REMOVED***
                switch (_c.label) ***REMOVED***
                    case 0:
                        _a = params.count, count = _a === void 0 ? 10 : _a, _b = params.period, period = _b === void 0 ? 'all' : _b, startDate = params.startDate, contextUserId = params.contextUserId;
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'bits/leaderboard',
                                scope: 'bits:read',
                                query: ***REMOVED***
                                    count: count.toString(),
                                    period: period,
                                    started_at: startDate ? startDate.toISOString() : undefined,
                                    user_id: contextUserId
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, new HelixBitsLeaderboard(result, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves all available cheermotes.
     *
     * @param broadcaster The broadcaster to include custom cheermotes of.
     *
     * If not given, only retrieves global cheermotes.
     */
    HelixBitsApi.prototype.getCheermotes = function (broadcaster) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var broadcasterId, result;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        broadcasterId = broadcaster ? extractUserId(broadcaster) : undefined;
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'bits/cheermotes',
                                query: ***REMOVED***
                                    broadcaster_id: broadcasterId
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixCheermoteList(result.data, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixBitsApi = __decorate([
        rtfm('twitch', 'HelixBitsApi')
    ], HelixBitsApi);
    return HelixBitsApi;
***REMOVED***(BaseApi));
export ***REMOVED*** HelixBitsApi ***REMOVED***;
