// twitch/es/API/Helix/Team/HelixTeamApi.mjs
import ***REMOVED*** __awaiter, __decorate, __extends, __generator ***REMOVED*** from "tslib";
import ***REMOVED*** HttpStatusCodeError, TwitchApiCallType ***REMOVED*** from 'twitch-api-call';
import ***REMOVED*** extractUserId, rtfm ***REMOVED*** from 'twitch-common';
import ***REMOVED*** BaseApi ***REMOVED*** from "../../BaseApi.mjs";
import ***REMOVED*** HelixTeam ***REMOVED*** from "./HelixTeam.mjs";
import ***REMOVED*** HelixTeamWithUsers ***REMOVED*** from "./HelixTeamWithUsers.mjs";
/**
 * The Helix API methods that deal with teams.
 *
 * Can be accessed using `client.helix.teams` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const tags = await api.helix.teams.getChannelTeams('125328655');
 * ```
 */
var HelixTeamApi = /** @class */ (function (_super) ***REMOVED***
    __extends(HelixTeamApi, _super);
    function HelixTeamApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves a list of all teams a broadcaster is a member of.
     *
     * @param broadcaster The broadcaster to retrieve the teams of.
     */
    HelixTeamApi.prototype.getTeamsForBroadcaster = function (broadcaster) ***REMOVED***
        var _a, _b;
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            var _this = this;
            return __generator(this, function (_c) ***REMOVED***
                switch (_c.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'teams/channel',
                            query: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster)
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, (_b = (_a = result.data) === null || _a === void 0 ? void 0 : _a.map(function (data) ***REMOVED*** return new HelixTeam(data, _this._client); ***REMOVED***)) !== null && _b !== void 0 ? _b : []];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a team by ID.
     *
     * Returns null if there is no team with the given ID.
     *
     * @param id The ID of the team.
     */
    HelixTeamApi.prototype.getTeamById = function (id) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result, e_1;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'teams',
                                query: ***REMOVED***
                                    id: id
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixTeamWithUsers(result.data[0], this._client)];
                    case 2:
                        e_1 = _a.sent();
                        // Twitch, please...
                        if (e_1 instanceof HttpStatusCodeError && e_1.statusCode === 500) ***REMOVED***
                            return [2 /*return*/, null];
                        ***REMOVED***
                        throw e_1;
                    case 3: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a team by name.
     *
     * Returns null if there is no team with the given name.
     *
     * @param name The name of the team.
     */
    HelixTeamApi.prototype.getTeamByName = function (name) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result, e_2;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'teams',
                                query: ***REMOVED***
                                    name: name
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixTeamWithUsers(result.data[0], this._client)];
                    case 2:
                        e_2 = _a.sent();
                        // ...but this one is fine
                        if (e_2 instanceof HttpStatusCodeError && e_2.statusCode === 404) ***REMOVED***
                            return [2 /*return*/, null];
                        ***REMOVED***
                        throw e_2;
                    case 3: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixTeamApi = __decorate([
        rtfm('twitch', 'HelixTeamApi')
    ], HelixTeamApi);
    return HelixTeamApi;
***REMOVED***(BaseApi));
export ***REMOVED*** HelixTeamApi ***REMOVED***;
