// twitch/es/API/Kraken/Channel/ChannelApi.mjs
import ***REMOVED*** __awaiter, __decorate, __extends, __generator ***REMOVED*** from "tslib";
import ***REMOVED*** Cacheable, Cached, ClearsCache ***REMOVED*** from '@d-fischer/cache-decorators';
import ***REMOVED*** HttpStatusCodeError ***REMOVED*** from 'twitch-api-call';
import ***REMOVED*** extractUserId, rtfm ***REMOVED*** from 'twitch-common';
import ***REMOVED*** NoSubscriptionProgramError ***REMOVED*** from "../../../Errors/NoSubscriptionProgramError.mjs";
import ***REMOVED*** BaseApi ***REMOVED*** from "../../BaseApi.mjs";
import ***REMOVED*** Team ***REMOVED*** from "../Team/Team.mjs";
import ***REMOVED*** User ***REMOVED*** from "../User/User.mjs";
import ***REMOVED*** Channel ***REMOVED*** from "./Channel.mjs";
import ***REMOVED*** ChannelFollow ***REMOVED*** from "./ChannelFollow.mjs";
import ***REMOVED*** ChannelSubscription ***REMOVED*** from "./ChannelSubscription.mjs";
import ***REMOVED*** PrivilegedChannel ***REMOVED*** from "./PrivilegedChannel.mjs";
/**
 * The API methods that deal with channels.
 *
 * Can be accessed using `client.kraken.channels` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const channel = await api.kraken.channels.getMyChannel();
 * ```
 */
var ChannelApi = /** @class */ (function (_super) ***REMOVED***
    __extends(ChannelApi, _super);
    function ChannelApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Gets the channel the client is logged in to.
     */
    ChannelApi.prototype.getMyChannel = function () ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var _a;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        _a = PrivilegedChannel.bind;
                        return [4 /*yield*/, this._client.callApi(***REMOVED*** url: 'channel', scope: 'channel_read' ***REMOVED***)];
                    case 1: return [2 /*return*/, new (_a.apply(PrivilegedChannel, [void 0, _b.sent(), this._client]))()];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the channel for the given user.
     *
     * @param user The user you want to retrieve the channel for.
     */
    ChannelApi.prototype.getChannel = function (user) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var _a;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        _a = Channel.bind;
                        return [4 /*yield*/, this._client.callApi(***REMOVED*** url: "channels/" + extractUserId(user) ***REMOVED***)];
                    case 1: return [2 /*return*/, new (_a.apply(Channel, [void 0, _b.sent(), this._client]))()];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Updates the given channel with the given data.
     *
     * @param channel The channel you want to update.
     * @param data The updated channel data.
     */
    ChannelApi.prototype.updateChannel = function (channel, data) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var channelId;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        channelId = extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "channels/" + channelId,
                                method: 'PUT',
                                jsonBody: ***REMOVED*** channel: data ***REMOVED***,
                                scope: 'channel_editor'
                            ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the list of users that have editor rights to the given channel.
     *
     * @param channel The channel you want to retrieve the list of editors for.
     */
    ChannelApi.prototype.getChannelEditors = function (channel) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var channelId, data;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        channelId = extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "channels/" + channelId + "/editors",
                                scope: 'channel_read'
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.users.map(function (userData) ***REMOVED*** return new User(userData, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the list of followers of the given channel.
     *
     * @param channel The channel you want to retrieve the list of followers of.
     * @param page The result page you want to retrieve.
     * @param limit The number of results you want to retrieve.
     * @param orderDirection The direction to order in - ascending or descending.
     */
    ChannelApi.prototype.getChannelFollowers = function (channel, page, limit, orderDirection) ***REMOVED***
        if (limit === void 0) ***REMOVED*** limit = 25; ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var channelId, query, data;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        channelId = extractUserId(channel);
                        query = ***REMOVED*** limit: limit.toString() ***REMOVED***;
                        if (page) ***REMOVED***
                            query.offset = ((page - 1) * limit).toString();
                        ***REMOVED***
                        if (orderDirection) ***REMOVED***
                            query.direction = orderDirection;
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "channels/" + channelId + "/follows",
                                query: query
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.follows.map(function (followData) ***REMOVED*** return new ChannelFollow(followData, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the list of subscribers of the given channel.
     *
     * @param channel The channel you want to retrieve the list of subscribers of.
     * @param page The result page you want to retrieve.
     * @param limit The number of results you want to retrieve.
     * @param orderDirection The direction to order in - ascending or descending.
     */
    ChannelApi.prototype.getChannelSubscriptions = function (channel, page, limit, orderDirection) ***REMOVED***
        if (limit === void 0) ***REMOVED*** limit = 25; ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var data;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._getChannelSubscriptions(channel, page, limit, orderDirection)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.subscriptions.map(function (sub) ***REMOVED*** return new ChannelSubscription(sub, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the total number of subscribers for the given channel.
     *
     * @param channel The channel you want to retrieve the number of subscribers for.
     */
    ChannelApi.prototype.getChannelSubscriptionCount = function (channel) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var data;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._getChannelSubscriptions(channel, 0, 1)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data._total];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the subscription data for the given user to a given channel.
     *
     * Throws if the channel doesn't have a subscription program or the user is not subscribed to it.
     *
     * This method requires access to the channel. If you only have access to the user,
     * use ***REMOVED***@UserApi#getSubscriptionData***REMOVED*** instead.
     *
     * @param channel The channel to check the subscription to.
     * @param byUser The user to check the subscription for.
     */
    ChannelApi.prototype.getChannelSubscriptionByUser = function (channel, byUser) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var channelId, userId, _a, e_1;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        channelId = extractUserId(channel);
                        userId = extractUserId(byUser);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        _a = ChannelSubscription.bind;
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "channels/" + channelId + "/subscriptions/" + userId,
                                scope: 'channel_check_subscription'
                            ***REMOVED***)];
                    case 2: return [2 /*return*/, new (_a.apply(ChannelSubscription, [void 0, _b.sent(), this._client]))()];
                    case 3:
                        e_1 = _b.sent();
                        if (e_1 instanceof HttpStatusCodeError) ***REMOVED***
                            if (e_1.statusCode === 404) ***REMOVED***
                                return [2 /*return*/, null];
                            ***REMOVED***
                            else if (e_1.statusCode === 422) ***REMOVED***
                                throw new NoSubscriptionProgramError(channelId);
                            ***REMOVED***
                        ***REMOVED***
                        throw e_1;
                    case 4: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a list of teams of the given channel.
     *
     * @param channel The channel you want to retrieve the list of teams of.
     */
    ChannelApi.prototype.getChannelTeams = function (channel) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var channelId, data;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        channelId = extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "channels/" + channelId + "/teams",
                                method: 'GET'
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.teams.map(function (teamData) ***REMOVED*** return new Team(teamData, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Starts a commercial in the given channel.
     *
     * @param channel The channel to start the commercial in.
     * @param length The length of the commercial.
     */
    ChannelApi.prototype.startChannelCommercial = function (channel, length) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var channelId;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        channelId = extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "channels/" + channelId + "/commercial",
                                method: 'POST',
                                jsonBody: ***REMOVED*** length: length ***REMOVED***,
                                scope: 'channel_commercial'
                            ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Resets the given channel's stream key.
     *
     * @param channel The channel to reset the stream key for.
     */
    ChannelApi.prototype.resetChannelStreamKey = function (channel) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var channelId, channelData;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        channelId = extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "channels/" + channelId + "/stream_key",
                                method: 'DELETE',
                                scope: 'channel_stream'
                            ***REMOVED***)];
                    case 1:
                        channelData = _a.sent();
                        return [2 /*return*/, new PrivilegedChannel(channelData, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    ChannelApi.prototype._getChannelSubscriptions = function (channel, page, limit, orderDirection) ***REMOVED***
        if (limit === void 0) ***REMOVED*** limit = 25; ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var channelId, query, e_2;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        channelId = extractUserId(channel);
                        query = ***REMOVED*** limit: limit.toString() ***REMOVED***;
                        if (page) ***REMOVED***
                            query.offset = ((page - 1) * limit).toString();
                        ***REMOVED***
                        if (orderDirection) ***REMOVED***
                            query.direction = orderDirection;
                        ***REMOVED***
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "channels/" + channelId + "/subscriptions",
                                query: query,
                                scope: 'channel_subscriptions'
                            ***REMOVED***)];
                    case 2: return [2 /*return*/, _a.sent()];
                    case 3:
                        e_2 = _a.sent();
                        if (e_2 instanceof HttpStatusCodeError && e_2.statusCode === 422) ***REMOVED***
                            throw new NoSubscriptionProgramError(channelId);
                        ***REMOVED***
                        throw e_2;
                    case 4: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    __decorate([
        Cached(3600)
    ], ChannelApi.prototype, "getMyChannel", null);
    __decorate([
        Cached(3600)
    ], ChannelApi.prototype, "getChannel", null);
    __decorate([
        ClearsCache('getChannel', 1)
    ], ChannelApi.prototype, "updateChannel", null);
    __decorate([
        Cached(3600)
    ], ChannelApi.prototype, "getChannelEditors", null);
    __decorate([
        Cached(30)
    ], ChannelApi.prototype, "getChannelFollowers", null);
    __decorate([
        Cached(30)
    ], ChannelApi.prototype, "getChannelSubscriptions", null);
    __decorate([
        Cached(30)
    ], ChannelApi.prototype, "getChannelSubscriptionCount", null);
    __decorate([
        Cached(3600)
    ], ChannelApi.prototype, "getChannelSubscriptionByUser", null);
    __decorate([
        Cached(30)
    ], ChannelApi.prototype, "getChannelTeams", null);
    __decorate([
        ClearsCache('getMyChannel')
    ], ChannelApi.prototype, "resetChannelStreamKey", null);
    ChannelApi = __decorate([
        Cacheable,
        rtfm('twitch', 'ChannelApi')
    ], ChannelApi);
    return ChannelApi;
***REMOVED***(BaseApi));
export ***REMOVED*** ChannelApi ***REMOVED***;
