// twitch/es/API/Helix/Channel/HelixChannelApi.mjs
import ***REMOVED*** __awaiter, __decorate, __extends, __generator ***REMOVED*** from "tslib";
import ***REMOVED*** TwitchApiCallType ***REMOVED*** from 'twitch-api-call';
import ***REMOVED*** extractUserId, rtfm ***REMOVED*** from 'twitch-common';
import ***REMOVED*** BaseApi ***REMOVED*** from "../../BaseApi.mjs";
import ***REMOVED*** HelixChannel ***REMOVED*** from "./HelixChannel.mjs";
import ***REMOVED*** HelixChannelEditor ***REMOVED*** from "./HelixChannelEditor.mjs";
/**
 * The Helix API methods that deal with channels.
 *
 * Can be accessed using `client.helix.channels` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const channel = await api.helix.channels.getChannelInfo('125328655');
 * ```
 */
var HelixChannelApi = /** @class */ (function (_super) ***REMOVED***
    __extends(HelixChannelApi, _super);
    function HelixChannelApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves the channel data for the given user.
     *
     * @param user The user you want to get channel info for.
     */
    HelixChannelApi.prototype.getChannelInfo = function (user) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, result;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'channels',
                                query: ***REMOVED***
                                    broadcaster_id: userId
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.length ? new HelixChannel(result.data[0], this._client) : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Updates the given user's channel data.
     *
     * @param user The user you want to update channel info for.
     * @param data The channel info to set.
     */
    HelixChannelApi.prototype.updateChannelInfo = function (user, data) ***REMOVED***
        var _a;
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'channels',
                                method: 'PATCH',
                                scope: 'user:edit:broadcast',
                                query: ***REMOVED***
                                    broadcaster_id: userId
                            ***REMOVED***
                                jsonBody: ***REMOVED***
                                    game_id: data.gameId,
                                    broadcaster_language: data.language,
                                    title: data.title,
                                    delay: (_a = data.delay) === null || _a === void 0 ? void 0 : _a.toString()
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        _b.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Starts a commercial on a channel.
     *
     * @param broadcaster The broadcaster on whose channel the commercial is started.
     * @param length The length of the commercial, in seconds.
     */
    HelixChannelApi.prototype.startChannelCommercial = function (broadcaster, length) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'channels/commercial',
                            method: 'POST',
                            scope: 'channel:edit:commercial',
                            jsonBody: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster),
                                length: length
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a list of users who have editor permissions on your channel.
     */
    HelixChannelApi.prototype.getChannelEditors = function (broadcaster) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'channels/editors',
                            scope: 'channel:read:editors',
                            query: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster)
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.map(function (data) ***REMOVED*** return new HelixChannelEditor(data, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixChannelApi = __decorate([
        rtfm('twitch', 'HelixChannelApi')
    ], HelixChannelApi);
    return HelixChannelApi;
***REMOVED***(BaseApi));
export ***REMOVED*** HelixChannelApi ***REMOVED***;
