// twitch/lib/API/Helix/Search/HelixSearchApi.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.HelixSearchApi = void 0;
var tslib_1 = require("tslib");
var twitch_api_call_1 = require("twitch-api-call");
var twitch_common_1 = require("twitch-common");
var BaseApi_1 = require("../../BaseApi");
var HelixGame_1 = require("../Game/HelixGame");
var HelixPaginatedRequest_1 = require("../HelixPaginatedRequest");
var HelixPaginatedResult_1 = require("../HelixPaginatedResult");
var HelixPagination_1 = require("../HelixPagination");
var HelixChannelSearchResult_1 = require("./HelixChannelSearchResult");
/**
 * The Helix API methods that run searches.
 *
 * Can be accessed using `client.helix.search` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const channels = await api.helix.search.searchChannels('pear');
 * ```
 */
var HelixSearchApi = /** @class */ (function (_super) ***REMOVED***
    tslib_1.__extends(HelixSearchApi, _super);
    function HelixSearchApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Search categories/games for an exact or partial match.
     *
     * @param query The search term.
     * @param pagination
     *
     * @expandParams
     */
    HelixSearchApi.prototype.searchCategories = function (query, pagination) ***REMOVED***
        if (pagination === void 0) ***REMOVED*** pagination = ***REMOVED******REMOVED***; ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'search/categories',
                            query: tslib_1.__assign(***REMOVED*** query: query ***REMOVED***, HelixPagination_1.makePaginationQuery(pagination))
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, HelixPaginatedResult_1.createPaginatedResult(result, HelixGame_1.HelixGame, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for a category/game search.
     *
     * @param query The search term.
     */
    HelixSearchApi.prototype.searchCategoriesPaginated = function (query) ***REMOVED***
        var _this = this;
        return new HelixPaginatedRequest_1.HelixPaginatedRequest(***REMOVED***
            url: 'search/categories',
            query: ***REMOVED***
                query: query
            ***REMOVED***
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixGame_1.HelixGame(data, _this._client); ***REMOVED***);
    ***REMOVED***;
    /**
     * Search channels for an exact or partial match.
     *
     * @param query The search term.
     * @param filter
     *
     * @expandParams
     */
    HelixSearchApi.prototype.searchChannels = function (query, filter) ***REMOVED***
        var _a;
        if (filter === void 0) ***REMOVED*** filter = ***REMOVED******REMOVED***; ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            return tslib_1.__generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'search/channels',
                            query: tslib_1.__assign(***REMOVED*** query: query, live_only: (_a = filter.liveOnly) === null || _a === void 0 ? void 0 : _a.toString() ***REMOVED***, HelixPagination_1.makePaginationQuery(filter))
                        ***REMOVED***)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, HelixPaginatedResult_1.createPaginatedResult(result, HelixChannelSearchResult_1.HelixChannelSearchResult, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for a channel search.
     *
     * @param query The search term.
     * @param filter
     *
     * @expandParams
     */
    HelixSearchApi.prototype.searchChannelsPaginated = function (query, filter) ***REMOVED***
        var _this = this;
        var _a;
        if (filter === void 0) ***REMOVED*** filter = ***REMOVED******REMOVED***; ***REMOVED***
        return new HelixPaginatedRequest_1.HelixPaginatedRequest(***REMOVED***
            url: 'search/channels',
            query: ***REMOVED***
                query: query,
                live_only: (_a = filter.liveOnly) === null || _a === void 0 ? void 0 : _a.toString()
            ***REMOVED***
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixChannelSearchResult_1.HelixChannelSearchResult(data, _this._client); ***REMOVED***);
    ***REMOVED***;
    HelixSearchApi = tslib_1.__decorate([
        twitch_common_1.rtfm('twitch', 'HelixSearchApi')
    ], HelixSearchApi);
    return HelixSearchApi;
***REMOVED***(BaseApi_1.BaseApi));
exports.HelixSearchApi = HelixSearchApi;
