// twitch-chat-client/lib/StandardCommands/TwitchPrivateMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.TwitchPrivateMessage = void 0;
var tslib_1 = require("tslib");
var ircv3_1 = require("ircv3");
var twitch_common_1 = require("twitch-common");
var ChatUser_1 = require("../ChatUser");
var EmoteTools_1 = require("../Toolkit/EmoteTools");
/**
 * An IRC PRIVMSG, with easy accessors for commonly used data from its tags.
 */
var TwitchPrivateMessage = /** @class */ (function (_super) ***REMOVED***
    tslib_1.__extends(TwitchPrivateMessage, _super);
    function TwitchPrivateMessage() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Object.defineProperty(TwitchPrivateMessage.prototype, "id", ***REMOVED***
        /**
         * The ID of the message.
         */
        get: function () ***REMOVED***
            return this._tags.get('id');
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "userInfo", ***REMOVED***
        /**
         * Info about the user that send the message, like their user ID and their status in the current channel.
         */
        get: function () ***REMOVED***
            return new ChatUser_1.ChatUser(this._prefix.nick, this._tags);
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "channelId", ***REMOVED***
        /**
         * The ID of the channel the message is in.
         */
        get: function () ***REMOVED***
            var _a;
            return (_a = this._tags.get('room-id')) !== null && _a !== void 0 ? _a : null;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "isCheer", ***REMOVED***
        /**
         * Whether the message is a cheer.
         */
        get: function () ***REMOVED***
            return this._tags.has('bits');
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "bits", ***REMOVED***
        /**
         * The number of bits cheered with the message.
         */
        get: function () ***REMOVED***
            var _a;
            return Number((_a = this._tags.get('bits')) !== null && _a !== void 0 ? _a : 0);
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "totalBits", ***REMOVED***
        /**
         * The number of bits cheered with the message.
         *
         * @deprecated Use ***REMOVED***@TwitchPrivateMessage#bits***REMOVED*** instead.
         */
        get: function () ***REMOVED***
            return this.bits;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "emoteOffsets", ***REMOVED***
        /**
         * The offsets of emote usages in the message.
         */
        get: function () ***REMOVED***
            return EmoteTools_1.parseEmoteOffsets(this._tags.get('emotes'));
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    /**
     * Parses the message, separating text from emote usages.
     */
    TwitchPrivateMessage.prototype.parseEmotes = function () ***REMOVED***
        var messageText = this.params.message;
        var foundEmotes = EmoteTools_1.parseEmotePositions(messageText, this.emoteOffsets);
        return EmoteTools_1.fillTextPositions(messageText, foundEmotes);
    ***REMOVED***;
    /**
     * Parses the message, separating text from emote usages and cheers.
     *
     * @param cheermotes A list of cheermotes
     */
    TwitchPrivateMessage.prototype.parseEmotesAndBits = function (cheermotes) ***REMOVED***
        var messageText = this.params.message;
        var foundCheermotes = cheermotes.parseMessage(messageText);
        var foundEmotesAndCheermotes = tslib_1.__spread(EmoteTools_1.parseEmotePositions(messageText, this.emoteOffsets), foundCheermotes.map(function (cheermote) ***REMOVED*** return (***REMOVED***
            type: 'cheer',
            position: cheermote.position,
            length: cheermote.length,
            name: cheermote.name,
            amount: cheermote.amount,
            displayInfo: cheermote.displayInfo
        ***REMOVED***); ***REMOVED***));
        foundEmotesAndCheermotes.sort(function (a, b) ***REMOVED*** return a.position - b.position; ***REMOVED***);
        return EmoteTools_1.fillTextPositions(messageText, foundEmotesAndCheermotes);
    ***REMOVED***;
    TwitchPrivateMessage = tslib_1.__decorate([
        twitch_common_1.rtfm('twitch-chat-client', 'TwitchPrivateMessage', 'id')
    ], TwitchPrivateMessage);
    return TwitchPrivateMessage;
***REMOVED***(ircv3_1.MessageTypes.Commands.PrivateMessage));
exports.TwitchPrivateMessage = TwitchPrivateMessage;
