// twitch/lib/API/Kraken/User/UserApi.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.UserApi = void 0;
var tslib_1 = require("tslib");
var cache_decorators_1 = require("@d-fischer/cache-decorators");
var shared_utils_1 = require("@d-fischer/shared-utils");
var twitch_api_call_1 = require("twitch-api-call");
var twitch_common_1 = require("twitch-common");
var NoSubscriptionProgramError_1 = require("../../../Errors/NoSubscriptionProgramError");
var BaseApi_1 = require("../../BaseApi");
var EmoteSetList_1 = require("../Channel/EmoteSetList");
var PrivilegedUser_1 = require("./PrivilegedUser");
var User_1 = require("./User");
var UserBlock_1 = require("./UserBlock");
var UserChatInfo_1 = require("./UserChatInfo");
var UserFollow_1 = require("./UserFollow");
var UserSubscription_1 = require("./UserSubscription");
/**
 * The API methods that deal with users.
 *
 * Can be accessed using `client.kraken.users` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const user = await api.kraken.users.getUser('125328655');
 * ```
 */
var UserApi = /** @class */ (function (_super) ***REMOVED***
    tslib_1.__extends(UserApi, _super);
    function UserApi() ***REMOVED***
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._userByNameCache = new Map();
        return _this;
    ***REMOVED***
    /**
     * Retrieves the user data of the currently authenticated user.
     */
    UserApi.prototype.getMe = function () ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var _a;
            return tslib_1.__generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        _a = PrivilegedUser_1.PrivilegedUser.bind;
                        return [4 /*yield*/, this._client.callApi(***REMOVED*** url: 'user', scope: 'user_read' ***REMOVED***)];
                    case 1: return [2 /*return*/, new (_a.apply(PrivilegedUser_1.PrivilegedUser, [void 0, _b.sent(), this._client]))()];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the user data for the given user ID.
     *
     * @param userId The user ID you want to look up.
     */
    UserApi.prototype.getUser = function (userId) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userData;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED*** url: "users/" + twitch_common_1.extractUserId(userId) ***REMOVED***)];
                    case 1:
                        userData = _a.sent();
                        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                        if (!userData) ***REMOVED***
                            throw new twitch_common_1.HellFreezesOverError('Could not get authenticated user');
                        ***REMOVED***
                        return [2 /*return*/, new User_1.User(userData, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the user data for the given user name.
     *
     * @param userName The user name you want to look up.
     */
    UserApi.prototype.getUserByName = function (userName) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var users, user;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        // not using the decorator's cache here as users-by-name is slightly more complex to cache
                        this._cleanUserCache();
                        if (this._userByNameCache.has(userName)) ***REMOVED***
                            return [2 /*return*/, this._userByNameCache.get(userName).value];
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: 'users',
                                query: ***REMOVED*** login: userName ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        users = (_a.sent()).users;
                        if (users.length === 0) ***REMOVED***
                            return [2 /*return*/, null];
                        ***REMOVED***
                        user = new User_1.User(users[0], this._client);
                        this._userByNameCache.set(userName, ***REMOVED***
                            value: user,
                            expires: Date.now() + 3600 * 1000
                        ***REMOVED***);
                        return [2 /*return*/, user];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the user data for the given user names.
     *
     * @param userNames The user names you want to look up.
     */
    UserApi.prototype.getUsersByNames = function (userNames) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var cachedEntries, cachedObject, cachedUsers, toFetch, usersData, usersArr, users;
            var _this = this;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        this._cleanUserCache();
                        userNames = userNames.map(function (name) ***REMOVED*** return name.toLowerCase(); ***REMOVED***);
                        cachedEntries = Array.from(this._userByNameCache.entries()).filter(function (_a) ***REMOVED***
                            var _b = tslib_1.__read(_a, 1), key = _b[0];
                            return userNames.includes(key);
                        ***REMOVED***);
                        cachedObject = shared_utils_1.entriesToObject(cachedEntries);
                        cachedUsers = shared_utils_1.mapObject(cachedObject, function (entry) ***REMOVED*** return entry.value; ***REMOVED***);
                        toFetch = userNames.filter(function (name) ***REMOVED*** return !(name in cachedUsers); ***REMOVED***);
                        if (!toFetch.length) ***REMOVED***
                            return [2 /*return*/, cachedUsers];
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: 'users',
                                query: ***REMOVED*** login: toFetch.join(',') ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        usersData = _a.sent();
                        usersArr = usersData.users.map(function (data) ***REMOVED*** return new User_1.User(data, _this._client); ***REMOVED***);
                        usersArr.forEach(function (user) ***REMOVED***
                            return _this._userByNameCache.set(user.name, ***REMOVED***
                                value: user,
                                expires: Date.now() + 3600 * 1000
                            ***REMOVED***);
                        ***REMOVED***);
                        users = shared_utils_1.indexBy(usersArr, 'name');
                        return [2 /*return*/, tslib_1.__assign(tslib_1.__assign(***REMOVED******REMOVED***, cachedUsers), users)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves information about the user's chat appearance and privileges.
     *
     * @param user The user you want to get chat info for.
     */
    UserApi.prototype.getChatInfo = function (user) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, data;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        return [4 /*yield*/, this._client.callApi(***REMOVED*** url: "users/" + userId + "/chat" ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, new UserChatInfo_1.UserChatInfo(data, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the emotes a given user can use.
     *
     * @param user The user you want to get emotes for.
     */
    UserApi.prototype.getUserEmotes = function (user) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, data;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/emotes",
                                scope: 'user_subscriptions'
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, new EmoteSetList_1.EmoteSetList(data.emoticon_sets)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the subscription data for a given user to a given channel.
     *
     * @param user The user to retrieve the subscription data of.
     * @param toChannel The channel you want to retrieve the subscription data to.
     */
    UserApi.prototype.getSubscriptionData = function (user, toChannel) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, channelId, _a, e_1;
            return tslib_1.__generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        channelId = twitch_common_1.extractUserId(toChannel);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        _a = UserSubscription_1.UserSubscription.bind;
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/subscriptions/" + channelId,
                                scope: 'user_subscriptions'
                            ***REMOVED***)];
                    case 2: return [2 /*return*/, new (_a.apply(UserSubscription_1.UserSubscription, [void 0, _b.sent(), this._client]))()];
                    case 3:
                        e_1 = _b.sent();
                        if (e_1 instanceof twitch_api_call_1.HttpStatusCodeError) ***REMOVED***
                            if (e_1.statusCode === 404) ***REMOVED***
                                return [2 /*return*/, null];
                            ***REMOVED***
                            else if (e_1.statusCode === 422) ***REMOVED***
                                throw new NoSubscriptionProgramError_1.NoSubscriptionProgramError(channelId);
                            ***REMOVED***
                        ***REMOVED***
                        throw e_1;
                    case 4: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Get a list of channels a given user follows.
     *
     * @param user The user you want to retrieve the follows of.
     * @param page The result page you want to retrieve.
     * @param limit The number of results you want to retrieve.
     * @param orderBy The field to order by.
     * @param orderDirection The direction to order in - ascending or descending.
     */
    UserApi.prototype.getFollowedChannels = function (user, page, limit, orderBy, orderDirection) ***REMOVED***
        if (limit === void 0) ***REMOVED*** limit = 25; ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, query, data;
            var _this = this;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        query = ***REMOVED******REMOVED***;
                        if (page) ***REMOVED***
                            query.offset = ((page - 1) * limit).toString();
                        ***REMOVED***
                        query.limit = limit.toString();
                        if (orderBy) ***REMOVED***
                            query.sortby = orderBy;
                        ***REMOVED***
                        if (orderDirection) ***REMOVED***
                            query.direction = orderDirection;
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/follows/channels",
                                query: query
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.follows.map(function (follow) ***REMOVED*** return new UserFollow_1.UserFollow(follow, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Get follow data for a given user to a given channel.
     *
     * @param user The user you want to retrieve follow data of.
     * @param channel The channel you want to retrieve follow data to.
     */
    UserApi.prototype.getFollowedChannel = function (user, channel) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, channelId, data, e_2;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        channelId = twitch_common_1.extractUserId(channel);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/follows/channels/" + channelId
                            ***REMOVED***)];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, new UserFollow_1.UserFollow(data, this._client)];
                    case 3:
                        e_2 = _a.sent();
                        if (e_2 instanceof twitch_api_call_1.HttpStatusCodeError) ***REMOVED***
                            if (e_2.statusCode === 404) ***REMOVED***
                                return [2 /*return*/, null];
                            ***REMOVED***
                        ***REMOVED***
                        throw e_2;
                    case 4: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Follows a given channel with a given user.
     *
     * @param user The user you want to follow with.
     * @param channel The channel to follow.
     * @param notifications Whether the user will receive notifications.
     */
    UserApi.prototype.followChannel = function (user, channel, notifications) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, channelId, data;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        channelId = twitch_common_1.extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/follows/channels/" + channelId,
                                method: 'PUT',
                                scope: 'user_follows_edit',
                                body: ***REMOVED*** notifications: Boolean(notifications).toString() ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, new UserFollow_1.UserFollow(data, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Unfollows a given channel with a given user.
     *
     * @param user The user you want to unfollow with.
     * @param channel The channel to unfollow.
     */
    UserApi.prototype.unfollowChannel = function (user, channel) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, channelId;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        channelId = twitch_common_1.extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/follows/channels/" + channelId,
                                scope: 'user_follows_edit',
                                method: 'DELETE'
                            ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a list of users a given user has blocked.
     *
     * @param user The user you want to retrieve the block list of.
     * @param page The result page you want to retrieve.
     * @param limit The number of results you want to retrieve.
     */
    UserApi.prototype.getBlockedUsers = function (user, page, limit) ***REMOVED***
        if (limit === void 0) ***REMOVED*** limit = 25; ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, query, data;
            var _this = this;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        query = ***REMOVED*** limit: limit.toString() ***REMOVED***;
                        if (page) ***REMOVED***
                            query.offset = ((page - 1) * limit).toString();
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/blocks",
                                query: query,
                                scope: 'user_blocks_read'
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.blocks.map(function (block) ***REMOVED*** return new UserBlock_1.UserBlock(block, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Blocks a given user with another given user.
     *
     * @param user The user you want to block with.
     * @param userToBlock The user to block.
     */
    UserApi.prototype.blockUser = function (user, userToBlock) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, userIdToBlock, data;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        userIdToBlock = twitch_common_1.extractUserId(userToBlock);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/blocks/" + userIdToBlock,
                                method: 'PUT',
                                scope: 'user_blocks_edit'
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, new UserBlock_1.UserBlock(data, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Unblocks a given user with another given user.
     *
     * @param user The user you want to unblock with.
     * @param userToUnblock The user to unblock.
     */
    UserApi.prototype.unblockUser = function (user, userToUnblock) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, userIdToUnblock;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        userIdToUnblock = twitch_common_1.extractUserId(userToUnblock);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/blocks/" + userIdToUnblock,
                                method: 'DELETE',
                                scope: 'user_blocks_edit'
                            ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    UserApi.prototype._cleanUserCache = function () ***REMOVED***
        var _this = this;
        var now = Date.now();
        this._userByNameCache.forEach(function (val, key) ***REMOVED***
            if (val.expires < now) ***REMOVED***
                _this._userByNameCache.delete(key);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***;
    tslib_1.__decorate([
        cache_decorators_1.Cached(3600)
    ], UserApi.prototype, "getMe", null);
    tslib_1.__decorate([
        cache_decorators_1.Cached(3600)
    ], UserApi.prototype, "getUser", null);
    tslib_1.__decorate([
        cache_decorators_1.Cached(3600)
    ], UserApi.prototype, "getChatInfo", null);
    tslib_1.__decorate([
        cache_decorators_1.Cached(3600)
    ], UserApi.prototype, "getUserEmotes", null);
    tslib_1.__decorate([
        cache_decorators_1.Cached(3600)
    ], UserApi.prototype, "getSubscriptionData", null);
    tslib_1.__decorate([
        cache_decorators_1.Cached(300)
    ], UserApi.prototype, "getFollowedChannels", null);
    tslib_1.__decorate([
        cache_decorators_1.Cached(300)
    ], UserApi.prototype, "getFollowedChannel", null);
    tslib_1.__decorate([
        cache_decorators_1.ClearsCache('getFollowedChannels', 1),
        cache_decorators_1.ClearsCache('getFollowedChannel', 2)
    ], UserApi.prototype, "followChannel", null);
    tslib_1.__decorate([
        cache_decorators_1.ClearsCache('getFollowedChannels', 1),
        cache_decorators_1.ClearsCache('getFollowedChannel', 2)
    ], UserApi.prototype, "unfollowChannel", null);
    tslib_1.__decorate([
        cache_decorators_1.Cached(3600)
    ], UserApi.prototype, "getBlockedUsers", null);
    tslib_1.__decorate([
        cache_decorators_1.ClearsCache('getBlockedUsers', 1)
    ], UserApi.prototype, "blockUser", null);
    tslib_1.__decorate([
        cache_decorators_1.ClearsCache('getBlockedUsers', 1)
    ], UserApi.prototype, "unblockUser", null);
    UserApi = tslib_1.__decorate([
        cache_decorators_1.Cacheable,
        twitch_common_1.rtfm('twitch', 'UserApi')
    ], UserApi);
    return UserApi;
***REMOVED***(BaseApi_1.BaseApi));
exports.UserApi = UserApi;
