// @d-fischer/logger/es/BaseLogger.mjs
import ***REMOVED*** LogLevel, resolveLogLevel ***REMOVED*** from "./LogLevel.mjs";
import * as isNode from 'detect-node';
var BaseLogger = /** @class */ (function () ***REMOVED***
    function BaseLogger(_a) ***REMOVED***
        var name = _a.name, _b = _a.minLevel, minLevel = _b === void 0 ? LogLevel.WARNING : _b, _c = _a.emoji, emoji = _c === void 0 ? false : _c, _d = _a.colors, colors = _d === void 0 ? true : _d, _e = _a.timestamps, timestamps = _e === void 0 ? isNode : _e;
        this._name = name;
        this._minLevel = resolveLogLevel(minLevel);
        this._emoji = emoji;
        this._colors = colors;
        this._timestamps = timestamps;
    ***REMOVED***
    // region convenience methods
    BaseLogger.prototype.crit = function (message) ***REMOVED***
        this.log(LogLevel.CRITICAL, message);
    ***REMOVED***;
    BaseLogger.prototype.error = function (message) ***REMOVED***
        this.log(LogLevel.ERROR, message);
    ***REMOVED***;
    BaseLogger.prototype.warn = function (message) ***REMOVED***
        this.log(LogLevel.WARNING, message);
    ***REMOVED***;
    BaseLogger.prototype.info = function (message) ***REMOVED***
        this.log(LogLevel.INFO, message);
    ***REMOVED***;
    BaseLogger.prototype.debug = function (message) ***REMOVED***
        this.log(LogLevel.DEBUG, message);
    ***REMOVED***;
    BaseLogger.prototype.trace = function (message) ***REMOVED***
        this.log(LogLevel.TRACE, message);
    ***REMOVED***;
    return BaseLogger;
***REMOVED***());
export ***REMOVED*** BaseLogger ***REMOVED***;
