// twitch/es/API/Helix/Schedule/HelixPaginatedScheduleSegmentRequest.mjs
import ***REMOVED*** __awaiter, __extends, __generator ***REMOVED*** from "tslib";
import ***REMOVED*** extractUserId ***REMOVED*** from 'twitch-common';
import ***REMOVED*** HelixPaginatedRequest ***REMOVED*** from "../HelixPaginatedRequest.mjs";
import ***REMOVED*** HelixScheduleSegment ***REMOVED*** from "./HelixScheduleSegment.mjs";
/**
 * A paginator specifically for schedule segments.
 */
var HelixPaginatedScheduleSegmentRequest = /** @class */ (function (_super) ***REMOVED***
    __extends(HelixPaginatedScheduleSegmentRequest, _super);
    /** @private */
    function HelixPaginatedScheduleSegmentRequest(broadcaster, client, filter) ***REMOVED***
        var _a;
        return _super.call(this, ***REMOVED***
            url: 'schedule',
            query: ***REMOVED***
                broadcaster_id: extractUserId(broadcaster),
                start_time: filter === null || filter === void 0 ? void 0 : filter.startDate,
                utc_offset: (_a = filter === null || filter === void 0 ? void 0 : filter.utcOffset) === null || _a === void 0 ? void 0 : _a.toString()
            ***REMOVED***
    ***REMOVED*** client, function (data) ***REMOVED*** return new HelixScheduleSegment(data, client); ***REMOVED***, 25) || this;
    ***REMOVED***
    // sadly, this hack is necessary to work around the weird data model of schedules
    // while still keeping the pagination code as generic as possible
    /** @private */
    HelixPaginatedScheduleSegmentRequest.prototype._fetchData = function (additionalOptions) ***REMOVED***
        if (additionalOptions === void 0) ***REMOVED*** additionalOptions = ***REMOVED******REMOVED***; ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var origData;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, _super.prototype._fetchData.call(this, additionalOptions)];
                    case 1:
                        origData = (_a.sent());
                        return [2 /*return*/, ***REMOVED***
                                data: origData.data.segments,
                                pagination: origData.pagination
                            ***REMOVED***];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    return HelixPaginatedScheduleSegmentRequest;
***REMOVED***(HelixPaginatedRequest));
export ***REMOVED*** HelixPaginatedScheduleSegmentRequest ***REMOVED***;
