// twitch/es/API/Helix/Schedule/HelixScheduleApi.mjs
import ***REMOVED*** __assign, __awaiter, __extends, __generator ***REMOVED*** from "tslib";
import ***REMOVED*** TwitchApiCallType ***REMOVED*** from 'twitch-api-call';
import ***REMOVED*** extractUserId ***REMOVED*** from 'twitch-common';
import ***REMOVED*** BaseApi ***REMOVED*** from "../../BaseApi.mjs";
import ***REMOVED*** makePaginationQuery ***REMOVED*** from "../HelixPagination.mjs";
import ***REMOVED*** HelixPaginatedScheduleSegmentRequest ***REMOVED*** from "./HelixPaginatedScheduleSegmentRequest.mjs";
import ***REMOVED*** HelixSchedule ***REMOVED*** from "./HelixSchedule.mjs";
import ***REMOVED*** HelixScheduleSegment ***REMOVED*** from "./HelixScheduleSegment.mjs";
/**
 * The Helix API methods that deal with schedules.
 *
 * Can be accessed using `client.helix.schedule` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const ***REMOVED*** data: schedule ***REMOVED*** = await api.helix.schedule.getSchedule('61369223');
 */
var HelixScheduleApi = /** @class */ (function (_super) ***REMOVED***
    __extends(HelixScheduleApi, _super);
    function HelixScheduleApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves the schedule for a given broadcaster.
     *
     * @param broadcaster The broadcaster to get the schedule of.
     * @param filter
     *
     * @expandParams
     */
    HelixScheduleApi.prototype.getSchedule = function (broadcaster, filter) ***REMOVED***
        var _a;
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'schedule',
                            query: __assign(***REMOVED*** broadcaster_id: extractUserId(broadcaster), start_time: filter === null || filter === void 0 ? void 0 : filter.startDate, utc_offset: (_a = filter === null || filter === void 0 ? void 0 : filter.utcOffset) === null || _a === void 0 ? void 0 : _a.toString() ***REMOVED***, makePaginationQuery(filter))
                        ***REMOVED***)];
                    case 1:
                        result = _b.sent();
                        return [2 /*return*/, ***REMOVED***
                                data: new HelixSchedule(result.data, this._client),
                                cursor: result.pagination.cursor
                            ***REMOVED***];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for schedule segments for a given broadcaster.
     *
     * @param broadcaster The broadcaster to get the schedule segments of.
     * @param filter
     *
     * @expandParams
     */
    HelixScheduleApi.prototype.getScheduleSegmentsPaginated = function (broadcaster, filter) ***REMOVED***
        return new HelixPaginatedScheduleSegmentRequest(broadcaster, this._client, filter);
    ***REMOVED***;
    /**
     * Retrieves a set of schedule segments by IDs.
     *
     * @param broadcaster The broadcaster to get schedule segments of.
     * @param ids The IDs of the schedule segments.
     */
    HelixScheduleApi.prototype.getScheduleSegmentsByIds = function (broadcaster, ids) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'schedule',
                            query: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster),
                                id: ids
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.segments.map(function (data) ***REMOVED*** return new HelixScheduleSegment(data, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a single schedule segment by ID.
     *
     * @param broadcaster The broadcaster to get a schedule segment of.
     * @param id The ID of the schedule segment.
     */
    HelixScheduleApi.prototype.getScheduleSegmentById = function (broadcaster, id) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var segments;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.getScheduleSegmentsByIds(broadcaster, [id])];
                    case 1:
                        segments = _a.sent();
                        return [2 /*return*/, segments.length ? segments[0] : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the schedule for a given broadcaster in iCal format.
     *
     * @param broadcaster The broadcaster to get the schedule for.
     */
    HelixScheduleApi.prototype.getScheduleAsIcal = function (broadcaster) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                return [2 /*return*/, this._client.callApi(***REMOVED***
                        type: TwitchApiCallType.Helix,
                        url: 'schedule/icalendar',
                        query: ***REMOVED***
                            broadcaster_id: extractUserId(broadcaster)
                        ***REMOVED***
                    ***REMOVED***)];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Updates the schedule settings of a given broadcaster.
     *
     * @param broadcaster The broadcaster to update the schedule settings for.
     * @param settings
     *
     * @expandParams
     */
    HelixScheduleApi.prototype.updateScheduleSettings = function (broadcaster, settings) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var vacationUpdateQuery;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        vacationUpdateQuery = settings.vacation
                            ? ***REMOVED***
                                is_vacation_enabled: 'true',
                                vacation_start_time: settings.vacation.startDate,
                                vacation_end_time: settings.vacation.endDate,
                                timezone: settings.vacation.timezone
                            ***REMOVED***
                            : ***REMOVED***
                                is_vacation_enabled: 'false'
                            ***REMOVED***;
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'schedule/settings',
                                method: 'PATCH',
                                scope: 'channel:manage:schedule',
                                query: __assign(***REMOVED*** broadcaster_id: extractUserId(broadcaster) ***REMOVED***, vacationUpdateQuery)
                            ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a new segment in a given broadcaster's schedule.
     *
     * @param broadcaster The broadcaster to create a new schedule segment for.
     * @param data
     *
     * @expandParams
     */
    HelixScheduleApi.prototype.createScheduleSegment = function (broadcaster, data) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'schedule/segment',
                            method: 'POST',
                            scope: 'channel:manage:schedule',
                            query: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster)
                        ***REMOVED***
                            jsonBody: ***REMOVED***
                                start_time: data.startDate,
                                timezone: data.timezone,
                                is_recurring: data.isRecurring,
                                duration: data.duration,
                                category_id: data.categoryId,
                                title: data.title
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixScheduleSegment(result.data.segments[0], this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Updates a segment in a given broadcaster's schedule.
     *
     * @param broadcaster The broadcaster to create a new schedule segment for.
     * @param segmentId The ID of the segment to update.
     * @param data
     *
     * @expandParams
     */
    HelixScheduleApi.prototype.updateScheduleSegment = function (broadcaster, segmentId, data) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'schedule/segment',
                            method: 'PATCH',
                            scope: 'channel:manage:schedule',
                            query: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster),
                                id: segmentId
                        ***REMOVED***
                            jsonBody: ***REMOVED***
                                start_time: data.startDate,
                                timezone: data.timezone,
                                is_canceled: data.isCanceled,
                                duration: data.duration,
                                category_id: data.categoryId,
                                title: data.title
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixScheduleSegment(result.data.segments[0], this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Deletes a segment in a given broadcaster's schedule.
     *
     * @param broadcaster The broadcaster to create a new schedule segment for.
     * @param segmentId The ID of the segment to update.
     */
    HelixScheduleApi.prototype.deleteScheduleSegment = function (broadcaster, segmentId) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'schedule/segment',
                            method: 'DELETE',
                            scope: 'channel:manage:schedule',
                            query: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster),
                                id: segmentId
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    return HelixScheduleApi;
***REMOVED***(BaseApi));
export ***REMOVED*** HelixScheduleApi ***REMOVED***;
