// twitch/lib/API/Helix/Channel/HelixChannelApi.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.HelixChannelApi = void 0;
var tslib_1 = require("tslib");
var twitch_api_call_1 = require("twitch-api-call");
var twitch_common_1 = require("twitch-common");
var BaseApi_1 = require("../../BaseApi");
var HelixChannel_1 = require("./HelixChannel");
var HelixChannelEditor_1 = require("./HelixChannelEditor");
/**
 * The Helix API methods that deal with channels.
 *
 * Can be accessed using `client.helix.channels` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const channel = await api.helix.channels.getChannelInfo('125328655');
 * ```
 */
var HelixChannelApi = /** @class */ (function (_super) ***REMOVED***
    tslib_1.__extends(HelixChannelApi, _super);
    function HelixChannelApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves the channel data for the given user.
     *
     * @param user The user you want to get channel info for.
     */
    HelixChannelApi.prototype.getChannelInfo = function (user) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId, result;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: twitch_api_call_1.TwitchApiCallType.Helix,
                                url: 'channels',
                                query: ***REMOVED***
                                    broadcaster_id: userId
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.length ? new HelixChannel_1.HelixChannel(result.data[0], this._client) : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Updates the given user's channel data.
     *
     * @param user The user you want to update channel info for.
     * @param data The channel info to set.
     */
    HelixChannelApi.prototype.updateChannelInfo = function (user, data) ***REMOVED***
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var userId;
            return tslib_1.__generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        userId = twitch_common_1.extractUserId(user);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: twitch_api_call_1.TwitchApiCallType.Helix,
                                url: 'channels',
                                method: 'PATCH',
                                scope: 'user:edit:broadcast',
                                query: ***REMOVED***
                                    broadcaster_id: userId
                            ***REMOVED***
                                jsonBody: ***REMOVED***
                                    game_id: data.gameId,
                                    broadcaster_language: data.language,
                                    title: data.title,
                                    delay: (_a = data.delay) === null || _a === void 0 ? void 0 : _a.toString()
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        _b.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Starts a commercial on a channel.
     *
     * @param broadcaster The broadcaster on whose channel the commercial is started.
     * @param length The length of the commercial, in seconds.
     */
    HelixChannelApi.prototype.startChannelCommercial = function (broadcaster, length) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'channels/commercial',
                            method: 'POST',
                            scope: 'channel:edit:commercial',
                            jsonBody: ***REMOVED***
                                broadcaster_id: twitch_common_1.extractUserId(broadcaster),
                                length: length
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a list of users who have editor permissions on your channel.
     */
    HelixChannelApi.prototype.getChannelEditors = function (broadcaster) ***REMOVED***
        return tslib_1.__awaiter(this, void 0, void 0, function () ***REMOVED***
            var result;
            var _this = this;
            return tslib_1.__generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: twitch_api_call_1.TwitchApiCallType.Helix,
                            url: 'channels/editors',
                            scope: 'channel:read:editors',
                            query: ***REMOVED***
                                broadcaster_id: twitch_common_1.extractUserId(broadcaster)
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.map(function (data) ***REMOVED*** return new HelixChannelEditor_1.HelixChannelEditor(data, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixChannelApi = tslib_1.__decorate([
        twitch_common_1.rtfm('twitch', 'HelixChannelApi')
    ], HelixChannelApi);
    return HelixChannelApi;
***REMOVED***(BaseApi_1.BaseApi));
exports.HelixChannelApi = HelixChannelApi;
