// twitch-auth/es/legacy.mjs
import ***REMOVED*** __awaiter, __generator ***REMOVED*** from "tslib";
import deprecate from '@d-fischer/deprecate';
import ***REMOVED*** callTwitchApi, TwitchApiCallType ***REMOVED*** from 'twitch-api-call';
/** @deprecated Use `exchangeCode` instead. */
export function getUserAccessToken(creds, code, redirectUri) ***REMOVED***
    if (redirectUri === void 0) ***REMOVED*** redirectUri = 'http://localhost'; ***REMOVED***
    return __awaiter(this, void 0, Promise, function () ***REMOVED***
        return __generator(this, function (_a) ***REMOVED***
            deprecate('[twitch-auth] getUserAccessToken', 'Use exchangeCode instead.');
            return [2 /*return*/, callTwitchApi(***REMOVED***
                    type: TwitchApiCallType.Auth,
                    url: 'token',
                    method: 'POST',
                    query: ***REMOVED***
                        grant_type: 'authorization_code',
                        client_id: creds.client_id,
                        client_secret: creds.client_secret,
                        code: code,
                        redirect_uri: redirectUri
                    ***REMOVED***
                ***REMOVED***)];
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***
/** @deprecated Use `refreshUserToken` instead. */
export function refreshUserAccessToken(creds, refreshToken) ***REMOVED***
    return __awaiter(this, void 0, Promise, function () ***REMOVED***
        return __generator(this, function (_a) ***REMOVED***
            deprecate('[twitch-auth] refreshUserAccessToken', 'Use refreshUserToken instead.');
            return [2 /*return*/, callTwitchApi(***REMOVED***
                    type: TwitchApiCallType.Auth,
                    url: 'token',
                    method: 'POST',
                    query: ***REMOVED***
                        grant_type: 'refresh_token',
                        client_id: creds.client_id,
                        client_secret: creds.client_secret,
                        refresh_token: refreshToken
                    ***REMOVED***
                ***REMOVED***)];
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***
/** @deprecated Use `getAppToken` instead. */
export function getAppAccessToken(creds) ***REMOVED***
    return __awaiter(this, void 0, Promise, function () ***REMOVED***
        return __generator(this, function (_a) ***REMOVED***
            deprecate('[twitch-auth] getAppAccessToken', 'Use getAppToken instead.');
            return [2 /*return*/, callTwitchApi(***REMOVED***
                    type: TwitchApiCallType.Auth,
                    url: 'token',
                    method: 'POST',
                    query: ***REMOVED***
                        grant_type: 'client_credentials',
                        client_id: creds.client_id,
                        client_secret: creds.client_secret
                    ***REMOVED***
                ***REMOVED***)];
        ***REMOVED***);
    ***REMOVED***);
***REMOVED***
