// node-fetch/lib/index.js
'use strict';

Object.defineProperty(exports, '__esModule', ***REMOVED*** value: true ***REMOVED***);

function _interopDefault (ex) ***REMOVED*** return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; ***REMOVED***

var Stream = _interopDefault(require('stream'));
var http = _interopDefault(require('http'));
var Url = _interopDefault(require('url'));
var whatwgUrl = _interopDefault(require('whatwg-url'));
var https = _interopDefault(require('https'));
var zlib = _interopDefault(require('zlib'));

// Based on https://github.com/tmpvar/jsdom/blob/aa85b2abf07766ff7bf5c1f6daafb3726f2f2db5/lib/jsdom/living/blob.js

// fix for "Readable" isn't a named export issue
const Readable = Stream.Readable;

const BUFFER = Symbol('buffer');
const TYPE = Symbol('type');

class Blob ***REMOVED***
	constructor() ***REMOVED***
		this[TYPE] = '';

		const blobParts = arguments[0];
		const options = arguments[1];

		const buffers = [];
		let size = 0;

		if (blobParts) ***REMOVED***
			const a = blobParts;
			const length = Number(a.length);
			for (let i = 0; i < length; i++) ***REMOVED***
				const element = a[i];
				let buffer;
				if (element instanceof Buffer) ***REMOVED***
					buffer = element;
				***REMOVED*** else if (ArrayBuffer.isView(element)) ***REMOVED***
					buffer = Buffer.from(element.buffer, element.byteOffset, element.byteLength);
				***REMOVED*** else if (element instanceof ArrayBuffer) ***REMOVED***
					buffer = Buffer.from(element);
				***REMOVED*** else if (element instanceof Blob) ***REMOVED***
					buffer = element[BUFFER];
				***REMOVED*** else ***REMOVED***
					buffer = Buffer.from(typeof element === 'string' ? element : String(element));
				***REMOVED***
				size += buffer.length;
				buffers.push(buffer);
			***REMOVED***
		***REMOVED***

		this[BUFFER] = Buffer.concat(buffers);

		let type = options && options.type !== undefined && String(options.type).toLowerCase();
		if (type && !/[^\u0020-\u007E]/.test(type)) ***REMOVED***
			this[TYPE] = type;
		***REMOVED***
	***REMOVED***
	get size() ***REMOVED***
		return this[BUFFER].length;
	***REMOVED***
	get type() ***REMOVED***
		return this[TYPE];
	***REMOVED***
	text() ***REMOVED***
		return Promise.resolve(this[BUFFER].toString());
	***REMOVED***
	arrayBuffer() ***REMOVED***
		const buf = this[BUFFER];
		const ab = buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
		return Promise.resolve(ab);
	***REMOVED***
	stream() ***REMOVED***
		const readable = new Readable();
		readable._read = function () ***REMOVED******REMOVED***;
		readable.push(this[BUFFER]);
		readable.push(null);
		return readable;
	***REMOVED***
	toString() ***REMOVED***
		return '[object Blob]';
	***REMOVED***
	slice() ***REMOVED***
		const size = this.size;

		const start = arguments[0];
		const end = arguments[1];
		let relativeStart, relativeEnd;
		if (start === undefined) ***REMOVED***
			relativeStart = 0;
		***REMOVED*** else if (start < 0) ***REMOVED***
			relativeStart = Math.max(size + start, 0);
		***REMOVED*** else ***REMOVED***
			relativeStart = Math.min(start, size);
		***REMOVED***
		if (end === undefined) ***REMOVED***
			relativeEnd = size;
		***REMOVED*** else if (end < 0) ***REMOVED***
			relativeEnd = Math.max(size + end, 0);
		***REMOVED*** else ***REMOVED***
			relativeEnd = Math.min(end, size);
		***REMOVED***
		const span = Math.max(relativeEnd - relativeStart, 0);

		const buffer = this[BUFFER];
		const slicedBuffer = buffer.slice(relativeStart, relativeStart + span);
		const blob = new Blob([], ***REMOVED*** type: arguments[2] ***REMOVED***);
		blob[BUFFER] = slicedBuffer;
		return blob;
	***REMOVED***
***REMOVED***

Object.defineProperties(Blob.prototype, ***REMOVED***
	size: ***REMOVED*** enumerable: true ***REMOVED***,
	type: ***REMOVED*** enumerable: true ***REMOVED***,
	slice: ***REMOVED*** enumerable: true ***REMOVED***
***REMOVED***);

Object.defineProperty(Blob.prototype, Symbol.toStringTag, ***REMOVED***
	value: 'Blob',
	writable: false,
	enumerable: false,
	configurable: true
***REMOVED***);

/**
 * fetch-error.js
 *
 * FetchError interface for operational errors
 */

/**
 * Create FetchError instance
 *
 * @param   String      message      Error message for human
 * @param   String      type         Error type for machine
 * @param   String      systemError  For Node.js system error
 * @return  FetchError
 */
function FetchError(message, type, systemError) ***REMOVED***
  Error.call(this, message);

  this.message = message;
  this.type = type;

  // when err.type is `system`, err.code contains system error code
  if (systemError) ***REMOVED***
    this.code = this.errno = systemError.code;
  ***REMOVED***

  // hide custom error implementation details from end-users
  Error.captureStackTrace(this, this.constructor);
***REMOVED***

FetchError.prototype = Object.create(Error.prototype);
FetchError.prototype.constructor = FetchError;
FetchError.prototype.name = 'FetchError';

let convert;
try ***REMOVED***
	convert = require('encoding').convert;
***REMOVED*** catch (e) ***REMOVED******REMOVED***

const INTERNALS = Symbol('Body internals');

// fix an issue where "PassThrough" isn't a named export for node <10
const PassThrough = Stream.PassThrough;

/**
 * Body mixin
 *
 * Ref: https://fetch.spec.whatwg.org/#body
 *
 * @param   Stream  body  Readable stream
 * @param   Object  opts  Response options
 * @return  Void
 */
function Body(body) ***REMOVED***
	var _this = this;

	var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : ***REMOVED******REMOVED***,
	    _ref$size = _ref.size;

	let size = _ref$size === undefined ? 0 : _ref$size;
	var _ref$timeout = _ref.timeout;
	let timeout = _ref$timeout === undefined ? 0 : _ref$timeout;

	if (body == null) ***REMOVED***
		// body is undefined or null
		body = null;
	***REMOVED*** else if (isURLSearchParams(body)) ***REMOVED***
		// body is a URLSearchParams
		body = Buffer.from(body.toString());
	***REMOVED*** else if (isBlob(body)) ; else if (Buffer.isBuffer(body)) ; else if (Object.prototype.toString.call(body) === '[object ArrayBuffer]') ***REMOVED***
		// body is ArrayBuffer
		body = Buffer.from(body);
	***REMOVED*** else if (ArrayBuffer.isView(body)) ***REMOVED***
		// body is ArrayBufferView
		body = Buffer.from(body.buffer, body.byteOffset, body.byteLength);
	***REMOVED*** else if (body instanceof Stream) ; else ***REMOVED***
		// none of the above
		// coerce to string then buffer
		body = Buffer.from(String(body));
	***REMOVED***
	this[INTERNALS] = ***REMOVED***
		body,
		disturbed: false,
		error: null
	***REMOVED***;
	this.size = size;
	this.timeout = timeout;

	if (body instanceof Stream) ***REMOVED***
		body.on('error', function (err) ***REMOVED***
			const error = err.name === 'AbortError' ? err : new FetchError(`Invalid response body while trying to fetch $***REMOVED***_this.url***REMOVED***: $***REMOVED***err.message***REMOVED***`, 'system', err);
			_this[INTERNALS].error = error;
		***REMOVED***);
	***REMOVED***
***REMOVED***

Body.prototype = ***REMOVED***
	get body() ***REMOVED***
		return this[INTERNALS].body;
	***REMOVED***,

	get bodyUsed() ***REMOVED***
		return this[INTERNALS].disturbed;
	***REMOVED***,

	/**
  * Decode response as ArrayBuffer
  *
  * @return  Promise
  */
	arrayBuffer() ***REMOVED***
		return consumeBody.call(this).then(function (buf) ***REMOVED***
			return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
		***REMOVED***);
	***REMOVED***,

	/**
  * Return raw response as Blob
  *
  * @return Promise
  */
	blob() ***REMOVED***
		let ct = this.headers && this.headers.get('content-type') || '';
		return consumeBody.call(this).then(function (buf) ***REMOVED***
			return Object.assign(
			// Prevent copying
			new Blob([], ***REMOVED***
				type: ct.toLowerCase()
			***REMOVED***), ***REMOVED***
				[BUFFER]: buf
			***REMOVED***);
		***REMOVED***);
	***REMOVED***,

	/**
  * Decode response as json
  *
  * @return  Promise
  */
	json() ***REMOVED***
		var _this2 = this;

		return consumeBody.call(this).then(function (buffer) ***REMOVED***
			try ***REMOVED***
				return JSON.parse(buffer.toString());
			***REMOVED*** catch (err) ***REMOVED***
				return Body.Promise.reject(new FetchError(`invalid json response body at $***REMOVED***_this2.url***REMOVED*** reason: $***REMOVED***err.message***REMOVED***`, 'invalid-json'));
			***REMOVED***
		***REMOVED***);
	***REMOVED***,

	/**
  * Decode response as text
  *
  * @return  Promise
  */
	text() ***REMOVED***
		return consumeBody.call(this).then(function (buffer) ***REMOVED***
			return buffer.toString();
		***REMOVED***);
	***REMOVED***,

	/**
  * Decode response as buffer (non-spec api)
  *
  * @return  Promise
  */
	buffer() ***REMOVED***
		return consumeBody.call(this);
	***REMOVED***,

	/**
  * Decode response as text, while automatically detecting the encoding and
  * trying to decode to UTF-8 (non-spec api)
  *
  * @return  Promise
  */
	textConverted() ***REMOVED***
		var _this3 = this;

		return consumeBody.call(this).then(function (buffer) ***REMOVED***
			return convertBody(buffer, _this3.headers);
		***REMOVED***);
	***REMOVED***
***REMOVED***;

// In browsers, all properties are enumerable.
Object.defineProperties(Body.prototype, ***REMOVED***
	body: ***REMOVED*** enumerable: true ***REMOVED***,
	bodyUsed: ***REMOVED*** enumerable: true ***REMOVED***,
	arrayBuffer: ***REMOVED*** enumerable: true ***REMOVED***,
	blob: ***REMOVED*** enumerable: true ***REMOVED***,
	json: ***REMOVED*** enumerable: true ***REMOVED***,
	text: ***REMOVED*** enumerable: true ***REMOVED***
***REMOVED***);

Body.mixIn = function (proto) ***REMOVED***
	for (const name of Object.getOwnPropertyNames(Body.prototype)) ***REMOVED***
		// istanbul ignore else: future proof
		if (!(name in proto)) ***REMOVED***
			const desc = Object.getOwnPropertyDescriptor(Body.prototype, name);
			Object.defineProperty(proto, name, desc);
		***REMOVED***
	***REMOVED***
***REMOVED***;

/**
 * Consume and convert an entire Body to a Buffer.
 *
 * Ref: https://fetch.spec.whatwg.org/#concept-body-consume-body
 *
 * @return  Promise
 */
function consumeBody() ***REMOVED***
	var _this4 = this;

	if (this[INTERNALS].disturbed) ***REMOVED***
		return Body.Promise.reject(new TypeError(`body used already for: $***REMOVED***this.url***REMOVED***`));
	***REMOVED***

	this[INTERNALS].disturbed = true;

	if (this[INTERNALS].error) ***REMOVED***
		return Body.Promise.reject(this[INTERNALS].error);
	***REMOVED***

	let body = this.body;

	// body is null
	if (body === null) ***REMOVED***
		return Body.Promise.resolve(Buffer.alloc(0));
	***REMOVED***

	// body is blob
	if (isBlob(body)) ***REMOVED***
		body = body.stream();
	***REMOVED***

	// body is buffer
	if (Buffer.isBuffer(body)) ***REMOVED***
		return Body.Promise.resolve(body);
	***REMOVED***

	// istanbul ignore if: should never happen
	if (!(body instanceof Stream)) ***REMOVED***
		return Body.Promise.resolve(Buffer.alloc(0));
	***REMOVED***

	// body is stream
	// get ready to actually consume the body
	let accum = [];
	let accumBytes = 0;
	let abort = false;

	return new Body.Promise(function (resolve, reject) ***REMOVED***
		let resTimeout;

		// allow timeout on slow response body
		if (_this4.timeout) ***REMOVED***
			resTimeout = setTimeout(function () ***REMOVED***
				abort = true;
				reject(new FetchError(`Response timeout while trying to fetch $***REMOVED***_this4.url***REMOVED*** (over $***REMOVED***_this4.timeout***REMOVED***ms)`, 'body-timeout'));
			***REMOVED***, _this4.timeout);
		***REMOVED***

		// handle stream errors
		body.on('error', function (err) ***REMOVED***
			if (err.name === 'AbortError') ***REMOVED***
				// if the request was aborted, reject with this Error
				abort = true;
				reject(err);
			***REMOVED*** else ***REMOVED***
				// other errors, such as incorrect content-encoding
				reject(new FetchError(`Invalid response body while trying to fetch $***REMOVED***_this4.url***REMOVED***: $***REMOVED***err.message***REMOVED***`, 'system', err));
			***REMOVED***
		***REMOVED***);

		body.on('data', function (chunk) ***REMOVED***
			if (abort || chunk === null) ***REMOVED***
				return;
			***REMOVED***

			if (_this4.size && accumBytes + chunk.length > _this4.size) ***REMOVED***
				abort = true;
				reject(new FetchError(`content size at $***REMOVED***_this4.url***REMOVED*** over limit: $***REMOVED***_this4.size***REMOVED***`, 'max-size'));
				return;
			***REMOVED***

			accumBytes += chunk.length;
			accum.push(chunk);
		***REMOVED***);

		body.on('end', function () ***REMOVED***
			if (abort) ***REMOVED***
				return;
			***REMOVED***

			clearTimeout(resTimeout);

			try ***REMOVED***
				resolve(Buffer.concat(accum, accumBytes));
			***REMOVED*** catch (err) ***REMOVED***
				// handle streams that have accumulated too much data (issue #414)
				reject(new FetchError(`Could not create Buffer from response body for $***REMOVED***_this4.url***REMOVED***: $***REMOVED***err.message***REMOVED***`, 'system', err));
			***REMOVED***
		***REMOVED***);
	***REMOVED***);
***REMOVED***

/**
 * Detect buffer encoding and convert to target encoding
 * ref: http://www.w3.org/TR/2011/WD-html5-20110113/parsing.html#determining-the-character-encoding
 *
 * @param   Buffer  buffer    Incoming buffer
 * @param   String  encoding  Target encoding
 * @return  String
 */
function convertBody(buffer, headers) ***REMOVED***
	if (typeof convert !== 'function') ***REMOVED***
		throw new Error('The package `encoding` must be installed to use the textConverted() function');
	***REMOVED***

	const ct = headers.get('content-type');
	let charset = 'utf-8';
	let res, str;

	// header
	if (ct) ***REMOVED***
		res = /charset=([^;]*)/i.exec(ct);
	***REMOVED***

	// no charset in content type, peek at response body for at most 1024 bytes
	str = buffer.slice(0, 1024).toString();

	// html5
	if (!res && str) ***REMOVED***
		res = /<meta.+?charset=(['"])(.+?)\1/i.exec(str);
	***REMOVED***

	// html4
	if (!res && str) ***REMOVED***
		res = /<meta[\s]+?http-equiv=(['"])content-type\1[\s]+?content=(['"])(.+?)\2/i.exec(str);
		if (!res) ***REMOVED***
			res = /<meta[\s]+?content=(['"])(.+?)\1[\s]+?http-equiv=(['"])content-type\3/i.exec(str);
			if (res) ***REMOVED***
				res.pop(); // drop last quote
			***REMOVED***
		***REMOVED***

		if (res) ***REMOVED***
			res = /charset=(.*)/i.exec(res.pop());
		***REMOVED***
	***REMOVED***

	// xml
	if (!res && str) ***REMOVED***
		res = /<\?xml.+?encoding=(['"])(.+?)\1/i.exec(str);
	***REMOVED***

	// found charset
	if (res) ***REMOVED***
		charset = res.pop();

		// prevent decode issues when sites use incorrect encoding
		// ref: https://hsivonen.fi/encoding-menu/
		if (charset === 'gb2312' || charset === 'gbk') ***REMOVED***
			charset = 'gb18030';
		***REMOVED***
	***REMOVED***

	// turn raw buffers into a single utf-8 buffer
	return convert(buffer, 'UTF-8', charset).toString();
***REMOVED***

/**
 * Detect a URLSearchParams object
 * ref: https://github.com/bitinn/node-fetch/issues/296#issuecomment-307598143
 *
 * @param   Object  obj     Object to detect by type or brand
 * @return  String
 */
function isURLSearchParams(obj) ***REMOVED***
	// Duck-typing as a necessary condition.
	if (typeof obj !== 'object' || typeof obj.append !== 'function' || typeof obj.delete !== 'function' || typeof obj.get !== 'function' || typeof obj.getAll !== 'function' || typeof obj.has !== 'function' || typeof obj.set !== 'function') ***REMOVED***
		return false;
	***REMOVED***

	// Brand-checking and more duck-typing as optional condition.
	return obj.constructor.name === 'URLSearchParams' || Object.prototype.toString.call(obj) === '[object URLSearchParams]' || typeof obj.sort === 'function';
***REMOVED***

/**
 * Check if `obj` is a W3C `Blob` object (which `File` inherits from)
 * @param  ***REMOVED*******REMOVED*** obj
 * @return ***REMOVED***boolean***REMOVED***
 */
function isBlob(obj) ***REMOVED***
	return typeof obj === 'object' && typeof obj.arrayBuffer === 'function' && typeof obj.type === 'string' && typeof obj.stream === 'function' && typeof obj.constructor === 'function' && typeof obj.constructor.name === 'string' && /^(Blob|File)$/.test(obj.constructor.name) && /^(Blob|File)$/.test(obj[Symbol.toStringTag]);
***REMOVED***

/**
 * Clone body given Res/Req instance
 *
 * @param   Mixed  instance  Response or Request instance
 * @return  Mixed
 */
function clone(instance) ***REMOVED***
	let p1, p2;
	let body = instance.body;

	// don't allow cloning a used body
	if (instance.bodyUsed) ***REMOVED***
		throw new Error('cannot clone body after it is used');
	***REMOVED***

	// check that body is a stream and not form-data object
	// note: we can't clone the form-data object without having it as a dependency
	if (body instanceof Stream && typeof body.getBoundary !== 'function') ***REMOVED***
		// tee instance body
		p1 = new PassThrough();
		p2 = new PassThrough();
		body.pipe(p1);
		body.pipe(p2);
		// set instance body to teed body and return the other teed body
		instance[INTERNALS].body = p1;
		body = p2;
	***REMOVED***

	return body;
***REMOVED***

/**
 * Performs the operation "extract a `Content-Type` value from |object|" as
 * specified in the specification:
 * https://fetch.spec.whatwg.org/#concept-bodyinit-extract
 *
 * This function assumes that instance.body is present.
 *
 * @param   Mixed  instance  Any options.body input
 */
function extractContentType(body) ***REMOVED***
	if (body === null) ***REMOVED***
		// body is null
		return null;
	***REMOVED*** else if (typeof body === 'string') ***REMOVED***
		// body is string
		return 'text/plain;charset=UTF-8';
	***REMOVED*** else if (isURLSearchParams(body)) ***REMOVED***
		// body is a URLSearchParams
		return 'application/x-www-form-urlencoded;charset=UTF-8';
	***REMOVED*** else if (isBlob(body)) ***REMOVED***
		// body is blob
		return body.type || null;
	***REMOVED*** else if (Buffer.isBuffer(body)) ***REMOVED***
		// body is buffer
		return null;
	***REMOVED*** else if (Object.prototype.toString.call(body) === '[object ArrayBuffer]') ***REMOVED***
		// body is ArrayBuffer
		return null;
	***REMOVED*** else if (ArrayBuffer.isView(body)) ***REMOVED***
		// body is ArrayBufferView
		return null;
	***REMOVED*** else if (typeof body.getBoundary === 'function') ***REMOVED***
		// detect form data input from form-data module
		return `multipart/form-data;boundary=$***REMOVED***body.getBoundary()***REMOVED***`;
	***REMOVED*** else if (body instanceof Stream) ***REMOVED***
		// body is stream
		// can't really do much about this
		return null;
	***REMOVED*** else ***REMOVED***
		// Body constructor defaults other things to string
		return 'text/plain;charset=UTF-8';
	***REMOVED***
***REMOVED***

/**
 * The Fetch Standard treats this as if "total bytes" is a property on the body.
 * For us, we have to explicitly get it with a function.
 *
 * ref: https://fetch.spec.whatwg.org/#concept-body-total-bytes
 *
 * @param   Body    instance   Instance of Body
 * @return  Number?            Number of bytes, or null if not possible
 */
function getTotalBytes(instance) ***REMOVED***
	const body = instance.body;


	if (body === null) ***REMOVED***
		// body is null
		return 0;
	***REMOVED*** else if (isBlob(body)) ***REMOVED***
		return body.size;
	***REMOVED*** else if (Buffer.isBuffer(body)) ***REMOVED***
		// body is buffer
		return body.length;
	***REMOVED*** else if (body && typeof body.getLengthSync === 'function') ***REMOVED***
		// detect form data input from form-data module
		if (body._lengthRetrievers && body._lengthRetrievers.length == 0 || // 1.x
		body.hasKnownLength && body.hasKnownLength()) ***REMOVED***
			// 2.x
			return body.getLengthSync();
		***REMOVED***
		return null;
	***REMOVED*** else ***REMOVED***
		// body is stream
		return null;
	***REMOVED***
***REMOVED***

/**
 * Write a Body to a Node.js WritableStream (e.g. http.Request) object.
 *
 * @param   Body    instance   Instance of Body
 * @return  Void
 */
function writeToStream(dest, instance) ***REMOVED***
	const body = instance.body;


	if (body === null) ***REMOVED***
		// body is null
		dest.end();
	***REMOVED*** else if (isBlob(body)) ***REMOVED***
		body.stream().pipe(dest);
	***REMOVED*** else if (Buffer.isBuffer(body)) ***REMOVED***
		// body is buffer
		dest.write(body);
		dest.end();
	***REMOVED*** else ***REMOVED***
		// body is stream
		body.pipe(dest);
	***REMOVED***
***REMOVED***

// expose Promise
Body.Promise = global.Promise;

/**
 * headers.js
 *
 * Headers class offers convenient helpers
 */

const invalidTokenRegex = /[^\^_`a-zA-Z\-0-9!#$%&'*+.|~]/;
const invalidHeaderCharRegex = /[^\t\x20-\x7e\x80-\xff]/;

function validateName(name) ***REMOVED***
	name = `$***REMOVED***name***REMOVED***`;
	if (invalidTokenRegex.test(name) || name === '') ***REMOVED***
		throw new TypeError(`$***REMOVED***name***REMOVED*** is not a legal HTTP header name`);
	***REMOVED***
***REMOVED***

function validateValue(value) ***REMOVED***
	value = `$***REMOVED***value***REMOVED***`;
	if (invalidHeaderCharRegex.test(value)) ***REMOVED***
		throw new TypeError(`$***REMOVED***value***REMOVED*** is not a legal HTTP header value`);
	***REMOVED***
***REMOVED***

/**
 * Find the key in the map object given a header name.
 *
 * Returns undefined if not found.
 *
 * @param   String  name  Header name
 * @return  String|Undefined
 */
function find(map, name) ***REMOVED***
	name = name.toLowerCase();
	for (const key in map) ***REMOVED***
		if (key.toLowerCase() === name) ***REMOVED***
			return key;
		***REMOVED***
	***REMOVED***
	return undefined;
***REMOVED***

const MAP = Symbol('map');
class Headers ***REMOVED***
	/**
  * Headers class
  *
  * @param   Object  headers  Response headers
  * @return  Void
  */
	constructor() ***REMOVED***
		let init = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : undefined;

		this[MAP] = Object.create(null);

		if (init instanceof Headers) ***REMOVED***
			const rawHeaders = init.raw();
			const headerNames = Object.keys(rawHeaders);

			for (const headerName of headerNames) ***REMOVED***
				for (const value of rawHeaders[headerName]) ***REMOVED***
					this.append(headerName, value);
				***REMOVED***
			***REMOVED***

			return;
		***REMOVED***

		// We don't worry about converting prop to ByteString here as append()
		// will handle it.
		if (init == null) ; else if (typeof init === 'object') ***REMOVED***
			const method = init[Symbol.iterator];
			if (method != null) ***REMOVED***
				if (typeof method !== 'function') ***REMOVED***
					throw new TypeError('Header pairs must be iterable');
				***REMOVED***

				// sequence<sequence<ByteString>>
				// Note: per spec we have to first exhaust the lists then process them
				const pairs = [];
				for (const pair of init) ***REMOVED***
					if (typeof pair !== 'object' || typeof pair[Symbol.iterator] !== 'function') ***REMOVED***
						throw new TypeError('Each header pair must be iterable');
					***REMOVED***
					pairs.push(Array.from(pair));
				***REMOVED***

				for (const pair of pairs) ***REMOVED***
					if (pair.length !== 2) ***REMOVED***
						throw new TypeError('Each header pair must be a name/value tuple');
					***REMOVED***
					this.append(pair[0], pair[1]);
				***REMOVED***
			***REMOVED*** else ***REMOVED***
				// record<ByteString, ByteString>
				for (const key of Object.keys(init)) ***REMOVED***
					const value = init[key];
					this.append(key, value);
				***REMOVED***
			***REMOVED***
		***REMOVED*** else ***REMOVED***
			throw new TypeError('Provided initializer must be an object');
		***REMOVED***
	***REMOVED***

	/**
  * Return combined header value given name
  *
  * @param   String  name  Header name
  * @return  Mixed
  */
	get(name) ***REMOVED***
		name = `$***REMOVED***name***REMOVED***`;
		validateName(name);
		const key = find(this[MAP], name);
		if (key === undefined) ***REMOVED***
			return null;
		***REMOVED***

		return this[MAP][key].join(', ');
	***REMOVED***

	/**
  * Iterate over all headers
  *
  * @param   Function  callback  Executed for each item with parameters (value, name, thisArg)
  * @param   Boolean   thisArg   `this` context for callback function
  * @return  Void
  */
	forEach(callback) ***REMOVED***
		let thisArg = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : undefined;

		let pairs = getHeaders(this);
		let i = 0;
		while (i < pairs.length) ***REMOVED***
			var _pairs$i = pairs[i];
			const name = _pairs$i[0],
			      value = _pairs$i[1];

			callback.call(thisArg, value, name, this);
			pairs = getHeaders(this);
			i++;
		***REMOVED***
	***REMOVED***

	/**
  * Overwrite header values given name
  *
  * @param   String  name   Header name
  * @param   String  value  Header value
  * @return  Void
  */
	set(name, value) ***REMOVED***
		name = `$***REMOVED***name***REMOVED***`;
		value = `$***REMOVED***value***REMOVED***`;
		validateName(name);
		validateValue(value);
		const key = find(this[MAP], name);
		this[MAP][key !== undefined ? key : name] = [value];
	***REMOVED***

	/**
  * Append a value onto existing header
  *
  * @param   String  name   Header name
  * @param   String  value  Header value
  * @return  Void
  */
	append(name, value) ***REMOVED***
		name = `$***REMOVED***name***REMOVED***`;
		value = `$***REMOVED***value***REMOVED***`;
		validateName(name);
		validateValue(value);
		const key = find(this[MAP], name);
		if (key !== undefined) ***REMOVED***
			this[MAP][key].push(value);
		***REMOVED*** else ***REMOVED***
			this[MAP][name] = [value];
		***REMOVED***
	***REMOVED***

	/**
  * Check for header name existence
  *
  * @param   String   name  Header name
  * @return  Boolean
  */
	has(name) ***REMOVED***
		name = `$***REMOVED***name***REMOVED***`;
		validateName(name);
		return find(this[MAP], name) !== undefined;
	***REMOVED***

	/**
  * Delete all header values given name
  *
  * @param   String  name  Header name
  * @return  Void
  */
	delete(name) ***REMOVED***
		name = `$***REMOVED***name***REMOVED***`;
		validateName(name);
		const key = find(this[MAP], name);
		if (key !== undefined) ***REMOVED***
			delete this[MAP][key];
		***REMOVED***
	***REMOVED***

	/**
  * Return raw headers (non-spec api)
  *
  * @return  Object
  */
	raw() ***REMOVED***
		return this[MAP];
	***REMOVED***

	/**
  * Get an iterator on keys.
  *
  * @return  Iterator
  */
	keys() ***REMOVED***
		return createHeadersIterator(this, 'key');
	***REMOVED***

	/**
  * Get an iterator on values.
  *
  * @return  Iterator
  */
	values() ***REMOVED***
		return createHeadersIterator(this, 'value');
	***REMOVED***

	/**
  * Get an iterator on entries.
  *
  * This is the default iterator of the Headers object.
  *
  * @return  Iterator
  */
	[Symbol.iterator]() ***REMOVED***
		return createHeadersIterator(this, 'key+value');
	***REMOVED***
***REMOVED***
Headers.prototype.entries = Headers.prototype[Symbol.iterator];

Object.defineProperty(Headers.prototype, Symbol.toStringTag, ***REMOVED***
	value: 'Headers',
	writable: false,
	enumerable: false,
	configurable: true
***REMOVED***);

Object.defineProperties(Headers.prototype, ***REMOVED***
	get: ***REMOVED*** enumerable: true ***REMOVED***,
	forEach: ***REMOVED*** enumerable: true ***REMOVED***,
	set: ***REMOVED*** enumerable: true ***REMOVED***,
	append: ***REMOVED*** enumerable: true ***REMOVED***,
	has: ***REMOVED*** enumerable: true ***REMOVED***,
	delete: ***REMOVED*** enumerable: true ***REMOVED***,
	keys: ***REMOVED*** enumerable: true ***REMOVED***,
	values: ***REMOVED*** enumerable: true ***REMOVED***,
	entries: ***REMOVED*** enumerable: true ***REMOVED***
***REMOVED***);

function getHeaders(headers) ***REMOVED***
	let kind = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'key+value';

	const keys = Object.keys(headers[MAP]).sort();
	return keys.map(kind === 'key' ? function (k) ***REMOVED***
		return k.toLowerCase();
	***REMOVED*** : kind === 'value' ? function (k) ***REMOVED***
		return headers[MAP][k].join(', ');
	***REMOVED*** : function (k) ***REMOVED***
		return [k.toLowerCase(), headers[MAP][k].join(', ')];
	***REMOVED***);
***REMOVED***

const INTERNAL = Symbol('internal');

function createHeadersIterator(target, kind) ***REMOVED***
	const iterator = Object.create(HeadersIteratorPrototype);
	iterator[INTERNAL] = ***REMOVED***
		target,
		kind,
		index: 0
	***REMOVED***;
	return iterator;
***REMOVED***

const HeadersIteratorPrototype = Object.setPrototypeOf(***REMOVED***
	next() ***REMOVED***
		// istanbul ignore if
		if (!this || Object.getPrototypeOf(this) !== HeadersIteratorPrototype) ***REMOVED***
			throw new TypeError('Value of `this` is not a HeadersIterator');
		***REMOVED***

		var _INTERNAL = this[INTERNAL];
		const target = _INTERNAL.target,
		      kind = _INTERNAL.kind,
		      index = _INTERNAL.index;

		const values = getHeaders(target, kind);
		const len = values.length;
		if (index >= len) ***REMOVED***
			return ***REMOVED***
				value: undefined,
				done: true
			***REMOVED***;
		***REMOVED***

		this[INTERNAL].index = index + 1;

		return ***REMOVED***
			value: values[index],
			done: false
		***REMOVED***;
	***REMOVED***
***REMOVED***, Object.getPrototypeOf(Object.getPrototypeOf([][Symbol.iterator]())));

Object.defineProperty(HeadersIteratorPrototype, Symbol.toStringTag, ***REMOVED***
	value: 'HeadersIterator',
	writable: false,
	enumerable: false,
	configurable: true
***REMOVED***);

/**
 * Export the Headers object in a form that Node.js can consume.
 *
 * @param   Headers  headers
 * @return  Object
 */
function exportNodeCompatibleHeaders(headers) ***REMOVED***
	const obj = Object.assign(***REMOVED*** __proto__: null ***REMOVED***, headers[MAP]);

	// http.request() only supports string as Host header. This hack makes
	// specifying custom Host header possible.
	const hostHeaderKey = find(headers[MAP], 'Host');
	if (hostHeaderKey !== undefined) ***REMOVED***
		obj[hostHeaderKey] = obj[hostHeaderKey][0];
	***REMOVED***

	return obj;
***REMOVED***

/**
 * Create a Headers object from an object of headers, ignoring those that do
 * not conform to HTTP grammar productions.
 *
 * @param   Object  obj  Object of headers
 * @return  Headers
 */
function createHeadersLenient(obj) ***REMOVED***
	const headers = new Headers();
	for (const name of Object.keys(obj)) ***REMOVED***
		if (invalidTokenRegex.test(name)) ***REMOVED***
			continue;
		***REMOVED***
		if (Array.isArray(obj[name])) ***REMOVED***
			for (const val of obj[name]) ***REMOVED***
				if (invalidHeaderCharRegex.test(val)) ***REMOVED***
					continue;
				***REMOVED***
				if (headers[MAP][name] === undefined) ***REMOVED***
					headers[MAP][name] = [val];
				***REMOVED*** else ***REMOVED***
					headers[MAP][name].push(val);
				***REMOVED***
			***REMOVED***
		***REMOVED*** else if (!invalidHeaderCharRegex.test(obj[name])) ***REMOVED***
			headers[MAP][name] = [obj[name]];
		***REMOVED***
	***REMOVED***
	return headers;
***REMOVED***

const INTERNALS$1 = Symbol('Response internals');

// fix an issue where "STATUS_CODES" aren't a named export for node <10
const STATUS_CODES = http.STATUS_CODES;

/**
 * Response class
 *
 * @param   Stream  body  Readable stream
 * @param   Object  opts  Response options
 * @return  Void
 */
class Response ***REMOVED***
	constructor() ***REMOVED***
		let body = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
		let opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : ***REMOVED******REMOVED***;

		Body.call(this, body, opts);

		const status = opts.status || 200;
		const headers = new Headers(opts.headers);

		if (body != null && !headers.has('Content-Type')) ***REMOVED***
			const contentType = extractContentType(body);
			if (contentType) ***REMOVED***
				headers.append('Content-Type', contentType);
			***REMOVED***
		***REMOVED***

		this[INTERNALS$1] = ***REMOVED***
			url: opts.url,
			status,
			statusText: opts.statusText || STATUS_CODES[status],
			headers,
			counter: opts.counter
		***REMOVED***;
	***REMOVED***

	get url() ***REMOVED***
		return this[INTERNALS$1].url || '';
	***REMOVED***

	get status() ***REMOVED***
		return this[INTERNALS$1].status;
	***REMOVED***

	/**
  * Convenience property representing if the request ended normally
  */
	get ok() ***REMOVED***
		return this[INTERNALS$1].status >= 200 && this[INTERNALS$1].status < 300;
	***REMOVED***

	get redirected() ***REMOVED***
		return this[INTERNALS$1].counter > 0;
	***REMOVED***

	get statusText() ***REMOVED***
		return this[INTERNALS$1].statusText;
	***REMOVED***

	get headers() ***REMOVED***
		return this[INTERNALS$1].headers;
	***REMOVED***

	/**
  * Clone this response
  *
  * @return  Response
  */
	clone() ***REMOVED***
		return new Response(clone(this), ***REMOVED***
			url: this.url,
			status: this.status,
			statusText: this.statusText,
			headers: this.headers,
			ok: this.ok,
			redirected: this.redirected
		***REMOVED***);
	***REMOVED***
***REMOVED***

Body.mixIn(Response.prototype);

Object.defineProperties(Response.prototype, ***REMOVED***
	url: ***REMOVED*** enumerable: true ***REMOVED***,
	status: ***REMOVED*** enumerable: true ***REMOVED***,
	ok: ***REMOVED*** enumerable: true ***REMOVED***,
	redirected: ***REMOVED*** enumerable: true ***REMOVED***,
	statusText: ***REMOVED*** enumerable: true ***REMOVED***,
	headers: ***REMOVED*** enumerable: true ***REMOVED***,
	clone: ***REMOVED*** enumerable: true ***REMOVED***
***REMOVED***);

Object.defineProperty(Response.prototype, Symbol.toStringTag, ***REMOVED***
	value: 'Response',
	writable: false,
	enumerable: false,
	configurable: true
***REMOVED***);

const INTERNALS$2 = Symbol('Request internals');
const URL = Url.URL || whatwgUrl.URL;

// fix an issue where "format", "parse" aren't a named export for node <10
const parse_url = Url.parse;
const format_url = Url.format;

/**
 * Wrapper around `new URL` to handle arbitrary URLs
 *
 * @param  ***REMOVED***string***REMOVED*** urlStr
 * @return ***REMOVED***void***REMOVED***
 */
function parseURL(urlStr) ***REMOVED***
	/*
 	Check whether the URL is absolute or not
 		Scheme: https://tools.ietf.org/html/rfc3986#section-3.1
 	Absolute URL: https://tools.ietf.org/html/rfc3986#section-4.3
 */
	if (/^[a-zA-Z][a-zA-Z\d+\-.]*:/.exec(urlStr)) ***REMOVED***
		urlStr = new URL(urlStr).toString();
	***REMOVED***

	// Fallback to old implementation for arbitrary URLs
	return parse_url(urlStr);
***REMOVED***

const streamDestructionSupported = 'destroy' in Stream.Readable.prototype;

/**
 * Check if a value is an instance of Request.
 *
 * @param   Mixed   input
 * @return  Boolean
 */
function isRequest(input) ***REMOVED***
	return typeof input === 'object' && typeof input[INTERNALS$2] === 'object';
***REMOVED***

function isAbortSignal(signal) ***REMOVED***
	const proto = signal && typeof signal === 'object' && Object.getPrototypeOf(signal);
	return !!(proto && proto.constructor.name === 'AbortSignal');
***REMOVED***

/**
 * Request class
 *
 * @param   Mixed   input  Url or Request instance
 * @param   Object  init   Custom options
 * @return  Void
 */
class Request ***REMOVED***
	constructor(input) ***REMOVED***
		let init = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : ***REMOVED******REMOVED***;

		let parsedURL;

		// normalize input
		if (!isRequest(input)) ***REMOVED***
			if (input && input.href) ***REMOVED***
				// in order to support Node.js' Url objects; though WHATWG's URL objects
				// will fall into this branch also (since their `toString()` will return
				// `href` property anyway)
				parsedURL = parseURL(input.href);
			***REMOVED*** else ***REMOVED***
				// coerce input to a string before attempting to parse
				parsedURL = parseURL(`$***REMOVED***input***REMOVED***`);
			***REMOVED***
			input = ***REMOVED******REMOVED***;
		***REMOVED*** else ***REMOVED***
			parsedURL = parseURL(input.url);
		***REMOVED***

		let method = init.method || input.method || 'GET';
		method = method.toUpperCase();

		if ((init.body != null || isRequest(input) && input.body !== null) && (method === 'GET' || method === 'HEAD')) ***REMOVED***
			throw new TypeError('Request with GET/HEAD method cannot have body');
		***REMOVED***

		let inputBody = init.body != null ? init.body : isRequest(input) && input.body !== null ? clone(input) : null;

		Body.call(this, inputBody, ***REMOVED***
			timeout: init.timeout || input.timeout || 0,
			size: init.size || input.size || 0
		***REMOVED***);

		const headers = new Headers(init.headers || input.headers || ***REMOVED******REMOVED***);

		if (inputBody != null && !headers.has('Content-Type')) ***REMOVED***
			const contentType = extractContentType(inputBody);
			if (contentType) ***REMOVED***
				headers.append('Content-Type', contentType);
			***REMOVED***
		***REMOVED***

		let signal = isRequest(input) ? input.signal : null;
		if ('signal' in init) signal = init.signal;

		if (signal != null && !isAbortSignal(signal)) ***REMOVED***
			throw new TypeError('Expected signal to be an instanceof AbortSignal');
		***REMOVED***

		this[INTERNALS$2] = ***REMOVED***
			method,
			redirect: init.redirect || input.redirect || 'follow',
			headers,
			parsedURL,
			signal
		***REMOVED***;

		// node-fetch-only options
		this.follow = init.follow !== undefined ? init.follow : input.follow !== undefined ? input.follow : 20;
		this.compress = init.compress !== undefined ? init.compress : input.compress !== undefined ? input.compress : true;
		this.counter = init.counter || input.counter || 0;
		this.agent = init.agent || input.agent;
	***REMOVED***

	get method() ***REMOVED***
		return this[INTERNALS$2].method;
	***REMOVED***

	get url() ***REMOVED***
		return format_url(this[INTERNALS$2].parsedURL);
	***REMOVED***

	get headers() ***REMOVED***
		return this[INTERNALS$2].headers;
	***REMOVED***

	get redirect() ***REMOVED***
		return this[INTERNALS$2].redirect;
	***REMOVED***

	get signal() ***REMOVED***
		return this[INTERNALS$2].signal;
	***REMOVED***

	/**
  * Clone this request
  *
  * @return  Request
  */
	clone() ***REMOVED***
		return new Request(this);
	***REMOVED***
***REMOVED***

Body.mixIn(Request.prototype);

Object.defineProperty(Request.prototype, Symbol.toStringTag, ***REMOVED***
	value: 'Request',
	writable: false,
	enumerable: false,
	configurable: true
***REMOVED***);

Object.defineProperties(Request.prototype, ***REMOVED***
	method: ***REMOVED*** enumerable: true ***REMOVED***,
	url: ***REMOVED*** enumerable: true ***REMOVED***,
	headers: ***REMOVED*** enumerable: true ***REMOVED***,
	redirect: ***REMOVED*** enumerable: true ***REMOVED***,
	clone: ***REMOVED*** enumerable: true ***REMOVED***,
	signal: ***REMOVED*** enumerable: true ***REMOVED***
***REMOVED***);

/**
 * Convert a Request to Node.js http request options.
 *
 * @param   Request  A Request instance
 * @return  Object   The options object to be passed to http.request
 */
function getNodeRequestOptions(request) ***REMOVED***
	const parsedURL = request[INTERNALS$2].parsedURL;
	const headers = new Headers(request[INTERNALS$2].headers);

	// fetch step 1.3
	if (!headers.has('Accept')) ***REMOVED***
		headers.set('Accept', '*/*');
	***REMOVED***

	// Basic fetch
	if (!parsedURL.protocol || !parsedURL.hostname) ***REMOVED***
		throw new TypeError('Only absolute URLs are supported');
	***REMOVED***

	if (!/^https?:$/.test(parsedURL.protocol)) ***REMOVED***
		throw new TypeError('Only HTTP(S) protocols are supported');
	***REMOVED***

	if (request.signal && request.body instanceof Stream.Readable && !streamDestructionSupported) ***REMOVED***
		throw new Error('Cancellation of streamed requests with AbortSignal is not supported in node < 8');
	***REMOVED***

	// HTTP-network-or-cache fetch steps 2.4-2.7
	let contentLengthValue = null;
	if (request.body == null && /^(POST|PUT)$/i.test(request.method)) ***REMOVED***
		contentLengthValue = '0';
	***REMOVED***
	if (request.body != null) ***REMOVED***
		const totalBytes = getTotalBytes(request);
		if (typeof totalBytes === 'number') ***REMOVED***
			contentLengthValue = String(totalBytes);
		***REMOVED***
	***REMOVED***
	if (contentLengthValue) ***REMOVED***
		headers.set('Content-Length', contentLengthValue);
	***REMOVED***

	// HTTP-network-or-cache fetch step 2.11
	if (!headers.has('User-Agent')) ***REMOVED***
		headers.set('User-Agent', 'node-fetch/1.0 (+https://github.com/bitinn/node-fetch)');
	***REMOVED***

	// HTTP-network-or-cache fetch step 2.15
	if (request.compress && !headers.has('Accept-Encoding')) ***REMOVED***
		headers.set('Accept-Encoding', 'gzip,deflate');
	***REMOVED***

	let agent = request.agent;
	if (typeof agent === 'function') ***REMOVED***
		agent = agent(parsedURL);
	***REMOVED***

	if (!headers.has('Connection') && !agent) ***REMOVED***
		headers.set('Connection', 'close');
	***REMOVED***

	// HTTP-network fetch step 4.2
	// chunked encoding is handled by Node.js

	return Object.assign(***REMOVED******REMOVED***, parsedURL, ***REMOVED***
		method: request.method,
		headers: exportNodeCompatibleHeaders(headers),
		agent
	***REMOVED***);
***REMOVED***

/**
 * abort-error.js
 *
 * AbortError interface for cancelled requests
 */

/**
 * Create AbortError instance
 *
 * @param   String      message      Error message for human
 * @return  AbortError
 */
function AbortError(message) ***REMOVED***
  Error.call(this, message);

  this.type = 'aborted';
  this.message = message;

  // hide custom error implementation details from end-users
  Error.captureStackTrace(this, this.constructor);
***REMOVED***

AbortError.prototype = Object.create(Error.prototype);
AbortError.prototype.constructor = AbortError;
AbortError.prototype.name = 'AbortError';

// fix an issue where "PassThrough", "resolve" aren't a named export for node <10
const PassThrough$1 = Stream.PassThrough;
const resolve_url = Url.resolve;

/**
 * Fetch function
 *
 * @param   Mixed    url   Absolute url or Request instance
 * @param   Object   opts  Fetch options
 * @return  Promise
 */
function fetch(url, opts) ***REMOVED***

	// allow custom promise
	if (!fetch.Promise) ***REMOVED***
		throw new Error('native promise missing, set fetch.Promise to your favorite alternative');
	***REMOVED***

	Body.Promise = fetch.Promise;

	// wrap http.request into fetch
	return new fetch.Promise(function (resolve, reject) ***REMOVED***
		// build request object
		const request = new Request(url, opts);
		const options = getNodeRequestOptions(request);

		const send = (options.protocol === 'https:' ? https : http).request;
		const signal = request.signal;

		let response = null;

		const abort = function abort() ***REMOVED***
			let error = new AbortError('The user aborted a request.');
			reject(error);
			if (request.body && request.body instanceof Stream.Readable) ***REMOVED***
				request.body.destroy(error);
			***REMOVED***
			if (!response || !response.body) return;
			response.body.emit('error', error);
		***REMOVED***;

		if (signal && signal.aborted) ***REMOVED***
			abort();
			return;
		***REMOVED***

		const abortAndFinalize = function abortAndFinalize() ***REMOVED***
			abort();
			finalize();
		***REMOVED***;

		// send request
		const req = send(options);
		let reqTimeout;

		if (signal) ***REMOVED***
			signal.addEventListener('abort', abortAndFinalize);
		***REMOVED***

		function finalize() ***REMOVED***
			req.abort();
			if (signal) signal.removeEventListener('abort', abortAndFinalize);
			clearTimeout(reqTimeout);
		***REMOVED***

		if (request.timeout) ***REMOVED***
			req.once('socket', function (socket) ***REMOVED***
				reqTimeout = setTimeout(function () ***REMOVED***
					reject(new FetchError(`network timeout at: $***REMOVED***request.url***REMOVED***`, 'request-timeout'));
					finalize();
				***REMOVED***, request.timeout);
			***REMOVED***);
		***REMOVED***

		req.on('error', function (err) ***REMOVED***
			reject(new FetchError(`request to $***REMOVED***request.url***REMOVED*** failed, reason: $***REMOVED***err.message***REMOVED***`, 'system', err));
			finalize();
		***REMOVED***);

		req.on('response', function (res) ***REMOVED***
			clearTimeout(reqTimeout);

			const headers = createHeadersLenient(res.headers);

			// HTTP fetch step 5
			if (fetch.isRedirect(res.statusCode)) ***REMOVED***
				// HTTP fetch step 5.2
				const location = headers.get('Location');

				// HTTP fetch step 5.3
				const locationURL = location === null ? null : resolve_url(request.url, location);

				// HTTP fetch step 5.5
				switch (request.redirect) ***REMOVED***
					case 'error':
						reject(new FetchError(`uri requested responds with a redirect, redirect mode is set to error: $***REMOVED***request.url***REMOVED***`, 'no-redirect'));
						finalize();
						return;
					case 'manual':
						// node-fetch-specific step: make manual redirect a bit easier to use by setting the Location header value to the resolved URL.
						if (locationURL !== null) ***REMOVED***
							// handle corrupted header
							try ***REMOVED***
								headers.set('Location', locationURL);
							***REMOVED*** catch (err) ***REMOVED***
								// istanbul ignore next: nodejs server prevent invalid response headers, we can't test this through normal request
								reject(err);
							***REMOVED***
						***REMOVED***
						break;
					case 'follow':
						// HTTP-redirect fetch step 2
						if (locationURL === null) ***REMOVED***
							break;
						***REMOVED***

						// HTTP-redirect fetch step 5
						if (request.counter >= request.follow) ***REMOVED***
							reject(new FetchError(`maximum redirect reached at: $***REMOVED***request.url***REMOVED***`, 'max-redirect'));
							finalize();
							return;
						***REMOVED***

						// HTTP-redirect fetch step 6 (counter increment)
						// Create a new Request object.
						const requestOpts = ***REMOVED***
							headers: new Headers(request.headers),
							follow: request.follow,
							counter: request.counter + 1,
							agent: request.agent,
							compress: request.compress,
							method: request.method,
							body: request.body,
							signal: request.signal,
							timeout: request.timeout,
							size: request.size
						***REMOVED***;

						// HTTP-redirect fetch step 9
						if (res.statusCode !== 303 && request.body && getTotalBytes(request) === null) ***REMOVED***
							reject(new FetchError('Cannot follow redirect with body being a readable stream', 'unsupported-redirect'));
							finalize();
							return;
						***REMOVED***

						// HTTP-redirect fetch step 11
						if (res.statusCode === 303 || (res.statusCode === 301 || res.statusCode === 302) && request.method === 'POST') ***REMOVED***
							requestOpts.method = 'GET';
							requestOpts.body = undefined;
							requestOpts.headers.delete('content-length');
						***REMOVED***

						// HTTP-redirect fetch step 15
						resolve(fetch(new Request(locationURL, requestOpts)));
						finalize();
						return;
				***REMOVED***
			***REMOVED***

			// prepare response
			res.once('end', function () ***REMOVED***
				if (signal) signal.removeEventListener('abort', abortAndFinalize);
			***REMOVED***);
			let body = res.pipe(new PassThrough$1());

			const response_options = ***REMOVED***
				url: request.url,
				status: res.statusCode,
				statusText: res.statusMessage,
				headers: headers,
				size: request.size,
				timeout: request.timeout,
				counter: request.counter
			***REMOVED***;

			// HTTP-network fetch step 12.1.1.3
			const codings = headers.get('Content-Encoding');

			// HTTP-network fetch step 12.1.1.4: handle content codings

			// in following scenarios we ignore compression support
			// 1. compression support is disabled
			// 2. HEAD request
			// 3. no Content-Encoding header
			// 4. no content response (204)
			// 5. content not modified response (304)
			if (!request.compress || request.method === 'HEAD' || codings === null || res.statusCode === 204 || res.statusCode === 304) ***REMOVED***
				response = new Response(body, response_options);
				resolve(response);
				return;
			***REMOVED***

			// For Node v6+
			// Be less strict when decoding compressed responses, since sometimes
			// servers send slightly invalid responses that are still accepted
			// by common browsers.
			// Always using Z_SYNC_FLUSH is what cURL does.
			const zlibOptions = ***REMOVED***
				flush: zlib.Z_SYNC_FLUSH,
				finishFlush: zlib.Z_SYNC_FLUSH
			***REMOVED***;

			// for gzip
			if (codings == 'gzip' || codings == 'x-gzip') ***REMOVED***
				body = body.pipe(zlib.createGunzip(zlibOptions));
				response = new Response(body, response_options);
				resolve(response);
				return;
			***REMOVED***

			// for deflate
			if (codings == 'deflate' || codings == 'x-deflate') ***REMOVED***
				// handle the infamous raw deflate response from old servers
				// a hack for old IIS and Apache servers
				const raw = res.pipe(new PassThrough$1());
				raw.once('data', function (chunk) ***REMOVED***
					// see http://stackoverflow.com/questions/37519828
					if ((chunk[0] & 0x0F) === 0x08) ***REMOVED***
						body = body.pipe(zlib.createInflate());
					***REMOVED*** else ***REMOVED***
						body = body.pipe(zlib.createInflateRaw());
					***REMOVED***
					response = new Response(body, response_options);
					resolve(response);
				***REMOVED***);
				return;
			***REMOVED***

			// for br
			if (codings == 'br' && typeof zlib.createBrotliDecompress === 'function') ***REMOVED***
				body = body.pipe(zlib.createBrotliDecompress());
				response = new Response(body, response_options);
				resolve(response);
				return;
			***REMOVED***

			// otherwise, use response as-is
			response = new Response(body, response_options);
			resolve(response);
		***REMOVED***);

		writeToStream(req, request);
	***REMOVED***);
***REMOVED***
/**
 * Redirect code matching
 *
 * @param   Number   code  Status code
 * @return  Boolean
 */
fetch.isRedirect = function (code) ***REMOVED***
	return code === 301 || code === 302 || code === 303 || code === 307 || code === 308;
***REMOVED***;

// expose Promise
fetch.Promise = global.Promise;

module.exports = exports = fetch;
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.default = exports;
exports.Headers = Headers;
exports.Request = Request;
exports.Response = Response;
exports.FetchError = FetchError;
