// ircv3/lib/Message/MessageTypes/Commands/CapabilityNegotiation.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.CapabilityNegotiation = void 0;
var tslib_1 = require("tslib");
var Message_1 = require("../../Message");
var MessageDefinition_1 = require("../../MessageDefinition");
var CapabilityNegotiation = /** @class */ (function (_super) ***REMOVED***
    tslib_1.__extends(CapabilityNegotiation, _super);
    function CapabilityNegotiation() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    CapabilityNegotiation_1 = CapabilityNegotiation;
    CapabilityNegotiation.prototype.isResponseTo = function (originalMessage) ***REMOVED***
        if (!(originalMessage instanceof CapabilityNegotiation_1)) ***REMOVED***
            return false;
        ***REMOVED***
        switch (this.params.subCommand) ***REMOVED***
            case 'ACK':
            case 'NAK': ***REMOVED***
                // trim is necessary because some networks seem to add trailing spaces (looking at you, Freenode)...
                return (originalMessage.params.subCommand === 'REQ' &&
                    originalMessage.params.capabilities === this.params.capabilities.trim());
            ***REMOVED***
            case 'LS':
            case 'LIST': ***REMOVED***
                return originalMessage.params.subCommand === this.params.subCommand;
            ***REMOVED***
            default: ***REMOVED***
                return false;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    CapabilityNegotiation.prototype.endsResponseTo = function (originalMessage) ***REMOVED***
        if (!(originalMessage instanceof CapabilityNegotiation_1)) ***REMOVED***
            return false;
        ***REMOVED***
        switch (this.params.subCommand) ***REMOVED***
            case 'LS':
            case 'LIST': ***REMOVED***
                return !this.params.continued;
            ***REMOVED***
            default: ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    var CapabilityNegotiation_1;
    CapabilityNegotiation.SUPPORTS_CAPTURE = true;
    tslib_1.__decorate([
        MessageDefinition_1.MessageParamDefinition(***REMOVED***
            match: /^(?:[a-z_\-\[\]\\^***REMOVED******REMOVED***|`][a-z0-9_\-\[\]\\^***REMOVED******REMOVED***|`]+|\*)$/i,
            optional: true,
            noClient: true
        ***REMOVED***)
    ], CapabilityNegotiation.prototype, "target", void 0);
    tslib_1.__decorate([
        MessageDefinition_1.MessageParamDefinition(***REMOVED***
            match: /^(?:LS|LIST|REQ|ACK|NAK|END|NEW|DEL)$/i
        ***REMOVED***)
    ], CapabilityNegotiation.prototype, "subCommand", void 0);
    tslib_1.__decorate([
        MessageDefinition_1.MessageParamDefinition(***REMOVED***
            match: /^\d+$/,
            optional: true
        ***REMOVED***)
    ], CapabilityNegotiation.prototype, "version", void 0);
    tslib_1.__decorate([
        MessageDefinition_1.MessageParamDefinition(***REMOVED***
            match: /^\*$/,
            optional: true
        ***REMOVED***)
    ], CapabilityNegotiation.prototype, "continued", void 0);
    tslib_1.__decorate([
        MessageDefinition_1.MessageParamDefinition(***REMOVED***
            trailing: true,
            optional: true
        ***REMOVED***)
    ], CapabilityNegotiation.prototype, "capabilities", void 0);
    CapabilityNegotiation = CapabilityNegotiation_1 = tslib_1.__decorate([
        MessageDefinition_1.MessageType('CAP')
    ], CapabilityNegotiation);
    return CapabilityNegotiation;
***REMOVED***(Message_1.Message));
exports.CapabilityNegotiation = CapabilityNegotiation;
