// ircv3/lib/Message/MessageTypes/Commands/index.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.CapabilityNegotiation = exports.IsOnQuery = exports.UserHostQuery = exports.Wallops = exports.Restart = exports.Rehash = exports.Away = exports.ErrorMessage = exports.Pong = exports.Ping = exports.Kill = exports.WhoWasQuery = exports.WhoIsQuery = exports.WhoQuery = exports.Notice = exports.PrivateMessage = exports.ChannelKick = exports.ChannelInvite = exports.ChannelList = exports.Names = exports.Topic = exports.Mode = exports.ChannelPart = exports.ChannelJoin = exports.ServerQuit = exports.ClientQuit = exports.OperLogin = exports.UserRegistration = exports.NickChange = exports.Password = void 0;
// region RFC1459
// region 4 Message details
// region 4.1 Connection Registration
// 4.1.1 Password message
var Password_1 = require("./Password");
Object.defineProperty(exports, "Password", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Password_1.Password; ***REMOVED*** ***REMOVED***);
// 4.1.2 Nickname message
var NickChange_1 = require("./NickChange");
Object.defineProperty(exports, "NickChange", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return NickChange_1.NickChange; ***REMOVED*** ***REMOVED***);
// 4.1.3 User message
var UserRegistration_1 = require("./UserRegistration");
Object.defineProperty(exports, "UserRegistration", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return UserRegistration_1.UserRegistration; ***REMOVED*** ***REMOVED***);
// 4.1.4 Server message
// We can't be a server yet
// 4.1.5 Operator message
var OperLogin_1 = require("./OperLogin");
Object.defineProperty(exports, "OperLogin", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return OperLogin_1.OperLogin; ***REMOVED*** ***REMOVED***);
// 4.1.6 Quit message
var ClientQuit_1 = require("./ClientQuit");
Object.defineProperty(exports, "ClientQuit", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ClientQuit_1.ClientQuit; ***REMOVED*** ***REMOVED***);
// 4.1.7 Server Quit message
var ServerQuit_1 = require("./ServerQuit");
Object.defineProperty(exports, "ServerQuit", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ServerQuit_1.ServerQuit; ***REMOVED*** ***REMOVED***);
// endregion
// region 4.2 Channel operations
// 4.2.1 Join message
var ChannelJoin_1 = require("./ChannelJoin");
Object.defineProperty(exports, "ChannelJoin", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ChannelJoin_1.ChannelJoin; ***REMOVED*** ***REMOVED***);
// 4.2.2 Part message
var ChannelPart_1 = require("./ChannelPart");
Object.defineProperty(exports, "ChannelPart", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ChannelPart_1.ChannelPart; ***REMOVED*** ***REMOVED***);
// 4.2.3 Mode message
var Mode_1 = require("./Mode");
Object.defineProperty(exports, "Mode", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Mode_1.Mode; ***REMOVED*** ***REMOVED***);
// 4.2.4 Topic message
var Topic_1 = require("./Topic");
Object.defineProperty(exports, "Topic", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Topic_1.Topic; ***REMOVED*** ***REMOVED***);
// 4.2.5 Names message
var Names_1 = require("./Names");
Object.defineProperty(exports, "Names", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Names_1.Names; ***REMOVED*** ***REMOVED***);
// 4.2.6 List message
var ChannelList_1 = require("./ChannelList");
Object.defineProperty(exports, "ChannelList", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ChannelList_1.ChannelList; ***REMOVED*** ***REMOVED***);
// 4.2.7 Invite message
var ChannelInvite_1 = require("./ChannelInvite");
Object.defineProperty(exports, "ChannelInvite", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ChannelInvite_1.ChannelInvite; ***REMOVED*** ***REMOVED***);
// 4.2.8 Kick message
var ChannelKick_1 = require("./ChannelKick");
Object.defineProperty(exports, "ChannelKick", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ChannelKick_1.ChannelKick; ***REMOVED*** ***REMOVED***);
// endregion
// region 4.3 Server queries and commands
// TODO
// endregion
// region 4.4 Sending messages
// 4.4.1 Private messages
var PrivateMessage_1 = require("./PrivateMessage");
Object.defineProperty(exports, "PrivateMessage", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return PrivateMessage_1.PrivateMessage; ***REMOVED*** ***REMOVED***);
// 4.4.2 Notice messages
var Notice_1 = require("./Notice");
Object.defineProperty(exports, "Notice", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Notice_1.Notice; ***REMOVED*** ***REMOVED***);
// endregion
// region 4.5 User-based queries
// 4.5.1 Who query
var WhoQuery_1 = require("./WhoQuery");
Object.defineProperty(exports, "WhoQuery", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return WhoQuery_1.WhoQuery; ***REMOVED*** ***REMOVED***);
// 4.5.2 Whois query
var WhoIsQuery_1 = require("./WhoIsQuery");
Object.defineProperty(exports, "WhoIsQuery", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return WhoIsQuery_1.WhoIsQuery; ***REMOVED*** ***REMOVED***);
// 4.5.3 Whowas
var WhoWasQuery_1 = require("./WhoWasQuery");
Object.defineProperty(exports, "WhoWasQuery", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return WhoWasQuery_1.WhoWasQuery; ***REMOVED*** ***REMOVED***);
// endregion
// region 4.6 Miscellaneous messages
// 4.6.1 Kill message
var Kill_1 = require("./Kill");
Object.defineProperty(exports, "Kill", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Kill_1.Kill; ***REMOVED*** ***REMOVED***);
// 4.6.2 Ping message
var Ping_1 = require("./Ping");
Object.defineProperty(exports, "Ping", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Ping_1.Ping; ***REMOVED*** ***REMOVED***);
// 4.6.3 Pong message
var Pong_1 = require("./Pong");
Object.defineProperty(exports, "Pong", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Pong_1.Pong; ***REMOVED*** ***REMOVED***);
// 4.6.4 Error message
var ErrorMessage_1 = require("./ErrorMessage");
Object.defineProperty(exports, "ErrorMessage", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ErrorMessage_1.ErrorMessage; ***REMOVED*** ***REMOVED***);
// endregion
// endregion
// region 5 Optionals
// 5.1 Away
var Away_1 = require("./Away");
Object.defineProperty(exports, "Away", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Away_1.Away; ***REMOVED*** ***REMOVED***);
// 5.2 Rehash message
var Rehash_1 = require("./Rehash");
Object.defineProperty(exports, "Rehash", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Rehash_1.Rehash; ***REMOVED*** ***REMOVED***);
// 5.3 Restart message
var Restart_1 = require("./Restart");
Object.defineProperty(exports, "Restart", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Restart_1.Restart; ***REMOVED*** ***REMOVED***);
// 5.4 Summon message
// really, would anyone use that?
// 5.5 Users message
// would anyone WANT to have that? it seems like a huge security risk.
// 5.6 Operwall message
var WallopsMessage_1 = require("./WallopsMessage");
Object.defineProperty(exports, "Wallops", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return WallopsMessage_1.WallopsMessage; ***REMOVED*** ***REMOVED***);
// 5.7 Userhost message
var UserHostQuery_1 = require("./UserHostQuery");
Object.defineProperty(exports, "UserHostQuery", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return UserHostQuery_1.UserHostQuery; ***REMOVED*** ***REMOVED***);
// 5.8 Ison message
var IsOnQuery_1 = require("./IsOnQuery");
Object.defineProperty(exports, "IsOnQuery", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return IsOnQuery_1.IsOnQuery; ***REMOVED*** ***REMOVED***);
// endregion
// endregion
// region IRCv3
// Capability Negotiation
var CapabilityNegotiation_1 = require("./CapabilityNegotiation");
Object.defineProperty(exports, "CapabilityNegotiation", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return CapabilityNegotiation_1.CapabilityNegotiation; ***REMOVED*** ***REMOVED***);
// endregion
