// twitch-chat-client/es/StandardCommands/TwitchPrivateMessage.mjs
import ***REMOVED*** __decorate, __extends, __read, __spread ***REMOVED*** from "tslib";
import ***REMOVED*** MessageTypes ***REMOVED*** from 'ircv3';
import ***REMOVED*** rtfm ***REMOVED*** from 'twitch-common';
import ***REMOVED*** ChatUser ***REMOVED*** from "../ChatUser.mjs";
import ***REMOVED*** fillTextPositions, parseEmoteOffsets, parseEmotePositions ***REMOVED*** from "../Toolkit/EmoteTools.mjs";
/**
 * An IRC PRIVMSG, with easy accessors for commonly used data from its tags.
 */
var TwitchPrivateMessage = /** @class */ (function (_super) ***REMOVED***
    __extends(TwitchPrivateMessage, _super);
    function TwitchPrivateMessage() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Object.defineProperty(TwitchPrivateMessage.prototype, "id", ***REMOVED***
        /**
         * The ID of the message.
         */
        get: function () ***REMOVED***
            return this._tags.get('id');
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "userInfo", ***REMOVED***
        /**
         * Info about the user that send the message, like their user ID and their status in the current channel.
         */
        get: function () ***REMOVED***
            return new ChatUser(this._prefix.nick, this._tags);
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "channelId", ***REMOVED***
        /**
         * The ID of the channel the message is in.
         */
        get: function () ***REMOVED***
            var _a;
            return (_a = this._tags.get('room-id')) !== null && _a !== void 0 ? _a : null;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "isCheer", ***REMOVED***
        /**
         * Whether the message is a cheer.
         */
        get: function () ***REMOVED***
            return this._tags.has('bits');
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "bits", ***REMOVED***
        /**
         * The number of bits cheered with the message.
         */
        get: function () ***REMOVED***
            var _a;
            return Number((_a = this._tags.get('bits')) !== null && _a !== void 0 ? _a : 0);
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "totalBits", ***REMOVED***
        /**
         * The number of bits cheered with the message.
         *
         * @deprecated Use ***REMOVED***@TwitchPrivateMessage#bits***REMOVED*** instead.
         */
        get: function () ***REMOVED***
            return this.bits;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(TwitchPrivateMessage.prototype, "emoteOffsets", ***REMOVED***
        /**
         * The offsets of emote usages in the message.
         */
        get: function () ***REMOVED***
            return parseEmoteOffsets(this._tags.get('emotes'));
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    /**
     * Parses the message, separating text from emote usages.
     */
    TwitchPrivateMessage.prototype.parseEmotes = function () ***REMOVED***
        var messageText = this.params.message;
        var foundEmotes = parseEmotePositions(messageText, this.emoteOffsets);
        return fillTextPositions(messageText, foundEmotes);
    ***REMOVED***;
    /**
     * Parses the message, separating text from emote usages and cheers.
     *
     * @param cheermotes A list of cheermotes
     */
    TwitchPrivateMessage.prototype.parseEmotesAndBits = function (cheermotes) ***REMOVED***
        var messageText = this.params.message;
        var foundCheermotes = cheermotes.parseMessage(messageText);
        var foundEmotesAndCheermotes = __spread(parseEmotePositions(messageText, this.emoteOffsets), foundCheermotes.map(function (cheermote) ***REMOVED*** return (***REMOVED***
            type: 'cheer',
            position: cheermote.position,
            length: cheermote.length,
            name: cheermote.name,
            amount: cheermote.amount,
            displayInfo: cheermote.displayInfo
        ***REMOVED***); ***REMOVED***));
        foundEmotesAndCheermotes.sort(function (a, b) ***REMOVED*** return a.position - b.position; ***REMOVED***);
        return fillTextPositions(messageText, foundEmotesAndCheermotes);
    ***REMOVED***;
    TwitchPrivateMessage = __decorate([
        rtfm('twitch-chat-client', 'TwitchPrivateMessage', 'id')
    ], TwitchPrivateMessage);
    return TwitchPrivateMessage;
***REMOVED***(MessageTypes.Commands.PrivateMessage));
export ***REMOVED*** TwitchPrivateMessage ***REMOVED***;
