// ws/lib/sender.js
/* eslint no-unused-vars: ["error", ***REMOVED*** "varsIgnorePattern": "^net|tls$" ***REMOVED***] */

'use strict';

const net = require('net');
const tls = require('tls');
const ***REMOVED*** randomFillSync ***REMOVED*** = require('crypto');

const PerMessageDeflate = require('./permessage-deflate');
const ***REMOVED*** EMPTY_BUFFER ***REMOVED*** = require('./constants');
const ***REMOVED*** isValidStatusCode ***REMOVED*** = require('./validation');
const ***REMOVED*** mask: applyMask, toBuffer ***REMOVED*** = require('./buffer-util');

const mask = Buffer.alloc(4);

/**
 * HyBi Sender implementation.
 */
class Sender ***REMOVED***
  /**
   * Creates a Sender instance.
   *
   * @param ***REMOVED***(net.Socket|tls.Socket)***REMOVED*** socket The connection socket
   * @param ***REMOVED***Object***REMOVED*** [extensions] An object containing the negotiated extensions
   */
  constructor(socket, extensions) ***REMOVED***
    this._extensions = extensions || ***REMOVED******REMOVED***;
    this._socket = socket;

    this._firstFragment = true;
    this._compress = false;

    this._bufferedBytes = 0;
    this._deflating = false;
    this._queue = [];
  ***REMOVED***

  /**
   * Frames a piece of data according to the HyBi WebSocket protocol.
   *
   * @param ***REMOVED***Buffer***REMOVED*** data The data to frame
   * @param ***REMOVED***Object***REMOVED*** options Options object
   * @param ***REMOVED***Number***REMOVED*** options.opcode The opcode
   * @param ***REMOVED***Boolean***REMOVED*** [options.readOnly=false] Specifies whether `data` can be
   *     modified
   * @param ***REMOVED***Boolean***REMOVED*** [options.fin=false] Specifies whether or not to set the
   *     FIN bit
   * @param ***REMOVED***Boolean***REMOVED*** [options.mask=false] Specifies whether or not to mask
   *     `data`
   * @param ***REMOVED***Boolean***REMOVED*** [options.rsv1=false] Specifies whether or not to set the
   *     RSV1 bit
   * @return ***REMOVED***Buffer[]***REMOVED*** The framed data as a list of `Buffer` instances
   * @public
   */
  static frame(data, options) ***REMOVED***
    const merge = options.mask && options.readOnly;
    let offset = options.mask ? 6 : 2;
    let payloadLength = data.length;

    if (data.length >= 65536) ***REMOVED***
      offset += 8;
      payloadLength = 127;
    ***REMOVED*** else if (data.length > 125) ***REMOVED***
      offset += 2;
      payloadLength = 126;
    ***REMOVED***

    const target = Buffer.allocUnsafe(merge ? data.length + offset : offset);

    target[0] = options.fin ? options.opcode | 0x80 : options.opcode;
    if (options.rsv1) target[0] |= 0x40;

    target[1] = payloadLength;

    if (payloadLength === 126) ***REMOVED***
      target.writeUInt16BE(data.length, 2);
    ***REMOVED*** else if (payloadLength === 127) ***REMOVED***
      target.writeUInt32BE(0, 2);
      target.writeUInt32BE(data.length, 6);
    ***REMOVED***

    if (!options.mask) return [target, data];

    randomFillSync(mask, 0, 4);

    target[1] |= 0x80;
    target[offset - 4] = mask[0];
    target[offset - 3] = mask[1];
    target[offset - 2] = mask[2];
    target[offset - 1] = mask[3];

    if (merge) ***REMOVED***
      applyMask(data, mask, target, offset, data.length);
      return [target];
    ***REMOVED***

    applyMask(data, mask, data, 0, data.length);
    return [target, data];
  ***REMOVED***

  /**
   * Sends a close message to the other peer.
   *
   * @param ***REMOVED***Number***REMOVED*** [code] The status code component of the body
   * @param ***REMOVED***String***REMOVED*** [data] The message component of the body
   * @param ***REMOVED***Boolean***REMOVED*** [mask=false] Specifies whether or not to mask the message
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @public
   */
  close(code, data, mask, cb) ***REMOVED***
    let buf;

    if (code === undefined) ***REMOVED***
      buf = EMPTY_BUFFER;
    ***REMOVED*** else if (typeof code !== 'number' || !isValidStatusCode(code)) ***REMOVED***
      throw new TypeError('First argument must be a valid error code number');
    ***REMOVED*** else if (data === undefined || data === '') ***REMOVED***
      buf = Buffer.allocUnsafe(2);
      buf.writeUInt16BE(code, 0);
    ***REMOVED*** else ***REMOVED***
      const length = Buffer.byteLength(data);

      if (length > 123) ***REMOVED***
        throw new RangeError('The message must not be greater than 123 bytes');
      ***REMOVED***

      buf = Buffer.allocUnsafe(2 + length);
      buf.writeUInt16BE(code, 0);
      buf.write(data, 2);
    ***REMOVED***

    if (this._deflating) ***REMOVED***
      this.enqueue([this.doClose, buf, mask, cb]);
    ***REMOVED*** else ***REMOVED***
      this.doClose(buf, mask, cb);
    ***REMOVED***
  ***REMOVED***

  /**
   * Frames and sends a close message.
   *
   * @param ***REMOVED***Buffer***REMOVED*** data The message to send
   * @param ***REMOVED***Boolean***REMOVED*** [mask=false] Specifies whether or not to mask `data`
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @private
   */
  doClose(data, mask, cb) ***REMOVED***
    this.sendFrame(
      Sender.frame(data, ***REMOVED***
        fin: true,
        rsv1: false,
        opcode: 0x08,
        mask,
        readOnly: false
      ***REMOVED***),
      cb
    );
  ***REMOVED***

  /**
   * Sends a ping message to the other peer.
   *
   * @param ***REMOVED*******REMOVED*** data The message to send
   * @param ***REMOVED***Boolean***REMOVED*** [mask=false] Specifies whether or not to mask `data`
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @public
   */
  ping(data, mask, cb) ***REMOVED***
    const buf = toBuffer(data);

    if (buf.length > 125) ***REMOVED***
      throw new RangeError('The data size must not be greater than 125 bytes');
    ***REMOVED***

    if (this._deflating) ***REMOVED***
      this.enqueue([this.doPing, buf, mask, toBuffer.readOnly, cb]);
    ***REMOVED*** else ***REMOVED***
      this.doPing(buf, mask, toBuffer.readOnly, cb);
    ***REMOVED***
  ***REMOVED***

  /**
   * Frames and sends a ping message.
   *
   * @param ***REMOVED***Buffer***REMOVED*** data The message to send
   * @param ***REMOVED***Boolean***REMOVED*** [mask=false] Specifies whether or not to mask `data`
   * @param ***REMOVED***Boolean***REMOVED*** [readOnly=false] Specifies whether `data` can be modified
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @private
   */
  doPing(data, mask, readOnly, cb) ***REMOVED***
    this.sendFrame(
      Sender.frame(data, ***REMOVED***
        fin: true,
        rsv1: false,
        opcode: 0x09,
        mask,
        readOnly
      ***REMOVED***),
      cb
    );
  ***REMOVED***

  /**
   * Sends a pong message to the other peer.
   *
   * @param ***REMOVED*******REMOVED*** data The message to send
   * @param ***REMOVED***Boolean***REMOVED*** [mask=false] Specifies whether or not to mask `data`
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @public
   */
  pong(data, mask, cb) ***REMOVED***
    const buf = toBuffer(data);

    if (buf.length > 125) ***REMOVED***
      throw new RangeError('The data size must not be greater than 125 bytes');
    ***REMOVED***

    if (this._deflating) ***REMOVED***
      this.enqueue([this.doPong, buf, mask, toBuffer.readOnly, cb]);
    ***REMOVED*** else ***REMOVED***
      this.doPong(buf, mask, toBuffer.readOnly, cb);
    ***REMOVED***
  ***REMOVED***

  /**
   * Frames and sends a pong message.
   *
   * @param ***REMOVED***Buffer***REMOVED*** data The message to send
   * @param ***REMOVED***Boolean***REMOVED*** [mask=false] Specifies whether or not to mask `data`
   * @param ***REMOVED***Boolean***REMOVED*** [readOnly=false] Specifies whether `data` can be modified
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @private
   */
  doPong(data, mask, readOnly, cb) ***REMOVED***
    this.sendFrame(
      Sender.frame(data, ***REMOVED***
        fin: true,
        rsv1: false,
        opcode: 0x0a,
        mask,
        readOnly
      ***REMOVED***),
      cb
    );
  ***REMOVED***

  /**
   * Sends a data message to the other peer.
   *
   * @param ***REMOVED*******REMOVED*** data The message to send
   * @param ***REMOVED***Object***REMOVED*** options Options object
   * @param ***REMOVED***Boolean***REMOVED*** [options.compress=false] Specifies whether or not to
   *     compress `data`
   * @param ***REMOVED***Boolean***REMOVED*** [options.binary=false] Specifies whether `data` is binary
   *     or text
   * @param ***REMOVED***Boolean***REMOVED*** [options.fin=false] Specifies whether the fragment is the
   *     last one
   * @param ***REMOVED***Boolean***REMOVED*** [options.mask=false] Specifies whether or not to mask
   *     `data`
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @public
   */
  send(data, options, cb) ***REMOVED***
    const buf = toBuffer(data);
    const perMessageDeflate = this._extensions[PerMessageDeflate.extensionName];
    let opcode = options.binary ? 2 : 1;
    let rsv1 = options.compress;

    if (this._firstFragment) ***REMOVED***
      this._firstFragment = false;
      if (rsv1 && perMessageDeflate) ***REMOVED***
        rsv1 = buf.length >= perMessageDeflate._threshold;
      ***REMOVED***
      this._compress = rsv1;
    ***REMOVED*** else ***REMOVED***
      rsv1 = false;
      opcode = 0;
    ***REMOVED***

    if (options.fin) this._firstFragment = true;

    if (perMessageDeflate) ***REMOVED***
      const opts = ***REMOVED***
        fin: options.fin,
        rsv1,
        opcode,
        mask: options.mask,
        readOnly: toBuffer.readOnly
      ***REMOVED***;

      if (this._deflating) ***REMOVED***
        this.enqueue([this.dispatch, buf, this._compress, opts, cb]);
      ***REMOVED*** else ***REMOVED***
        this.dispatch(buf, this._compress, opts, cb);
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      this.sendFrame(
        Sender.frame(buf, ***REMOVED***
          fin: options.fin,
          rsv1: false,
          opcode,
          mask: options.mask,
          readOnly: toBuffer.readOnly
        ***REMOVED***),
        cb
      );
    ***REMOVED***
  ***REMOVED***

  /**
   * Dispatches a data message.
   *
   * @param ***REMOVED***Buffer***REMOVED*** data The message to send
   * @param ***REMOVED***Boolean***REMOVED*** [compress=false] Specifies whether or not to compress
   *     `data`
   * @param ***REMOVED***Object***REMOVED*** options Options object
   * @param ***REMOVED***Number***REMOVED*** options.opcode The opcode
   * @param ***REMOVED***Boolean***REMOVED*** [options.readOnly=false] Specifies whether `data` can be
   *     modified
   * @param ***REMOVED***Boolean***REMOVED*** [options.fin=false] Specifies whether or not to set the
   *     FIN bit
   * @param ***REMOVED***Boolean***REMOVED*** [options.mask=false] Specifies whether or not to mask
   *     `data`
   * @param ***REMOVED***Boolean***REMOVED*** [options.rsv1=false] Specifies whether or not to set the
   *     RSV1 bit
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @private
   */
  dispatch(data, compress, options, cb) ***REMOVED***
    if (!compress) ***REMOVED***
      this.sendFrame(Sender.frame(data, options), cb);
      return;
    ***REMOVED***

    const perMessageDeflate = this._extensions[PerMessageDeflate.extensionName];

    this._bufferedBytes += data.length;
    this._deflating = true;
    perMessageDeflate.compress(data, options.fin, (_, buf) => ***REMOVED***
      if (this._socket.destroyed) ***REMOVED***
        const err = new Error(
          'The socket was closed while data was being compressed'
        );

        if (typeof cb === 'function') cb(err);

        for (let i = 0; i < this._queue.length; i++) ***REMOVED***
          const callback = this._queue[i][4];

          if (typeof callback === 'function') callback(err);
        ***REMOVED***

        return;
      ***REMOVED***

      this._bufferedBytes -= data.length;
      this._deflating = false;
      options.readOnly = false;
      this.sendFrame(Sender.frame(buf, options), cb);
      this.dequeue();
    ***REMOVED***);
  ***REMOVED***

  /**
   * Executes queued send operations.
   *
   * @private
   */
  dequeue() ***REMOVED***
    while (!this._deflating && this._queue.length) ***REMOVED***
      const params = this._queue.shift();

      this._bufferedBytes -= params[1].length;
      Reflect.apply(params[0], this, params.slice(1));
    ***REMOVED***
  ***REMOVED***

  /**
   * Enqueues a send operation.
   *
   * @param ***REMOVED***Array***REMOVED*** params Send operation parameters.
   * @private
   */
  enqueue(params) ***REMOVED***
    this._bufferedBytes += params[1].length;
    this._queue.push(params);
  ***REMOVED***

  /**
   * Sends a frame.
   *
   * @param ***REMOVED***Buffer[]***REMOVED*** list The frame to send
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback
   * @private
   */
  sendFrame(list, cb) ***REMOVED***
    if (list.length === 2) ***REMOVED***
      this._socket.cork();
      this._socket.write(list[0]);
      this._socket.write(list[1], cb);
      this._socket.uncork();
    ***REMOVED*** else ***REMOVED***
      this._socket.write(list[0], cb);
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = Sender;
