// ws/lib/websocket.js
/* eslint no-unused-vars: ["error", ***REMOVED*** "varsIgnorePattern": "^Readable$" ***REMOVED***] */

'use strict';

const EventEmitter = require('events');
const https = require('https');
const http = require('http');
const net = require('net');
const tls = require('tls');
const ***REMOVED*** randomBytes, createHash ***REMOVED*** = require('crypto');
const ***REMOVED*** Readable ***REMOVED*** = require('stream');
const ***REMOVED*** URL ***REMOVED*** = require('url');

const PerMessageDeflate = require('./permessage-deflate');
const Receiver = require('./receiver');
const Sender = require('./sender');
const ***REMOVED***
  BINARY_TYPES,
  EMPTY_BUFFER,
  GUID,
  kStatusCode,
  kWebSocket,
  NOOP
***REMOVED*** = require('./constants');
const ***REMOVED*** addEventListener, removeEventListener ***REMOVED*** = require('./event-target');
const ***REMOVED*** format, parse ***REMOVED*** = require('./extension');
const ***REMOVED*** toBuffer ***REMOVED*** = require('./buffer-util');

const readyStates = ['CONNECTING', 'OPEN', 'CLOSING', 'CLOSED'];
const protocolVersions = [8, 13];
const closeTimeout = 30 * 1000;

/**
 * Class representing a WebSocket.
 *
 * @extends EventEmitter
 */
class WebSocket extends EventEmitter ***REMOVED***
  /**
   * Create a new `WebSocket`.
   *
   * @param ***REMOVED***(String|URL)***REMOVED*** address The URL to which to connect
   * @param ***REMOVED***(String|String[])***REMOVED*** [protocols] The subprotocols
   * @param ***REMOVED***Object***REMOVED*** [options] Connection options
   */
  constructor(address, protocols, options) ***REMOVED***
    super();

    this._binaryType = BINARY_TYPES[0];
    this._closeCode = 1006;
    this._closeFrameReceived = false;
    this._closeFrameSent = false;
    this._closeMessage = '';
    this._closeTimer = null;
    this._extensions = ***REMOVED******REMOVED***;
    this._protocol = '';
    this._readyState = WebSocket.CONNECTING;
    this._receiver = null;
    this._sender = null;
    this._socket = null;

    if (address !== null) ***REMOVED***
      this._bufferedAmount = 0;
      this._isServer = false;
      this._redirects = 0;

      if (Array.isArray(protocols)) ***REMOVED***
        protocols = protocols.join(', ');
      ***REMOVED*** else if (typeof protocols === 'object' && protocols !== null) ***REMOVED***
        options = protocols;
        protocols = undefined;
      ***REMOVED***

      initAsClient(this, address, protocols, options);
    ***REMOVED*** else ***REMOVED***
      this._isServer = true;
    ***REMOVED***
  ***REMOVED***

  /**
   * This deviates from the WHATWG interface since ws doesn't support the
   * required default "blob" type (instead we define a custom "nodebuffer"
   * type).
   *
   * @type ***REMOVED***String***REMOVED***
   */
  get binaryType() ***REMOVED***
    return this._binaryType;
  ***REMOVED***

  set binaryType(type) ***REMOVED***
    if (!BINARY_TYPES.includes(type)) return;

    this._binaryType = type;

    //
    // Allow to change `binaryType` on the fly.
    //
    if (this._receiver) this._receiver._binaryType = type;
  ***REMOVED***

  /**
   * @type ***REMOVED***Number***REMOVED***
   */
  get bufferedAmount() ***REMOVED***
    if (!this._socket) return this._bufferedAmount;

    return this._socket._writableState.length + this._sender._bufferedBytes;
  ***REMOVED***

  /**
   * @type ***REMOVED***String***REMOVED***
   */
  get extensions() ***REMOVED***
    return Object.keys(this._extensions).join();
  ***REMOVED***

  /**
   * @type ***REMOVED***Function***REMOVED***
   */
  /* istanbul ignore next */
  get onclose() ***REMOVED***
    return undefined;
  ***REMOVED***

  /* istanbul ignore next */
  set onclose(listener) ***REMOVED******REMOVED***

  /**
   * @type ***REMOVED***Function***REMOVED***
   */
  /* istanbul ignore next */
  get onerror() ***REMOVED***
    return undefined;
  ***REMOVED***

  /* istanbul ignore next */
  set onerror(listener) ***REMOVED******REMOVED***

  /**
   * @type ***REMOVED***Function***REMOVED***
   */
  /* istanbul ignore next */
  get onopen() ***REMOVED***
    return undefined;
  ***REMOVED***

  /* istanbul ignore next */
  set onopen(listener) ***REMOVED******REMOVED***

  /**
   * @type ***REMOVED***Function***REMOVED***
   */
  /* istanbul ignore next */
  get onmessage() ***REMOVED***
    return undefined;
  ***REMOVED***

  /* istanbul ignore next */
  set onmessage(listener) ***REMOVED******REMOVED***

  /**
   * @type ***REMOVED***String***REMOVED***
   */
  get protocol() ***REMOVED***
    return this._protocol;
  ***REMOVED***

  /**
   * @type ***REMOVED***Number***REMOVED***
   */
  get readyState() ***REMOVED***
    return this._readyState;
  ***REMOVED***

  /**
   * @type ***REMOVED***String***REMOVED***
   */
  get url() ***REMOVED***
    return this._url;
  ***REMOVED***

  /**
   * Set up the socket and the internal resources.
   *
   * @param ***REMOVED***(net.Socket|tls.Socket)***REMOVED*** socket The network socket between the
   *     server and client
   * @param ***REMOVED***Buffer***REMOVED*** head The first packet of the upgraded stream
   * @param ***REMOVED***Number***REMOVED*** [maxPayload=0] The maximum allowed message size
   * @private
   */
  setSocket(socket, head, maxPayload) ***REMOVED***
    const receiver = new Receiver(
      this.binaryType,
      this._extensions,
      this._isServer,
      maxPayload
    );

    this._sender = new Sender(socket, this._extensions);
    this._receiver = receiver;
    this._socket = socket;

    receiver[kWebSocket] = this;
    socket[kWebSocket] = this;

    receiver.on('conclude', receiverOnConclude);
    receiver.on('drain', receiverOnDrain);
    receiver.on('error', receiverOnError);
    receiver.on('message', receiverOnMessage);
    receiver.on('ping', receiverOnPing);
    receiver.on('pong', receiverOnPong);

    socket.setTimeout(0);
    socket.setNoDelay();

    if (head.length > 0) socket.unshift(head);

    socket.on('close', socketOnClose);
    socket.on('data', socketOnData);
    socket.on('end', socketOnEnd);
    socket.on('error', socketOnError);

    this._readyState = WebSocket.OPEN;
    this.emit('open');
  ***REMOVED***

  /**
   * Emit the `'close'` event.
   *
   * @private
   */
  emitClose() ***REMOVED***
    if (!this._socket) ***REMOVED***
      this._readyState = WebSocket.CLOSED;
      this.emit('close', this._closeCode, this._closeMessage);
      return;
    ***REMOVED***

    if (this._extensions[PerMessageDeflate.extensionName]) ***REMOVED***
      this._extensions[PerMessageDeflate.extensionName].cleanup();
    ***REMOVED***

    this._receiver.removeAllListeners();
    this._readyState = WebSocket.CLOSED;
    this.emit('close', this._closeCode, this._closeMessage);
  ***REMOVED***

  /**
   * Start a closing handshake.
   *
   *          +----------+   +-----------+   +----------+
   *     - - -|ws.close()|-->|close frame|-->|ws.close()|- - -
   *    |     +----------+   +-----------+   +----------+     |
   *          +----------+   +-----------+         |
   * CLOSING  |ws.close()|<--|close frame|<--+-----+       CLOSING
   *          +----------+   +-----------+   |
   *    |           |                        |   +---+        |
   *                +------------------------+-->|fin| - - - -
   *    |         +---+                      |   +---+
   *     - - - - -|fin|<---------------------+
   *              +---+
   *
   * @param ***REMOVED***Number***REMOVED*** [code] Status code explaining why the connection is closing
   * @param ***REMOVED***String***REMOVED*** [data] A string explaining why the connection is closing
   * @public
   */
  close(code, data) ***REMOVED***
    if (this.readyState === WebSocket.CLOSED) return;
    if (this.readyState === WebSocket.CONNECTING) ***REMOVED***
      const msg = 'WebSocket was closed before the connection was established';
      return abortHandshake(this, this._req, msg);
    ***REMOVED***

    if (this.readyState === WebSocket.CLOSING) ***REMOVED***
      if (
        this._closeFrameSent &&
        (this._closeFrameReceived || this._receiver._writableState.errorEmitted)
      ) ***REMOVED***
        this._socket.end();
      ***REMOVED***

      return;
    ***REMOVED***

    this._readyState = WebSocket.CLOSING;
    this._sender.close(code, data, !this._isServer, (err) => ***REMOVED***
      //
      // This error is handled by the `'error'` listener on the socket. We only
      // want to know if the close frame has been sent here.
      //
      if (err) return;

      this._closeFrameSent = true;

      if (
        this._closeFrameReceived ||
        this._receiver._writableState.errorEmitted
      ) ***REMOVED***
        this._socket.end();
      ***REMOVED***
    ***REMOVED***);

    //
    // Specify a timeout for the closing handshake to complete.
    //
    this._closeTimer = setTimeout(
      this._socket.destroy.bind(this._socket),
      closeTimeout
    );
  ***REMOVED***

  /**
   * Send a ping.
   *
   * @param ***REMOVED*******REMOVED*** [data] The data to send
   * @param ***REMOVED***Boolean***REMOVED*** [mask] Indicates whether or not to mask `data`
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback which is executed when the ping is sent
   * @public
   */
  ping(data, mask, cb) ***REMOVED***
    if (this.readyState === WebSocket.CONNECTING) ***REMOVED***
      throw new Error('WebSocket is not open: readyState 0 (CONNECTING)');
    ***REMOVED***

    if (typeof data === 'function') ***REMOVED***
      cb = data;
      data = mask = undefined;
    ***REMOVED*** else if (typeof mask === 'function') ***REMOVED***
      cb = mask;
      mask = undefined;
    ***REMOVED***

    if (typeof data === 'number') data = data.toString();

    if (this.readyState !== WebSocket.OPEN) ***REMOVED***
      sendAfterClose(this, data, cb);
      return;
    ***REMOVED***

    if (mask === undefined) mask = !this._isServer;
    this._sender.ping(data || EMPTY_BUFFER, mask, cb);
  ***REMOVED***

  /**
   * Send a pong.
   *
   * @param ***REMOVED*******REMOVED*** [data] The data to send
   * @param ***REMOVED***Boolean***REMOVED*** [mask] Indicates whether or not to mask `data`
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback which is executed when the pong is sent
   * @public
   */
  pong(data, mask, cb) ***REMOVED***
    if (this.readyState === WebSocket.CONNECTING) ***REMOVED***
      throw new Error('WebSocket is not open: readyState 0 (CONNECTING)');
    ***REMOVED***

    if (typeof data === 'function') ***REMOVED***
      cb = data;
      data = mask = undefined;
    ***REMOVED*** else if (typeof mask === 'function') ***REMOVED***
      cb = mask;
      mask = undefined;
    ***REMOVED***

    if (typeof data === 'number') data = data.toString();

    if (this.readyState !== WebSocket.OPEN) ***REMOVED***
      sendAfterClose(this, data, cb);
      return;
    ***REMOVED***

    if (mask === undefined) mask = !this._isServer;
    this._sender.pong(data || EMPTY_BUFFER, mask, cb);
  ***REMOVED***

  /**
   * Send a data message.
   *
   * @param ***REMOVED*******REMOVED*** data The message to send
   * @param ***REMOVED***Object***REMOVED*** [options] Options object
   * @param ***REMOVED***Boolean***REMOVED*** [options.compress] Specifies whether or not to compress
   *     `data`
   * @param ***REMOVED***Boolean***REMOVED*** [options.binary] Specifies whether `data` is binary or
   *     text
   * @param ***REMOVED***Boolean***REMOVED*** [options.fin=true] Specifies whether the fragment is the
   *     last one
   * @param ***REMOVED***Boolean***REMOVED*** [options.mask] Specifies whether or not to mask `data`
   * @param ***REMOVED***Function***REMOVED*** [cb] Callback which is executed when data is written out
   * @public
   */
  send(data, options, cb) ***REMOVED***
    if (this.readyState === WebSocket.CONNECTING) ***REMOVED***
      throw new Error('WebSocket is not open: readyState 0 (CONNECTING)');
    ***REMOVED***

    if (typeof options === 'function') ***REMOVED***
      cb = options;
      options = ***REMOVED******REMOVED***;
    ***REMOVED***

    if (typeof data === 'number') data = data.toString();

    if (this.readyState !== WebSocket.OPEN) ***REMOVED***
      sendAfterClose(this, data, cb);
      return;
    ***REMOVED***

    const opts = ***REMOVED***
      binary: typeof data !== 'string',
      mask: !this._isServer,
      compress: true,
      fin: true,
      ...options
    ***REMOVED***;

    if (!this._extensions[PerMessageDeflate.extensionName]) ***REMOVED***
      opts.compress = false;
    ***REMOVED***

    this._sender.send(data || EMPTY_BUFFER, opts, cb);
  ***REMOVED***

  /**
   * Forcibly close the connection.
   *
   * @public
   */
  terminate() ***REMOVED***
    if (this.readyState === WebSocket.CLOSED) return;
    if (this.readyState === WebSocket.CONNECTING) ***REMOVED***
      const msg = 'WebSocket was closed before the connection was established';
      return abortHandshake(this, this._req, msg);
    ***REMOVED***

    if (this._socket) ***REMOVED***
      this._readyState = WebSocket.CLOSING;
      this._socket.destroy();
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * @constant ***REMOVED***Number***REMOVED*** CONNECTING
 * @memberof WebSocket
 */
Object.defineProperty(WebSocket, 'CONNECTING', ***REMOVED***
  enumerable: true,
  value: readyStates.indexOf('CONNECTING')
***REMOVED***);

/**
 * @constant ***REMOVED***Number***REMOVED*** CONNECTING
 * @memberof WebSocket.prototype
 */
Object.defineProperty(WebSocket.prototype, 'CONNECTING', ***REMOVED***
  enumerable: true,
  value: readyStates.indexOf('CONNECTING')
***REMOVED***);

/**
 * @constant ***REMOVED***Number***REMOVED*** OPEN
 * @memberof WebSocket
 */
Object.defineProperty(WebSocket, 'OPEN', ***REMOVED***
  enumerable: true,
  value: readyStates.indexOf('OPEN')
***REMOVED***);

/**
 * @constant ***REMOVED***Number***REMOVED*** OPEN
 * @memberof WebSocket.prototype
 */
Object.defineProperty(WebSocket.prototype, 'OPEN', ***REMOVED***
  enumerable: true,
  value: readyStates.indexOf('OPEN')
***REMOVED***);

/**
 * @constant ***REMOVED***Number***REMOVED*** CLOSING
 * @memberof WebSocket
 */
Object.defineProperty(WebSocket, 'CLOSING', ***REMOVED***
  enumerable: true,
  value: readyStates.indexOf('CLOSING')
***REMOVED***);

/**
 * @constant ***REMOVED***Number***REMOVED*** CLOSING
 * @memberof WebSocket.prototype
 */
Object.defineProperty(WebSocket.prototype, 'CLOSING', ***REMOVED***
  enumerable: true,
  value: readyStates.indexOf('CLOSING')
***REMOVED***);

/**
 * @constant ***REMOVED***Number***REMOVED*** CLOSED
 * @memberof WebSocket
 */
Object.defineProperty(WebSocket, 'CLOSED', ***REMOVED***
  enumerable: true,
  value: readyStates.indexOf('CLOSED')
***REMOVED***);

/**
 * @constant ***REMOVED***Number***REMOVED*** CLOSED
 * @memberof WebSocket.prototype
 */
Object.defineProperty(WebSocket.prototype, 'CLOSED', ***REMOVED***
  enumerable: true,
  value: readyStates.indexOf('CLOSED')
***REMOVED***);

[
  'binaryType',
  'bufferedAmount',
  'extensions',
  'protocol',
  'readyState',
  'url'
].forEach((property) => ***REMOVED***
  Object.defineProperty(WebSocket.prototype, property, ***REMOVED*** enumerable: true ***REMOVED***);
***REMOVED***);

//
// Add the `onopen`, `onerror`, `onclose`, and `onmessage` attributes.
// See https://html.spec.whatwg.org/multipage/comms.html#the-websocket-interface
//
['open', 'error', 'close', 'message'].forEach((method) => ***REMOVED***
  Object.defineProperty(WebSocket.prototype, `on$***REMOVED***method***REMOVED***`, ***REMOVED***
    enumerable: true,
    get() ***REMOVED***
      const listeners = this.listeners(method);
      for (let i = 0; i < listeners.length; i++) ***REMOVED***
        if (listeners[i]._listener) return listeners[i]._listener;
      ***REMOVED***

      return undefined;
***REMOVED***
    set(listener) ***REMOVED***
      const listeners = this.listeners(method);
      for (let i = 0; i < listeners.length; i++) ***REMOVED***
        //
        // Remove only the listeners added via `addEventListener`.
        //
        if (listeners[i]._listener) this.removeListener(method, listeners[i]);
      ***REMOVED***
      this.addEventListener(method, listener);
    ***REMOVED***
  ***REMOVED***);
***REMOVED***);

WebSocket.prototype.addEventListener = addEventListener;
WebSocket.prototype.removeEventListener = removeEventListener;

module.exports = WebSocket;

/**
 * Initialize a WebSocket client.
 *
 * @param ***REMOVED***WebSocket***REMOVED*** websocket The client to initialize
 * @param ***REMOVED***(String|URL)***REMOVED*** address The URL to which to connect
 * @param ***REMOVED***String***REMOVED*** [protocols] The subprotocols
 * @param ***REMOVED***Object***REMOVED*** [options] Connection options
 * @param ***REMOVED***(Boolean|Object)***REMOVED*** [options.perMessageDeflate=true] Enable/disable
 *     permessage-deflate
 * @param ***REMOVED***Number***REMOVED*** [options.handshakeTimeout] Timeout in milliseconds for the
 *     handshake request
 * @param ***REMOVED***Number***REMOVED*** [options.protocolVersion=13] Value of the
 *     `Sec-WebSocket-Version` header
 * @param ***REMOVED***String***REMOVED*** [options.origin] Value of the `Origin` or
 *     `Sec-WebSocket-Origin` header
 * @param ***REMOVED***Number***REMOVED*** [options.maxPayload=104857600] The maximum allowed message
 *     size
 * @param ***REMOVED***Boolean***REMOVED*** [options.followRedirects=false] Whether or not to follow
 *     redirects
 * @param ***REMOVED***Number***REMOVED*** [options.maxRedirects=10] The maximum number of redirects
 *     allowed
 * @private
 */
function initAsClient(websocket, address, protocols, options) ***REMOVED***
  const opts = ***REMOVED***
    protocolVersion: protocolVersions[1],
    maxPayload: 100 * 1024 * 1024,
    perMessageDeflate: true,
    followRedirects: false,
    maxRedirects: 10,
    ...options,
    createConnection: undefined,
    socketPath: undefined,
    hostname: undefined,
    protocol: undefined,
    timeout: undefined,
    method: undefined,
    host: undefined,
    path: undefined,
    port: undefined
  ***REMOVED***;

  if (!protocolVersions.includes(opts.protocolVersion)) ***REMOVED***
    throw new RangeError(
      `Unsupported protocol version: $***REMOVED***opts.protocolVersion***REMOVED*** ` +
        `(supported versions: $***REMOVED***protocolVersions.join(', ')***REMOVED***)`
    );
  ***REMOVED***

  let parsedUrl;

  if (address instanceof URL) ***REMOVED***
    parsedUrl = address;
    websocket._url = address.href;
  ***REMOVED*** else ***REMOVED***
    parsedUrl = new URL(address);
    websocket._url = address;
  ***REMOVED***

  const isUnixSocket = parsedUrl.protocol === 'ws+unix:';

  if (!parsedUrl.host && (!isUnixSocket || !parsedUrl.pathname)) ***REMOVED***
    const err = new Error(`Invalid URL: $***REMOVED***websocket.url***REMOVED***`);

    if (websocket._redirects === 0) ***REMOVED***
      throw err;
    ***REMOVED*** else ***REMOVED***
      emitErrorAndClose(websocket, err);
      return;
    ***REMOVED***
  ***REMOVED***

  const isSecure =
    parsedUrl.protocol === 'wss:' || parsedUrl.protocol === 'https:';
  const defaultPort = isSecure ? 443 : 80;
  const key = randomBytes(16).toString('base64');
  const get = isSecure ? https.get : http.get;
  let perMessageDeflate;

  opts.createConnection = isSecure ? tlsConnect : netConnect;
  opts.defaultPort = opts.defaultPort || defaultPort;
  opts.port = parsedUrl.port || defaultPort;
  opts.host = parsedUrl.hostname.startsWith('[')
    ? parsedUrl.hostname.slice(1, -1)
    : parsedUrl.hostname;
  opts.headers = ***REMOVED***
    'Sec-WebSocket-Version': opts.protocolVersion,
    'Sec-WebSocket-Key': key,
    Connection: 'Upgrade',
    Upgrade: 'websocket',
    ...opts.headers
  ***REMOVED***;
  opts.path = parsedUrl.pathname + parsedUrl.search;
  opts.timeout = opts.handshakeTimeout;

  if (opts.perMessageDeflate) ***REMOVED***
    perMessageDeflate = new PerMessageDeflate(
      opts.perMessageDeflate !== true ? opts.perMessageDeflate : ***REMOVED******REMOVED***,
      false,
      opts.maxPayload
    );
    opts.headers['Sec-WebSocket-Extensions'] = format(***REMOVED***
      [PerMessageDeflate.extensionName]: perMessageDeflate.offer()
    ***REMOVED***);
  ***REMOVED***
  if (protocols) ***REMOVED***
    opts.headers['Sec-WebSocket-Protocol'] = protocols;
  ***REMOVED***
  if (opts.origin) ***REMOVED***
    if (opts.protocolVersion < 13) ***REMOVED***
      opts.headers['Sec-WebSocket-Origin'] = opts.origin;
    ***REMOVED*** else ***REMOVED***
      opts.headers.Origin = opts.origin;
    ***REMOVED***
  ***REMOVED***
  if (parsedUrl.username || parsedUrl.password) ***REMOVED***
    opts.auth = `$***REMOVED***parsedUrl.username***REMOVED***:$***REMOVED***parsedUrl.password***REMOVED***`;
  ***REMOVED***

  if (isUnixSocket) ***REMOVED***
    const parts = opts.path.split(':');

    opts.socketPath = parts[0];
    opts.path = parts[1];
  ***REMOVED***

  let req = (websocket._req = get(opts));

  if (opts.timeout) ***REMOVED***
    req.on('timeout', () => ***REMOVED***
      abortHandshake(websocket, req, 'Opening handshake has timed out');
    ***REMOVED***);
  ***REMOVED***

  req.on('error', (err) => ***REMOVED***
    if (req === null || req.aborted) return;

    req = websocket._req = null;
    emitErrorAndClose(websocket, err);
  ***REMOVED***);

  req.on('response', (res) => ***REMOVED***
    const location = res.headers.location;
    const statusCode = res.statusCode;

    if (
      location &&
      opts.followRedirects &&
      statusCode >= 300 &&
      statusCode < 400
    ) ***REMOVED***
      if (++websocket._redirects > opts.maxRedirects) ***REMOVED***
        abortHandshake(websocket, req, 'Maximum redirects exceeded');
        return;
      ***REMOVED***

      req.abort();

      let addr;

      try ***REMOVED***
        addr = new URL(location, address);
      ***REMOVED*** catch (err) ***REMOVED***
        emitErrorAndClose(websocket, err);
        return;
      ***REMOVED***

      initAsClient(websocket, addr, protocols, options);
    ***REMOVED*** else if (!websocket.emit('unexpected-response', req, res)) ***REMOVED***
      abortHandshake(
        websocket,
        req,
        `Unexpected server response: $***REMOVED***res.statusCode***REMOVED***`
      );
    ***REMOVED***
  ***REMOVED***);

  req.on('upgrade', (res, socket, head) => ***REMOVED***
    websocket.emit('upgrade', res);

    //
    // The user may have closed the connection from a listener of the `upgrade`
    // event.
    //
    if (websocket.readyState !== WebSocket.CONNECTING) return;

    req = websocket._req = null;

    const digest = createHash('sha1')
      .update(key + GUID)
      .digest('base64');

    if (res.headers['sec-websocket-accept'] !== digest) ***REMOVED***
      abortHandshake(websocket, socket, 'Invalid Sec-WebSocket-Accept header');
      return;
    ***REMOVED***

    const serverProt = res.headers['sec-websocket-protocol'];
    const protList = (protocols || '').split(/, */);
    let protError;

    if (!protocols && serverProt) ***REMOVED***
      protError = 'Server sent a subprotocol but none was requested';
    ***REMOVED*** else if (protocols && !serverProt) ***REMOVED***
      protError = 'Server sent no subprotocol';
    ***REMOVED*** else if (serverProt && !protList.includes(serverProt)) ***REMOVED***
      protError = 'Server sent an invalid subprotocol';
    ***REMOVED***

    if (protError) ***REMOVED***
      abortHandshake(websocket, socket, protError);
      return;
    ***REMOVED***

    if (serverProt) websocket._protocol = serverProt;

    const secWebSocketExtensions = res.headers['sec-websocket-extensions'];

    if (secWebSocketExtensions !== undefined) ***REMOVED***
      if (!perMessageDeflate) ***REMOVED***
        const message =
          'Server sent a Sec-WebSocket-Extensions header but no extension ' +
          'was requested';
        abortHandshake(websocket, socket, message);
        return;
      ***REMOVED***

      let extensions;

      try ***REMOVED***
        extensions = parse(secWebSocketExtensions);
      ***REMOVED*** catch (err) ***REMOVED***
        const message = 'Invalid Sec-WebSocket-Extensions header';
        abortHandshake(websocket, socket, message);
        return;
      ***REMOVED***

      const extensionNames = Object.keys(extensions);

      if (extensionNames.length) ***REMOVED***
        if (
          extensionNames.length !== 1 ||
          extensionNames[0] !== PerMessageDeflate.extensionName
        ) ***REMOVED***
          const message =
            'Server indicated an extension that was not requested';
          abortHandshake(websocket, socket, message);
          return;
        ***REMOVED***

        try ***REMOVED***
          perMessageDeflate.accept(extensions[PerMessageDeflate.extensionName]);
        ***REMOVED*** catch (err) ***REMOVED***
          const message = 'Invalid Sec-WebSocket-Extensions header';
          abortHandshake(websocket, socket, message);
          return;
        ***REMOVED***

        websocket._extensions[PerMessageDeflate.extensionName] =
          perMessageDeflate;
      ***REMOVED***
    ***REMOVED***

    websocket.setSocket(socket, head, opts.maxPayload);
  ***REMOVED***);
***REMOVED***

/**
 * Emit the `'error'` and `'close'` event.
 *
 * @param ***REMOVED***WebSocket***REMOVED*** websocket The WebSocket instance
 * @param ***REMOVED***Error***REMOVED*** The error to emit
 * @private
 */
function emitErrorAndClose(websocket, err) ***REMOVED***
  websocket._readyState = WebSocket.CLOSING;
  websocket.emit('error', err);
  websocket.emitClose();
***REMOVED***

/**
 * Create a `net.Socket` and initiate a connection.
 *
 * @param ***REMOVED***Object***REMOVED*** options Connection options
 * @return ***REMOVED***net.Socket***REMOVED*** The newly created socket used to start the connection
 * @private
 */
function netConnect(options) ***REMOVED***
  options.path = options.socketPath;
  return net.connect(options);
***REMOVED***

/**
 * Create a `tls.TLSSocket` and initiate a connection.
 *
 * @param ***REMOVED***Object***REMOVED*** options Connection options
 * @return ***REMOVED***tls.TLSSocket***REMOVED*** The newly created socket used to start the connection
 * @private
 */
function tlsConnect(options) ***REMOVED***
  options.path = undefined;

  if (!options.servername && options.servername !== '') ***REMOVED***
    options.servername = net.isIP(options.host) ? '' : options.host;
  ***REMOVED***

  return tls.connect(options);
***REMOVED***

/**
 * Abort the handshake and emit an error.
 *
 * @param ***REMOVED***WebSocket***REMOVED*** websocket The WebSocket instance
 * @param ***REMOVED***(http.ClientRequest|net.Socket|tls.Socket)***REMOVED*** stream The request to
 *     abort or the socket to destroy
 * @param ***REMOVED***String***REMOVED*** message The error message
 * @private
 */
function abortHandshake(websocket, stream, message) ***REMOVED***
  websocket._readyState = WebSocket.CLOSING;

  const err = new Error(message);
  Error.captureStackTrace(err, abortHandshake);

  if (stream.setHeader) ***REMOVED***
    stream.abort();

    if (stream.socket && !stream.socket.destroyed) ***REMOVED***
      //
      // On Node.js >= 14.3.0 `request.abort()` does not destroy the socket if
      // called after the request completed. See
      // https://github.com/websockets/ws/issues/1869.
      //
      stream.socket.destroy();
    ***REMOVED***

    stream.once('abort', websocket.emitClose.bind(websocket));
    websocket.emit('error', err);
  ***REMOVED*** else ***REMOVED***
    stream.destroy(err);
    stream.once('error', websocket.emit.bind(websocket, 'error'));
    stream.once('close', websocket.emitClose.bind(websocket));
  ***REMOVED***
***REMOVED***

/**
 * Handle cases where the `ping()`, `pong()`, or `send()` methods are called
 * when the `readyState` attribute is `CLOSING` or `CLOSED`.
 *
 * @param ***REMOVED***WebSocket***REMOVED*** websocket The WebSocket instance
 * @param ***REMOVED*******REMOVED*** [data] The data to send
 * @param ***REMOVED***Function***REMOVED*** [cb] Callback
 * @private
 */
function sendAfterClose(websocket, data, cb) ***REMOVED***
  if (data) ***REMOVED***
    const length = toBuffer(data).length;

    //
    // The `_bufferedAmount` property is used only when the peer is a client and
    // the opening handshake fails. Under these circumstances, in fact, the
    // `setSocket()` method is not called, so the `_socket` and `_sender`
    // properties are set to `null`.
    //
    if (websocket._socket) websocket._sender._bufferedBytes += length;
    else websocket._bufferedAmount += length;
  ***REMOVED***

  if (cb) ***REMOVED***
    const err = new Error(
      `WebSocket is not open: readyState $***REMOVED***websocket.readyState***REMOVED*** ` +
        `($***REMOVED***readyStates[websocket.readyState]***REMOVED***)`
    );
    cb(err);
  ***REMOVED***
***REMOVED***

/**
 * The listener of the `Receiver` `'conclude'` event.
 *
 * @param ***REMOVED***Number***REMOVED*** code The status code
 * @param ***REMOVED***String***REMOVED*** reason The reason for closing
 * @private
 */
function receiverOnConclude(code, reason) ***REMOVED***
  const websocket = this[kWebSocket];

  websocket._closeFrameReceived = true;
  websocket._closeMessage = reason;
  websocket._closeCode = code;

  if (websocket._socket[kWebSocket] === undefined) return;

  websocket._socket.removeListener('data', socketOnData);
  process.nextTick(resume, websocket._socket);

  if (code === 1005) websocket.close();
  else websocket.close(code, reason);
***REMOVED***

/**
 * The listener of the `Receiver` `'drain'` event.
 *
 * @private
 */
function receiverOnDrain() ***REMOVED***
  this[kWebSocket]._socket.resume();
***REMOVED***

/**
 * The listener of the `Receiver` `'error'` event.
 *
 * @param ***REMOVED***(RangeError|Error)***REMOVED*** err The emitted error
 * @private
 */
function receiverOnError(err) ***REMOVED***
  const websocket = this[kWebSocket];

  if (websocket._socket[kWebSocket] !== undefined) ***REMOVED***
    websocket._socket.removeListener('data', socketOnData);

    //
    // On Node.js < 14.0.0 the `'error'` event is emitted synchronously. See
    // https://github.com/websockets/ws/issues/1940.
    //
    process.nextTick(resume, websocket._socket);

    websocket.close(err[kStatusCode]);
  ***REMOVED***

  websocket.emit('error', err);
***REMOVED***

/**
 * The listener of the `Receiver` `'finish'` event.
 *
 * @private
 */
function receiverOnFinish() ***REMOVED***
  this[kWebSocket].emitClose();
***REMOVED***

/**
 * The listener of the `Receiver` `'message'` event.
 *
 * @param ***REMOVED***(String|Buffer|ArrayBuffer|Buffer[])***REMOVED*** data The message
 * @private
 */
function receiverOnMessage(data) ***REMOVED***
  this[kWebSocket].emit('message', data);
***REMOVED***

/**
 * The listener of the `Receiver` `'ping'` event.
 *
 * @param ***REMOVED***Buffer***REMOVED*** data The data included in the ping frame
 * @private
 */
function receiverOnPing(data) ***REMOVED***
  const websocket = this[kWebSocket];

  websocket.pong(data, !websocket._isServer, NOOP);
  websocket.emit('ping', data);
***REMOVED***

/**
 * The listener of the `Receiver` `'pong'` event.
 *
 * @param ***REMOVED***Buffer***REMOVED*** data The data included in the pong frame
 * @private
 */
function receiverOnPong(data) ***REMOVED***
  this[kWebSocket].emit('pong', data);
***REMOVED***

/**
 * Resume a readable stream
 *
 * @param ***REMOVED***Readable***REMOVED*** stream The readable stream
 * @private
 */
function resume(stream) ***REMOVED***
  stream.resume();
***REMOVED***

/**
 * The listener of the `net.Socket` `'close'` event.
 *
 * @private
 */
function socketOnClose() ***REMOVED***
  const websocket = this[kWebSocket];

  this.removeListener('close', socketOnClose);
  this.removeListener('data', socketOnData);
  this.removeListener('end', socketOnEnd);

  websocket._readyState = WebSocket.CLOSING;

  let chunk;

  //
  // The close frame might not have been received or the `'end'` event emitted,
  // for example, if the socket was destroyed due to an error. Ensure that the
  // `receiver` stream is closed after writing any remaining buffered data to
  // it. If the readable side of the socket is in flowing mode then there is no
  // buffered data as everything has been already written and `readable.read()`
  // will return `null`. If instead, the socket is paused, any possible buffered
  // data will be read as a single chunk.
  //
  if (
    !this._readableState.endEmitted &&
    !websocket._closeFrameReceived &&
    !websocket._receiver._writableState.errorEmitted &&
    (chunk = websocket._socket.read()) !== null
  ) ***REMOVED***
    websocket._receiver.write(chunk);
  ***REMOVED***

  websocket._receiver.end();

  this[kWebSocket] = undefined;

  clearTimeout(websocket._closeTimer);

  if (
    websocket._receiver._writableState.finished ||
    websocket._receiver._writableState.errorEmitted
  ) ***REMOVED***
    websocket.emitClose();
  ***REMOVED*** else ***REMOVED***
    websocket._receiver.on('error', receiverOnFinish);
    websocket._receiver.on('finish', receiverOnFinish);
  ***REMOVED***
***REMOVED***

/**
 * The listener of the `net.Socket` `'data'` event.
 *
 * @param ***REMOVED***Buffer***REMOVED*** chunk A chunk of data
 * @private
 */
function socketOnData(chunk) ***REMOVED***
  if (!this[kWebSocket]._receiver.write(chunk)) ***REMOVED***
    this.pause();
  ***REMOVED***
***REMOVED***

/**
 * The listener of the `net.Socket` `'end'` event.
 *
 * @private
 */
function socketOnEnd() ***REMOVED***
  const websocket = this[kWebSocket];

  websocket._readyState = WebSocket.CLOSING;
  websocket._receiver.end();
  this.end();
***REMOVED***

/**
 * The listener of the `net.Socket` `'error'` event.
 *
 * @private
 */
function socketOnError() ***REMOVED***
  const websocket = this[kWebSocket];

  this.removeListener('error', socketOnError);
  this.on('error', NOOP);

  if (websocket) ***REMOVED***
    websocket._readyState = WebSocket.CLOSING;
    this.destroy();
  ***REMOVED***
***REMOVED***
