// @d-fischer/logger/lib/LogLevel.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.LogLevelToConsoleFunction = exports.resolveLogLevel = exports.LogLevel = void 0;
var isNode = require("detect-node");
var LogLevel;
(function (LogLevel) ***REMOVED***
    LogLevel[LogLevel["CRITICAL"] = 0] = "CRITICAL";
    LogLevel[LogLevel["ERROR"] = 1] = "ERROR";
    LogLevel[LogLevel["WARNING"] = 2] = "WARNING";
    LogLevel[LogLevel["INFO"] = 3] = "INFO";
    LogLevel[LogLevel["DEBUG"] = 4] = "DEBUG";
    LogLevel[LogLevel["TRACE"] = 7] = "TRACE";
***REMOVED***)(LogLevel || (LogLevel = ***REMOVED******REMOVED***));
exports.LogLevel = LogLevel;
function resolveLogLevel(level) ***REMOVED***
    if (typeof level === 'number') ***REMOVED***
        if (Object.prototype.hasOwnProperty.call(LogLevel, level)) ***REMOVED***
            return level;
        ***REMOVED***
        var eligibleLevels = Object.keys(LogLevel)
            .map(function (k) ***REMOVED*** return parseInt(k, 10); ***REMOVED***)
            .filter(function (k) ***REMOVED*** return !isNaN(k) && k < level; ***REMOVED***);
        if (!eligibleLevels.length) ***REMOVED***
            return LogLevel.WARNING;
        ***REMOVED***
        return Math.max.apply(Math, eligibleLevels);
    ***REMOVED***
    // TODO drop the replace for next major, it keeps the old deprecated debug1/2/3 levels running
    var strLevel = level.replace(/\d+$/, '').toUpperCase();
    if (!Object.prototype.hasOwnProperty.call(LogLevel, strLevel)) ***REMOVED***
        throw new Error("Unknown log level string: " + level);
    ***REMOVED***
    return LogLevel[strLevel];
***REMOVED***
exports.resolveLogLevel = resolveLogLevel;
// Node 8+ defines console.debug as noop, and earlier versions don't define it at all
var debugFunction = isNode ? console.log.bind(console) : console.debug.bind(console);
// eslint-disable-next-line @typescript-eslint/no-explicit-any
exports.LogLevelToConsoleFunction = (_a = ***REMOVED******REMOVED***,
    _a[LogLevel.CRITICAL] = console.error.bind(console),
    _a[LogLevel.ERROR] = console.error.bind(console),
    _a[LogLevel.WARNING] = console.warn.bind(console),
    _a[LogLevel.INFO] = console.info.bind(console),
    _a[LogLevel.DEBUG] = debugFunction.bind(console),
    _a[LogLevel.TRACE] = console.trace.bind(console),
    _a);
