// ircv3/es/IrcClient.mjs
import ***REMOVED*** __assign, __awaiter, __decorate, __extends, __generator, __read, __spread, __values ***REMOVED*** from "tslib";
import ***REMOVED*** DirectConnection, PersistentConnection, WebSocketConnection ***REMOVED*** from '@d-fischer/connection';
import ***REMOVED*** createLogger ***REMOVED*** from '@d-fischer/logger';
import ***REMOVED*** arrayToObject, Enumerable, forEachObjectEntry, padLeft, resolveConfigValue, splitWithLimit ***REMOVED*** from '@d-fischer/shared-utils';
import ***REMOVED*** EventEmitter ***REMOVED*** from '@d-fischer/typed-event-emitter';
import ***REMOVED*** klona ***REMOVED*** from 'klona/json';
import * as CoreCapabilities from "./Capability/CoreCapabilities/index.mjs";
import ***REMOVED*** MessageError ***REMOVED*** from "./Errors/MessageError.mjs";
import ***REMOVED*** createMessage ***REMOVED*** from "./Message/Message.mjs";
import ***REMOVED*** MessageCollector ***REMOVED*** from "./Message/MessageCollector.mjs";
import ***REMOVED*** parseMessage ***REMOVED*** from "./Message/MessageParser.mjs";
import * as MessageTypes from "./Message/MessageTypes/index.mjs";
import ***REMOVED*** CapabilityNegotiation, ChannelJoin, ChannelPart, ClientQuit, ErrorMessage, NickChange, Notice, Password, Ping, Pong, PrivateMessage, UserRegistration ***REMOVED*** from "./Message/MessageTypes/Commands/index.mjs";
import ***REMOVED*** Error422NoMotd, Error462AlreadyRegistered, Reply001Welcome, Reply004ServerInfo, Reply005Isupport, Reply376EndOfMotd ***REMOVED*** from "./Message/MessageTypes/Numerics/index.mjs";
import ***REMOVED*** defaultServerProperties ***REMOVED*** from "./ServerProperties.mjs";
import ***REMOVED*** decodeCtcp ***REMOVED*** from "./Toolkit/StringTools.mjs";
var IrcClient = /** @class */ (function (_super) ***REMOVED***
    __extends(IrcClient, _super);
    function IrcClient(options) ***REMOVED***
        var e_1, _a;
        var _this = _super.call(this) || this;
        _this._registered = false;
        _this._supportsCapabilities = true;
        _this._events = new Map();
        _this._registeredMessageTypes = new Map();
        // emitted events
        _this.onConnect = _this.registerEvent();
        _this.onRegister = _this.registerEvent();
        _this.onDisconnect = _this.registerEvent();
        _this.onPrivmsg = _this.registerEvent();
        _this.onAction = _this.registerEvent();
        _this.onNotice = _this.registerEvent();
        _this.onNickChange = _this.registerEvent();
        _this.onCtcp = _this.registerEvent();
        _this.onCtcpReply = _this.registerEvent();
        _this.onAnyMessage = _this.registerEvent();
        _this._serverProperties = klona(defaultServerProperties);
        _this._supportedFeatures = ***REMOVED******REMOVED***;
        _this._collectors = [];
        _this._clientCapabilities = new Map();
        _this._serverCapabilities = new Map();
        _this._negotiatedCapabilities = new Map();
        _this._initialConnectionSetupDone = false;
        var connection = options.connection, credentials = options.credentials, channels = options.channels, channelTypes = options.channelTypes, webSocket = options.webSocket, _b = options.logger, logger = _b === void 0 ? ***REMOVED******REMOVED*** : _b;
        _this._options = options;
        var _c = connection.pingOnInactivity, pingOnInactivity = _c === void 0 ? 60 : _c, _d = connection.pingTimeout, pingTimeout = _d === void 0 ? 10 : _d;
        _this._pingOnInactivity = pingOnInactivity;
        _this._pingTimeout = pingTimeout;
        _this._currentNick = credentials.nick;
        _this._logger = createLogger(__assign(***REMOVED*** name: 'ircv3', emoji: true ***REMOVED***, logger));
        _this.registerCoreMessageTypes();
        var hostName = connection.hostName, secure = connection.secure, _e = connection.reconnect, reconnect = _e === void 0 ? true : _e;
        var connectionOptions = ***REMOVED***
            hostName: hostName,
            port: _this.port,
            secure: secure,
            lineBased: true
        ***REMOVED***;
        var ConnectionType = webSocket ? WebSocketConnection : DirectConnection;
        if (reconnect) ***REMOVED***
            _this._connection = new PersistentConnection(ConnectionType, connectionOptions, ***REMOVED*** logger: _this._logger ***REMOVED***, options.connectionOptions);
        ***REMOVED***
        else ***REMOVED***
            _this._connection = new ConnectionType(connectionOptions, _this._logger, options.connectionOptions);
        ***REMOVED***
        try ***REMOVED***
            for (var _f = __values(Object.values(CoreCapabilities)), _g = _f.next(); !_g.done; _g = _f.next()) ***REMOVED***
                var cap = _g.value;
                _this.addCapability(cap);
            ***REMOVED***
        ***REMOVED***
        catch (e_1_1) ***REMOVED*** e_1 = ***REMOVED*** error: e_1_1 ***REMOVED***; ***REMOVED***
        finally ***REMOVED***
            try ***REMOVED***
                if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
            ***REMOVED***
            finally ***REMOVED*** if (e_1) throw e_1.error; ***REMOVED***
        ***REMOVED***
        if (channels) ***REMOVED***
            _this.onRegister(function () ***REMOVED*** return __awaiter(_this, void 0, void 0, function () ***REMOVED***
                var resolvedChannels, resolvedChannels_1, resolvedChannels_1_1, channel;
                var e_2, _a;
                return __generator(this, function (_b) ***REMOVED***
                    switch (_b.label) ***REMOVED***
                        case 0: return [4 /*yield*/, resolveConfigValue(channels)];
                        case 1:
                            resolvedChannels = _b.sent();
                            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                            if (resolvedChannels) ***REMOVED***
                                try ***REMOVED***
                                    for (resolvedChannels_1 = __values(resolvedChannels), resolvedChannels_1_1 = resolvedChannels_1.next(); !resolvedChannels_1_1.done; resolvedChannels_1_1 = resolvedChannels_1.next()) ***REMOVED***
                                        channel = resolvedChannels_1_1.value;
                                        this.join(channel);
                                    ***REMOVED***
                                ***REMOVED***
                                catch (e_2_1) ***REMOVED*** e_2 = ***REMOVED*** error: e_2_1 ***REMOVED***; ***REMOVED***
                                finally ***REMOVED***
                                    try ***REMOVED***
                                        if (resolvedChannels_1_1 && !resolvedChannels_1_1.done && (_a = resolvedChannels_1.return)) _a.call(resolvedChannels_1);
                                    ***REMOVED***
                                    finally ***REMOVED*** if (e_2) throw e_2.error; ***REMOVED***
                                ***REMOVED***
                            ***REMOVED***
                            return [2 /*return*/];
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***); ***REMOVED***);
        ***REMOVED***
        _this.onTypedMessage(CapabilityNegotiation, function (_a) ***REMOVED***
            var _b = _a.params, subCommand = _b.subCommand, capabilities = _b.capabilities;
            return __awaiter(_this, void 0, void 0, function () ***REMOVED***
                var caps, _c, capList, _d, _e, _f, name, cap, capNames_1, caps_1, caps_1_1, cap;
                var e_3, _g, e_4, _h;
                return __generator(this, function (_j) ***REMOVED***
                    switch (_j.label) ***REMOVED***
                        case 0:
                            caps = capabilities.split(' ');
                            _c = subCommand.toUpperCase();
                            switch (_c) ***REMOVED***
                                case 'NEW': return [3 /*break*/, 1];
                                case 'DEL': return [3 /*break*/, 3];
                            ***REMOVED***
                            return [3 /*break*/, 4];
                        case 1:
                            this._logger.debug("Server registered new capabilities: " + caps.join(', '));
                            capList = arrayToObject(caps, function (part) ***REMOVED***
                                var _a;
                                if (!part) ***REMOVED***
                                    return ***REMOVED******REMOVED***;
                                ***REMOVED***
                                var _b = __read(splitWithLimit(part, '=', 2), 2), cap = _b[0], param = _b[1];
                                return _a = ***REMOVED******REMOVED***,
                                    _a[cap] = ***REMOVED***
                                        name: cap,
                                        param: param || true
                                ***REMOVED***
                                    _a;
                            ***REMOVED***);
                            try ***REMOVED***
                                for (_d = __values(Object.entries(capList)), _e = _d.next(); !_e.done; _e = _d.next()) ***REMOVED***
                                    _f = __read(_e.value, 2), name = _f[0], cap = _f[1];
                                    this._serverCapabilities.set(name, cap);
                                ***REMOVED***
                            ***REMOVED***
                            catch (e_3_1) ***REMOVED*** e_3 = ***REMOVED*** error: e_3_1 ***REMOVED***; ***REMOVED***
                            finally ***REMOVED***
                                try ***REMOVED***
                                    if (_e && !_e.done && (_g = _d.return)) _g.call(_d);
                                ***REMOVED***
                                finally ***REMOVED*** if (e_3) throw e_3.error; ***REMOVED***
                            ***REMOVED***
                            capNames_1 = Object.keys(capList);
                            return [4 /*yield*/, this._negotiateCapabilities(Array.from(this._clientCapabilities.entries())
                                    .filter(function (_a) ***REMOVED***
                                    var _b = __read(_a, 1), name = _b[0];
                                    return capNames_1.includes(name);
                                ***REMOVED***)
                                    .map(function (_a) ***REMOVED***
                                    var _b = __read(_a, 2), cap = _b[1];
                                    return cap;
                                ***REMOVED***))];
                        case 2:
                            _j.sent();
                            return [3 /*break*/, 4];
                        case 3:
                            ***REMOVED***
                                this._logger.debug("Server removed capabilities: " + caps.join(', '));
                                try ***REMOVED***
                                    for (caps_1 = __values(caps), caps_1_1 = caps_1.next(); !caps_1_1.done; caps_1_1 = caps_1.next()) ***REMOVED***
                                        cap = caps_1_1.value;
                                        this._serverCapabilities.delete(cap);
                                        this._negotiatedCapabilities.delete(cap);
                                    ***REMOVED***
                                ***REMOVED***
                                catch (e_4_1) ***REMOVED*** e_4 = ***REMOVED*** error: e_4_1 ***REMOVED***; ***REMOVED***
                                finally ***REMOVED***
                                    try ***REMOVED***
                                        if (caps_1_1 && !caps_1_1.done && (_h = caps_1.return)) _h.call(caps_1);
                                    ***REMOVED***
                                    finally ***REMOVED*** if (e_4) throw e_4.error; ***REMOVED***
                                ***REMOVED***
                            ***REMOVED***
                            _j.label = 4;
                        case 4: return [2 /*return*/];
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***);
        _this.onTypedMessage(Ping, function (_a) ***REMOVED***
            var message = _a.params.message;
            _this.sendMessage(Pong, ***REMOVED*** message: message ***REMOVED***);
        ***REMOVED***);
        _this.onTypedMessage(Reply001Welcome, function (_a) ***REMOVED***
            var me = _a.params.me;
            _this._handleReceivedClientNick(me);
        ***REMOVED***);
        _this.onTypedMessage(Reply004ServerInfo, function (_a) ***REMOVED***
            var userModes = _a.params.userModes;
            if (userModes) ***REMOVED***
                _this._serverProperties.supportedUserModes = userModes;
            ***REMOVED***
        ***REMOVED***);
        _this.onTypedMessage(Reply005Isupport, function (_a) ***REMOVED***
            var supports = _a.params.supports;
            var newFeatures = arrayToObject(supports.split(' '), function (part) ***REMOVED***
                var _a;
                var _b = __read(splitWithLimit(part, '=', 2), 2), support = _b[0], param = _b[1];
                return _a = ***REMOVED******REMOVED***, _a[support] = param || true, _a;
            ***REMOVED***);
            _this._supportedFeatures = __assign(__assign(***REMOVED******REMOVED***, _this._supportedFeatures), newFeatures);
        ***REMOVED***);
        _this.onTypedMessage(Reply376EndOfMotd, function (_a) ***REMOVED***
            var me = _a.params.me;
            if (!_this._registered) ***REMOVED***
                _this._handleReceivedClientNick(me);
                _this._registered = true;
                _this.emit(_this.onRegister);
            ***REMOVED***
        ***REMOVED***);
        _this.onTypedMessage(Error422NoMotd, function (_a) ***REMOVED***
            var me = _a.params.me;
            if (!_this._registered) ***REMOVED***
                _this._handleReceivedClientNick(me);
                _this._registered = true;
                _this.emit(_this.onRegister);
            ***REMOVED***
        ***REMOVED***);
        _this.onTypedMessage(Error462AlreadyRegistered, function (_a) ***REMOVED***
            var me = _a.params.me;
            // what, I thought we are not registered yet?
            if (!_this._registered) ***REMOVED***
                // screw this, we are now.
                _this._logger.warn("We thought we're not registered yet, but we actually are");
                _this._handleReceivedClientNick(me);
                _this._registered = true;
                _this.emit(_this.onRegister);
            ***REMOVED***
        ***REMOVED***);
        _this.onTypedMessage(PrivateMessage, function (msg) ***REMOVED***
            var _a;
            var _b = msg.params, target = _b.target, message = _b.message;
            var ctcpMessage = decodeCtcp(message);
            var nick = (_a = msg.prefix) === null || _a === void 0 ? void 0 : _a.nick;
            if (ctcpMessage) ***REMOVED***
                if (ctcpMessage.command === 'ACTION') ***REMOVED***
                    _this.emit(_this.onAction, target, nick, ctcpMessage.params, msg);
                ***REMOVED***
                else ***REMOVED***
                    _this.emit(_this.onCtcp, target, nick, ctcpMessage.command, ctcpMessage.params, msg);
                ***REMOVED***
            ***REMOVED***
            else ***REMOVED***
                _this.emit(_this.onPrivmsg, target, nick, message, msg);
            ***REMOVED***
        ***REMOVED***);
        _this.onTypedMessage(NickChange, function (msg) ***REMOVED***
            var _a;
            var newNick = msg.params.nick;
            var oldNick = (_a = msg.prefix) === null || _a === void 0 ? void 0 : _a.nick;
            if (oldNick === _this._currentNick) ***REMOVED***
                _this._currentNick = newNick;
            ***REMOVED***
            _this.emit(_this.onNickChange, oldNick, newNick, msg);
        ***REMOVED***);
        _this.onTypedMessage(Notice, function (msg) ***REMOVED***
            var _a;
            var _b = msg.params, target = _b.target, message = _b.message;
            var ctcpMessage = decodeCtcp(message);
            var nick = (_a = msg.prefix) === null || _a === void 0 ? void 0 : _a.nick;
            if (ctcpMessage) ***REMOVED***
                _this.emit(_this.onCtcpReply, target, nick, ctcpMessage.command, ctcpMessage.params, msg);
            ***REMOVED***
            _this.emit(_this.onNotice, target, nick, message, msg);
        ***REMOVED***);
        _this.onRegister(function () ***REMOVED*** return _this._startPingCheckTimer(); ***REMOVED***);
        _this._credentials = __assign(***REMOVED******REMOVED***, credentials);
        if (channelTypes) ***REMOVED***
            _this._serverProperties.channelTypes = channelTypes;
        ***REMOVED***
        return _this;
    ***REMOVED***
    IrcClient.prototype.receiveLine = function (line) ***REMOVED***
        var _a;
        this._logger.debug("Received message: " + line);
        // eslint-disable-next-line @typescript-eslint/init-declarations
        var parsedMessage;
        try ***REMOVED***
            parsedMessage = parseMessage(line, this._serverProperties, this._registeredMessageTypes, true, this._options.nonConformingCommands);
        ***REMOVED***
        catch (e) ***REMOVED***
            this._logger.error("Error parsing message: " + e.message);
            this._logger.trace((_a = e.stack) !== null && _a !== void 0 ? _a : 'No stack available');
            return;
        ***REMOVED***
        this._logger.trace("Parsed message: " + JSON.stringify(parsedMessage));
        this._startPingCheckTimer();
        this.emit(this.onAnyMessage, parsedMessage);
        this._handleEvents(parsedMessage);
    ***REMOVED***;
    Object.defineProperty(IrcClient.prototype, "serverProperties", ***REMOVED***
        get: function () ***REMOVED***
            return klona(this._serverProperties);
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(IrcClient.prototype, "port", ***REMOVED***
        get: function () ***REMOVED***
            var _a = this._options, webSocket = _a.webSocket, _b = _a.connection, port = _b.port, secure = _b.secure;
            if (port) ***REMOVED***
                return port;
            ***REMOVED***
            if (webSocket) ***REMOVED***
                return secure ? 443 : 80;
            ***REMOVED***
            return secure ? 6697 : 6667;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    IrcClient.prototype.pingCheck = function () ***REMOVED***
        var _this = this;
        var now = Date.now();
        var nowStr = now.toString();
        var handler = this.onTypedMessage(Pong, function (msg) ***REMOVED***
            var message = msg.params.message;
            if (message === nowStr) ***REMOVED***
                _this._logger.debug("Current ping: " + (Date.now() - now) + "ms");
                if (_this._pingTimeoutTimer) ***REMOVED***
                    clearTimeout(_this._pingTimeoutTimer);
                ***REMOVED***
                _this.removeMessageListener(handler);
            ***REMOVED***
        ***REMOVED***);
        this._pingTimeoutTimer = setTimeout(function () ***REMOVED*** return __awaiter(_this, void 0, void 0, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                this.removeMessageListener(handler);
                // eslint-disable-next-line no-restricted-syntax
                if (this._options.connection.reconnect === false) ***REMOVED***
                    this._logger.error("Disconnecting because the last ping took over " + this._pingTimeout + " seconds");
                ***REMOVED***
                else ***REMOVED***
                    this._logger.warn("Reconnecting because the last ping took over " + this._pingTimeout + " seconds");
                ***REMOVED***
                this._connection.assumeExternalDisconnect();
                return [2 /*return*/];
            ***REMOVED***);
        ***REMOVED***); ***REMOVED***, this._pingTimeout * 1000);
        this.sendMessage(Ping, ***REMOVED*** message: nowStr ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.reconnect = function (message) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.quit(message)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, this.connect()];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.registerMessageType = function (cls) ***REMOVED***
        if (cls.COMMAND !== '') ***REMOVED***
            this._logger.trace("Registering message type " + cls.COMMAND);
            this._registeredMessageTypes.set(cls.COMMAND.toUpperCase(), cls);
        ***REMOVED***
    ***REMOVED***;
    IrcClient.prototype.knowsCommand = function (command) ***REMOVED***
        return this._registeredMessageTypes.has(command.toUpperCase());
    ***REMOVED***;
    IrcClient.prototype.getCommandClass = function (command) ***REMOVED***
        return this._registeredMessageTypes.get(command.toUpperCase());
    ***REMOVED***;
    IrcClient.prototype.connect = function () ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        this._supportsCapabilities = false;
                        this._negotiatedCapabilities = new Map();
                        this._currentNick = this._credentials.nick;
                        return [4 /*yield*/, this._setupConnection()];
                    case 1:
                        _a.sent();
                        this._logger.info("Connecting to " + this._connection.host + ":" + this._connection.port);
                        return [4 /*yield*/, this._connection.connect()];
                    case 2:
                        _a.sent();
                        this.emit(this.onConnect);
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.waitForRegistration = function () ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                if (this._registered) ***REMOVED***
                    return [2 /*return*/, undefined];
                ***REMOVED***
                return [2 /*return*/, new Promise(function (resolve, reject) ***REMOVED***
                        // eslint-disable-next-line @typescript-eslint/init-declarations
                        var errorListener;
                        // eslint-disable-next-line @typescript-eslint/init-declarations
                        var disconnectListener;
                        var registerListener = _this.onRegister(function () ***REMOVED***
                            registerListener.unbind();
                            _this.removeMessageListener(errorListener);
                            disconnectListener.unbind();
                            resolve();
                        ***REMOVED***);
                        errorListener = _this.onTypedMessage(ErrorMessage, function (msg) ***REMOVED***
                            registerListener.unbind();
                            _this.removeMessageListener(errorListener);
                            disconnectListener.unbind();
                            reject(new MessageError(msg));
                        ***REMOVED***);
                        disconnectListener = _this.onDisconnect(function (reason) ***REMOVED***
                            registerListener.unbind();
                            _this.removeMessageListener(errorListener);
                            disconnectListener.unbind();
                            reject(reason);
                        ***REMOVED***);
                    ***REMOVED***)];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.addCapability = function (cap) ***REMOVED***
        var e_5, _a;
        this._clientCapabilities.set(cap.name, cap);
        if (cap.messageTypes) ***REMOVED***
            try ***REMOVED***
                for (var _b = __values(cap.messageTypes), _c = _b.next(); !_c.done; _c = _b.next()) ***REMOVED***
                    var messageType = _c.value;
                    this.registerMessageType(messageType);
                ***REMOVED***
            ***REMOVED***
            catch (e_5_1) ***REMOVED*** e_5 = ***REMOVED*** error: e_5_1 ***REMOVED***; ***REMOVED***
            finally ***REMOVED***
                try ***REMOVED***
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                ***REMOVED***
                finally ***REMOVED*** if (e_5) throw e_5.error; ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    IrcClient.prototype.registerCapability = function (cap) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                this.addCapability(cap);
                if (this._serverCapabilities.has(cap.name)) ***REMOVED***
                    return [2 /*return*/, this._negotiateCapabilities([cap])];
                ***REMOVED***
                return [2 /*return*/, []];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.send = function (message) ***REMOVED***
        this.sendRaw(message.toString());
    ***REMOVED***;
    IrcClient.prototype.sendRaw = function (line) ***REMOVED***
        if (this._connection.isConnected) ***REMOVED***
            this._logger.debug("Sending message: " + line);
            this._connection.sendLine(line);
        ***REMOVED***
    ***REMOVED***;
    IrcClient.prototype.onNamedMessage = function (commandName, handler, handlerName) ***REMOVED***
        if (!this._events.has(commandName)) ***REMOVED***
            this._events.set(commandName, new Map());
        ***REMOVED***
        var handlerList = this._events.get(commandName);
        if (!handlerName) ***REMOVED***
            do ***REMOVED***
                handlerName = commandName + ":" + padLeft(Math.random() * 10000, 4, '0');
            ***REMOVED*** while (handlerList.has(handlerName));
        ***REMOVED***
        handlerList.set(handlerName, handler);
        return handlerName;
    ***REMOVED***;
    IrcClient.prototype.onTypedMessage = function (type, handler, handlerName) ***REMOVED***
        return this.onNamedMessage(type.COMMAND, handler, handlerName);
    ***REMOVED***;
    IrcClient.prototype.removeMessageListener = function (handlerName) ***REMOVED***
        var _a = __read(handlerName.split(':'), 1), commandName = _a[0];
        if (!this._events.has(commandName)) ***REMOVED***
            return;
        ***REMOVED***
        this._events.get(commandName).delete(handlerName);
    ***REMOVED***;
    IrcClient.prototype.createMessage = function (type, params, tags) ***REMOVED***
        var tagsMap = tags ? new Map(Object.entries(tags)) : undefined;
        return createMessage(type, params, undefined, tagsMap, this.serverProperties);
    ***REMOVED***;
    IrcClient.prototype.sendMessage = function (type, params, tags) ***REMOVED***
        this.send(this.createMessage(type, params, tags));
    ***REMOVED***;
    IrcClient.prototype.sendMessageAndCaptureReply = function (type, params) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var message, promise;
            return __generator(this, function (_a) ***REMOVED***
                if (!type.SUPPORTS_CAPTURE) ***REMOVED***
                    throw new Error("The command \"" + type.COMMAND + "\" does not support reply capture");
                ***REMOVED***
                message = this.createMessage(type, params);
                promise = this.collect(message).promise();
                this.send(message);
                return [2 /*return*/, promise];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    Object.defineProperty(IrcClient.prototype, "isConnected", ***REMOVED***
        get: function () ***REMOVED***
            return this._connection.isConnected;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(IrcClient.prototype, "isConnecting", ***REMOVED***
        get: function () ***REMOVED***
            return this._connection.isConnecting;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(IrcClient.prototype, "isRegistered", ***REMOVED***
        get: function () ***REMOVED***
            return this._registered;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    Object.defineProperty(IrcClient.prototype, "currentNick", ***REMOVED***
        get: function () ***REMOVED***
            return this._currentNick;
    ***REMOVED***
        enumerable: false,
        configurable: true
    ***REMOVED***);
    /** @private */
    IrcClient.prototype.collect = function (originalMessage) ***REMOVED***
        var types = [];
        for (var _i = 1; _i < arguments.length; _i++) ***REMOVED***
            types[_i - 1] = arguments[_i];
        ***REMOVED***
        var collector = new (MessageCollector.bind.apply(MessageCollector, __spread([void 0, this, originalMessage], types)))();
        this._collectors.push(collector);
        return collector;
    ***REMOVED***;
    /** @private */
    IrcClient.prototype.stopCollect = function (collector) ***REMOVED***
        this._collectors.splice(this._collectors.findIndex(function (value) ***REMOVED*** return value === collector; ***REMOVED***), 1);
    ***REMOVED***;
    // convenience methods
    IrcClient.prototype.join = function (channel, key) ***REMOVED***
        this.sendMessage(ChannelJoin, ***REMOVED*** channel: channel, key: key ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.part = function (channel) ***REMOVED***
        this.sendMessage(ChannelPart, ***REMOVED*** channel: channel ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.quit = function (message) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                this.sendMessage(ClientQuit, ***REMOVED*** message: message ***REMOVED***);
                void this._connection.disconnect().then(function () ***REMOVED***
                    _this._logger.debug('Finished cleaning up old connection');
                ***REMOVED***);
                return [2 /*return*/];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.say = function (target, message, tags) ***REMOVED***
        if (tags === void 0) ***REMOVED*** tags = ***REMOVED******REMOVED***; ***REMOVED***
        this.sendMessage(PrivateMessage, ***REMOVED*** target: target, message: message ***REMOVED***, tags);
    ***REMOVED***;
    IrcClient.prototype.sendCtcp = function (target, type, message) ***REMOVED***
        this.say(target, "\u0001" + type.toUpperCase() + " " + message + "\u0001");
    ***REMOVED***;
    IrcClient.prototype.action = function (target, message) ***REMOVED***
        this.sendCtcp(target, 'ACTION', message);
    ***REMOVED***;
    IrcClient.prototype.getPassword = function (currentPassword) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            return __generator(this, function (_a) ***REMOVED***
                return [2 /*return*/, currentPassword];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype.registerCoreMessageTypes = function () ***REMOVED***
        var _this = this;
        forEachObjectEntry(MessageTypes.Commands, function (type) ***REMOVED***
            _this.registerMessageType(type);
        ***REMOVED***);
        forEachObjectEntry(MessageTypes.Numerics, function (type) ***REMOVED***
            _this.registerMessageType(type);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype._negotiateCapabilityBatch = function (capabilities) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                return [2 /*return*/, Promise.all(capabilities
                        .filter(function (list) ***REMOVED*** return list.length; ***REMOVED***)
                        .map(function (capList) ***REMOVED*** return __awaiter(_this, void 0, void 0, function () ***REMOVED*** return __generator(this, function (_a) ***REMOVED***
                        return [2 /*return*/, this._negotiateCapabilities(capList)];
                    ***REMOVED***); ***REMOVED***); ***REMOVED***))];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype._negotiateCapabilities = function (capList) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var mappedCapList, messages, capReply, negotiatedCapNames, newNegotiatedCaps, newNegotiatedCaps_1, newNegotiatedCaps_1_1, newCap, mergedCap;
            var e_6, _a;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        mappedCapList = arrayToObject(capList, function (cap) ***REMOVED***
                            var _a;
                            return (_a = ***REMOVED******REMOVED***,
                                _a[cap.name] = cap,
                                _a);
                        ***REMOVED***);
                        return [4 /*yield*/, this.sendMessageAndCaptureReply(CapabilityNegotiation, ***REMOVED***
                                subCommand: 'REQ',
                                capabilities: capList.map(function (cap) ***REMOVED*** return cap.name; ***REMOVED***).join(' ')
                            ***REMOVED***)];
                    case 1:
                        messages = _b.sent();
                        capReply = messages.shift();
                        if (!capReply) ***REMOVED***
                            throw new Error('capability negotiation failed unexpectedly without any reply');
                        ***REMOVED***
                        if (!(capReply instanceof CapabilityNegotiation)) ***REMOVED***
                            throw new Error("capability negotiation failed unexpectedly with \"" + capReply.command + "\" command");
                        ***REMOVED***
                        negotiatedCapNames = capReply.params.capabilities.split(' ').filter(function (c) ***REMOVED*** return c; ***REMOVED***);
                        if (capReply.params.subCommand === 'ACK') ***REMOVED***
                            // filter is necessary because some networks seem to add trailing spaces...
                            this._logger.debug("Successfully negotiated capabilities: " + negotiatedCapNames.join(', '));
                            newNegotiatedCaps = negotiatedCapNames.map(function (capName) ***REMOVED*** return mappedCapList[capName]; ***REMOVED***);
                            try ***REMOVED***
                                for (newNegotiatedCaps_1 = __values(newNegotiatedCaps), newNegotiatedCaps_1_1 = newNegotiatedCaps_1.next(); !newNegotiatedCaps_1_1.done; newNegotiatedCaps_1_1 = newNegotiatedCaps_1.next()) ***REMOVED***
                                    newCap = newNegotiatedCaps_1_1.value;
                                    mergedCap = this._clientCapabilities.get(newCap.name);
                                    mergedCap.param = newCap.param;
                                    this._negotiatedCapabilities.set(mergedCap.name, mergedCap);
                                ***REMOVED***
                            ***REMOVED***
                            catch (e_6_1) ***REMOVED*** e_6 = ***REMOVED*** error: e_6_1 ***REMOVED***; ***REMOVED***
                            finally ***REMOVED***
                                try ***REMOVED***
                                    if (newNegotiatedCaps_1_1 && !newNegotiatedCaps_1_1.done && (_a = newNegotiatedCaps_1.return)) _a.call(newNegotiatedCaps_1);
                                ***REMOVED***
                                finally ***REMOVED*** if (e_6) throw e_6.error; ***REMOVED***
                            ***REMOVED***
                            return [2 /*return*/, newNegotiatedCaps];
                        ***REMOVED***
                        else ***REMOVED***
                            this._logger.warn("Failed to negotiate capabilities: " + negotiatedCapNames.join(', '));
                            return [2 /*return*/, new Error('capabilities failed to negotiate')];
                        ***REMOVED***
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype._updateCredentials = function (newCredentials) ***REMOVED***
        this._credentials = __assign(__assign(***REMOVED******REMOVED***, this._credentials), newCredentials);
    ***REMOVED***;
    IrcClient.prototype._setupConnection = function () ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                if (this._initialConnectionSetupDone) ***REMOVED***
                    return [2 /*return*/];
                ***REMOVED***
                this._connection.onConnect(function () ***REMOVED*** return __awaiter(_this, void 0, void 0, function () ***REMOVED***
                    var _a, password;
                    var _this = this;
                    var _b, _c;
                    return __generator(this, function (_d) ***REMOVED***
                        switch (_d.label) ***REMOVED***
                            case 0:
                                this._logger.info("Connection to server " + this._connection.host + ":" + this._connection.port + " established");
                                this._logger.debug('Determining connection password');
                                return [4 /*yield*/, Promise.all([
                                        this.getPassword(this._credentials.password),
                                        this.sendMessageAndCaptureReply(CapabilityNegotiation, ***REMOVED***
                                            subCommand: 'LS',
                                            version: '302'
                                        ***REMOVED***)
                                            .then(function (capReply) ***REMOVED***
                                            if (!capReply.length || !(capReply[0] instanceof CapabilityNegotiation)) ***REMOVED***
                                                _this._logger.debug('Server does not support capabilities');
                                                return [];
                                            ***REMOVED***
                                            _this._supportsCapabilities = true;
                                            var capLists = capReply.map(function (line) ***REMOVED***
                                                return arrayToObject(line.params.capabilities.split(' '), function (part) ***REMOVED***
                                                    var _a;
                                                    if (!part) ***REMOVED***
                                                        return ***REMOVED******REMOVED***;
                                                    ***REMOVED***
                                                    var _b = __read(splitWithLimit(part, '=', 2), 2), cap = _b[0], param = _b[1];
                                                    return _a = ***REMOVED******REMOVED***,
                                                        _a[cap] = ***REMOVED***
                                                            name: cap,
                                                            param: param || true
                                                    ***REMOVED***
                                                        _a;
                                                ***REMOVED***);
                                            ***REMOVED***);
                                            _this._serverCapabilities = new Map(Object.entries(Object.assign.apply(Object, __spread([***REMOVED******REMOVED***], capLists))));
                                            _this._logger.debug("Capabilities supported by server: " + Array.from(_this._serverCapabilities.keys()).join(', '));
                                            var capabilitiesToNegotiate = capLists.map(function (list) ***REMOVED***
                                                var capNames = Object.keys(list);
                                                return Array.from(_this._clientCapabilities.entries())
                                                    .filter(function (_a) ***REMOVED***
                                                    var _b = __read(_a, 1), name = _b[0];
                                                    return capNames.includes(name);
                                                ***REMOVED***)
                                                    .map(function (_a) ***REMOVED***
                                                    var _b = __read(_a, 2), cap = _b[1];
                                                    return cap;
                                                ***REMOVED***);
                                            ***REMOVED***);
                                            return _this._negotiateCapabilityBatch(capabilitiesToNegotiate);
                                        ***REMOVED***)
                                            .then(function () ***REMOVED***
                                            _this.sendMessage(CapabilityNegotiation, ***REMOVED*** subCommand: 'END' ***REMOVED***);
                                        ***REMOVED***)
                                    ])];
                            case 1:
                                _a = __read.apply(void 0, [_d.sent(), 1]), password = _a[0];
                                if (password && password !== this._credentials.password) ***REMOVED***
                                    this._updateCredentials(***REMOVED*** password: password ***REMOVED***);
                                ***REMOVED***
                                if (password) ***REMOVED***
                                    this.sendMessage(Password, ***REMOVED*** password: password ***REMOVED***);
                                ***REMOVED***
                                this.sendMessage(NickChange, ***REMOVED*** nick: this._credentials.nick ***REMOVED***);
                                this.sendMessage(UserRegistration, ***REMOVED***
                                    user: (_b = this._credentials.userName) !== null && _b !== void 0 ? _b : this._credentials.nick,
                                    mode: '8',
                                    unused: '*',
                                    realName: (_c = this._credentials.realName) !== null && _c !== void 0 ? _c : this._credentials.nick
                                ***REMOVED***);
                                return [2 /*return*/];
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***); ***REMOVED***);
                this._initialConnectionSetupDone = true;
                this._connection.onReceive(function (line) ***REMOVED***
                    _this.receiveLine(line);
                ***REMOVED***);
                this._connection.onDisconnect(function (manually, reason) ***REMOVED***
                    _this._registered = false;
                    if (_this._pingCheckTimer) ***REMOVED***
                        clearTimeout(_this._pingCheckTimer);
                    ***REMOVED***
                    if (_this._pingTimeoutTimer) ***REMOVED***
                        clearTimeout(_this._pingTimeoutTimer);
                    ***REMOVED***
                    if (manually) ***REMOVED***
                        _this._logger.info('Disconnected');
                    ***REMOVED***
                    else ***REMOVED***
                        if (reason) ***REMOVED***
                            _this._logger.error("Disconnected unexpectedly: " + reason.message);
                        ***REMOVED***
                        else ***REMOVED***
                            _this._logger.error('Disconnected unexpectedly');
                        ***REMOVED***
                    ***REMOVED***
                    _this.emit(_this.onDisconnect, manually, reason);
                ***REMOVED***);
                // eslint-disable-next-line no-restricted-syntax
                if (this._options.connection.reconnect !== false) ***REMOVED***
                    this._connection.onEnd(function (manually) ***REMOVED***
                        if (!manually) ***REMOVED***
                            _this._logger.info('No further retries will be made');
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
                return [2 /*return*/];
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    IrcClient.prototype._handleReceivedClientNick = function (me) ***REMOVED***
        if (this._currentNick !== me) ***REMOVED***
            if (this._currentNick !== '') ***REMOVED***
                this._logger.warn("Mismatching nicks: passed " + this._currentNick + ", but you're actually " + me);
            ***REMOVED***
            this._currentNick = me;
        ***REMOVED***
    ***REMOVED***;
    // event helper
    IrcClient.prototype._handleEvents = function (message) ***REMOVED***
        var e_7, _a;
        this._collectors.some(function (collector) ***REMOVED*** return collector.collect(message); ***REMOVED***);
        var handlers = this._events.get(message.constructor.COMMAND);
        if (!handlers) ***REMOVED***
            return;
        ***REMOVED***
        try ***REMOVED***
            for (var _b = __values(handlers.values()), _c = _b.next(); !_c.done; _c = _b.next()) ***REMOVED***
                var handler = _c.value;
                handler(message);
            ***REMOVED***
        ***REMOVED***
        catch (e_7_1) ***REMOVED*** e_7 = ***REMOVED*** error: e_7_1 ***REMOVED***; ***REMOVED***
        finally ***REMOVED***
            try ***REMOVED***
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            ***REMOVED***
            finally ***REMOVED*** if (e_7) throw e_7.error; ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    IrcClient.prototype._startPingCheckTimer = function () ***REMOVED***
        var _this = this;
        if (this._pingCheckTimer) ***REMOVED***
            clearTimeout(this._pingCheckTimer);
        ***REMOVED***
        if (this._connection.isConnected) ***REMOVED***
            this._pingCheckTimer = setTimeout(function () ***REMOVED*** return _this.pingCheck(); ***REMOVED***, this._pingOnInactivity * 1000);
        ***REMOVED***
        else ***REMOVED***
            this._pingCheckTimer = undefined;
        ***REMOVED***
    ***REMOVED***;
    __decorate([
        Enumerable(false)
    ], IrcClient.prototype, "_options", void 0);
    __decorate([
        Enumerable(false)
    ], IrcClient.prototype, "_credentials", void 0);
    return IrcClient;
***REMOVED***(EventEmitter));
export ***REMOVED*** IrcClient ***REMOVED***;
