// whatwg-url/lib/URL.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");
const Impl = require(".//URL-impl.js");

const impl = utils.implSymbol;

function URL(url) ***REMOVED***
  if (!this || this[impl] || !(this instanceof URL)) ***REMOVED***
    throw new TypeError("Failed to construct 'URL': Please use the 'new' operator, this DOM object constructor cannot be called as a function.");
  ***REMOVED***
  if (arguments.length < 1) ***REMOVED***
    throw new TypeError("Failed to construct 'URL': 1 argument required, but only " + arguments.length + " present.");
  ***REMOVED***
  const args = [];
  for (let i = 0; i < arguments.length && i < 2; ++i) ***REMOVED***
    args[i] = arguments[i];
  ***REMOVED***
  args[0] = conversions["USVString"](args[0]);
  if (args[1] !== undefined) ***REMOVED***
  args[1] = conversions["USVString"](args[1]);
  ***REMOVED***

  module.exports.setup(this, args);
***REMOVED***

URL.prototype.toJSON = function toJSON() ***REMOVED***
  if (!this || !module.exports.is(this)) ***REMOVED***
    throw new TypeError("Illegal invocation");
  ***REMOVED***
  const args = [];
  for (let i = 0; i < arguments.length && i < 0; ++i) ***REMOVED***
    args[i] = arguments[i];
  ***REMOVED***
  return this[impl].toJSON.apply(this[impl], args);
***REMOVED***;
Object.defineProperty(URL.prototype, "href", ***REMOVED***
  get() ***REMOVED***
    return this[impl].href;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].href = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

URL.prototype.toString = function () ***REMOVED***
  if (!this || !module.exports.is(this)) ***REMOVED***
    throw new TypeError("Illegal invocation");
  ***REMOVED***
  return this.href;
***REMOVED***;

Object.defineProperty(URL.prototype, "origin", ***REMOVED***
  get() ***REMOVED***
    return this[impl].origin;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "protocol", ***REMOVED***
  get() ***REMOVED***
    return this[impl].protocol;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].protocol = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "username", ***REMOVED***
  get() ***REMOVED***
    return this[impl].username;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].username = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "password", ***REMOVED***
  get() ***REMOVED***
    return this[impl].password;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].password = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "host", ***REMOVED***
  get() ***REMOVED***
    return this[impl].host;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].host = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "hostname", ***REMOVED***
  get() ***REMOVED***
    return this[impl].hostname;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].hostname = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "port", ***REMOVED***
  get() ***REMOVED***
    return this[impl].port;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].port = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "pathname", ***REMOVED***
  get() ***REMOVED***
    return this[impl].pathname;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].pathname = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "search", ***REMOVED***
  get() ***REMOVED***
    return this[impl].search;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].search = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);

Object.defineProperty(URL.prototype, "hash", ***REMOVED***
  get() ***REMOVED***
    return this[impl].hash;
  ***REMOVED***,
  set(V) ***REMOVED***
    V = conversions["USVString"](V);
    this[impl].hash = V;
  ***REMOVED***,
  enumerable: true,
  configurable: true
***REMOVED***);


module.exports = ***REMOVED***
  is(obj) ***REMOVED***
    return !!obj && obj[impl] instanceof Impl.implementation;
  ***REMOVED***,
  create(constructorArgs, privateData) ***REMOVED***
    let obj = Object.create(URL.prototype);
    this.setup(obj, constructorArgs, privateData);
    return obj;
  ***REMOVED***,
  setup(obj, constructorArgs, privateData) ***REMOVED***
    if (!privateData) privateData = ***REMOVED******REMOVED***;
    privateData.wrapper = obj;

    obj[impl] = new Impl.implementation(constructorArgs, privateData);
    obj[impl][utils.wrapperSymbol] = obj;
  ***REMOVED***,
  interface: URL,
  expose: ***REMOVED***
    Window: ***REMOVED*** URL: URL ***REMOVED***,
    Worker: ***REMOVED*** URL: URL ***REMOVED***
  ***REMOVED***
***REMOVED***;
