// qs/lib/stringify.js
'use strict';

var utils = require('./utils');
var formats = require('./formats');
var has = Object.prototype.hasOwnProperty;

var arrayPrefixGenerators = ***REMOVED***
    brackets: function brackets(prefix) ***REMOVED*** // eslint-disable-line func-name-matching
        return prefix + '[]';
***REMOVED***
    comma: 'comma',
    indices: function indices(prefix, key) ***REMOVED*** // eslint-disable-line func-name-matching
        return prefix + '[' + key + ']';
***REMOVED***
    repeat: function repeat(prefix) ***REMOVED*** // eslint-disable-line func-name-matching
        return prefix;
    ***REMOVED***
***REMOVED***;

var isArray = Array.isArray;
var push = Array.prototype.push;
var pushToArray = function (arr, valueOrArray) ***REMOVED***
    push.apply(arr, isArray(valueOrArray) ? valueOrArray : [valueOrArray]);
***REMOVED***;

var toISO = Date.prototype.toISOString;

var defaults = ***REMOVED***
    addQueryPrefix: false,
    allowDots: false,
    charset: 'utf-8',
    charsetSentinel: false,
    delimiter: '&',
    encode: true,
    encoder: utils.encode,
    encodeValuesOnly: false,
    formatter: formats.formatters[formats['default']],
    // deprecated
    indices: false,
    serializeDate: function serializeDate(date) ***REMOVED*** // eslint-disable-line func-name-matching
        return toISO.call(date);
***REMOVED***
    skipNulls: false,
    strictNullHandling: false
***REMOVED***;

var stringify = function stringify( // eslint-disable-line func-name-matching
    object,
    prefix,
    generateArrayPrefix,
    strictNullHandling,
    skipNulls,
    encoder,
    filter,
    sort,
    allowDots,
    serializeDate,
    formatter,
    encodeValuesOnly,
    charset
) ***REMOVED***
    var obj = object;
    if (typeof filter === 'function') ***REMOVED***
        obj = filter(prefix, obj);
    ***REMOVED*** else if (obj instanceof Date) ***REMOVED***
        obj = serializeDate(obj);
    ***REMOVED*** else if (generateArrayPrefix === 'comma' && isArray(obj)) ***REMOVED***
        obj = obj.join(',');
    ***REMOVED***

    if (obj === null) ***REMOVED***
        if (strictNullHandling) ***REMOVED***
            return encoder && !encodeValuesOnly ? encoder(prefix, defaults.encoder, charset) : prefix;
        ***REMOVED***

        obj = '';
    ***REMOVED***

    if (typeof obj === 'string' || typeof obj === 'number' || typeof obj === 'boolean' || utils.isBuffer(obj)) ***REMOVED***
        if (encoder) ***REMOVED***
            var keyValue = encodeValuesOnly ? prefix : encoder(prefix, defaults.encoder, charset);
            return [formatter(keyValue) + '=' + formatter(encoder(obj, defaults.encoder, charset))];
        ***REMOVED***
        return [formatter(prefix) + '=' + formatter(String(obj))];
    ***REMOVED***

    var values = [];

    if (typeof obj === 'undefined') ***REMOVED***
        return values;
    ***REMOVED***

    var objKeys;
    if (isArray(filter)) ***REMOVED***
        objKeys = filter;
    ***REMOVED*** else ***REMOVED***
        var keys = Object.keys(obj);
        objKeys = sort ? keys.sort(sort) : keys;
    ***REMOVED***

    for (var i = 0; i < objKeys.length; ++i) ***REMOVED***
        var key = objKeys[i];

        if (skipNulls && obj[key] === null) ***REMOVED***
            continue;
        ***REMOVED***

        if (isArray(obj)) ***REMOVED***
            pushToArray(values, stringify(
                obj[key],
                typeof generateArrayPrefix === 'function' ? generateArrayPrefix(prefix, key) : prefix,
                generateArrayPrefix,
                strictNullHandling,
                skipNulls,
                encoder,
                filter,
                sort,
                allowDots,
                serializeDate,
                formatter,
                encodeValuesOnly,
                charset
            ));
        ***REMOVED*** else ***REMOVED***
            pushToArray(values, stringify(
                obj[key],
                prefix + (allowDots ? '.' + key : '[' + key + ']'),
                generateArrayPrefix,
                strictNullHandling,
                skipNulls,
                encoder,
                filter,
                sort,
                allowDots,
                serializeDate,
                formatter,
                encodeValuesOnly,
                charset
            ));
        ***REMOVED***
    ***REMOVED***

    return values;
***REMOVED***;

var normalizeStringifyOptions = function normalizeStringifyOptions(opts) ***REMOVED***
    if (!opts) ***REMOVED***
        return defaults;
    ***REMOVED***

    if (opts.encoder !== null && opts.encoder !== undefined && typeof opts.encoder !== 'function') ***REMOVED***
        throw new TypeError('Encoder has to be a function.');
    ***REMOVED***

    var charset = opts.charset || defaults.charset;
    if (typeof opts.charset !== 'undefined' && opts.charset !== 'utf-8' && opts.charset !== 'iso-8859-1') ***REMOVED***
        throw new TypeError('The charset option must be either utf-8, iso-8859-1, or undefined');
    ***REMOVED***

    var format = formats['default'];
    if (typeof opts.format !== 'undefined') ***REMOVED***
        if (!has.call(formats.formatters, opts.format)) ***REMOVED***
            throw new TypeError('Unknown format option provided.');
        ***REMOVED***
        format = opts.format;
    ***REMOVED***
    var formatter = formats.formatters[format];

    var filter = defaults.filter;
    if (typeof opts.filter === 'function' || isArray(opts.filter)) ***REMOVED***
        filter = opts.filter;
    ***REMOVED***

    return ***REMOVED***
        addQueryPrefix: typeof opts.addQueryPrefix === 'boolean' ? opts.addQueryPrefix : defaults.addQueryPrefix,
        allowDots: typeof opts.allowDots === 'undefined' ? defaults.allowDots : !!opts.allowDots,
        charset: charset,
        charsetSentinel: typeof opts.charsetSentinel === 'boolean' ? opts.charsetSentinel : defaults.charsetSentinel,
        delimiter: typeof opts.delimiter === 'undefined' ? defaults.delimiter : opts.delimiter,
        encode: typeof opts.encode === 'boolean' ? opts.encode : defaults.encode,
        encoder: typeof opts.encoder === 'function' ? opts.encoder : defaults.encoder,
        encodeValuesOnly: typeof opts.encodeValuesOnly === 'boolean' ? opts.encodeValuesOnly : defaults.encodeValuesOnly,
        filter: filter,
        formatter: formatter,
        serializeDate: typeof opts.serializeDate === 'function' ? opts.serializeDate : defaults.serializeDate,
        skipNulls: typeof opts.skipNulls === 'boolean' ? opts.skipNulls : defaults.skipNulls,
        sort: typeof opts.sort === 'function' ? opts.sort : null,
        strictNullHandling: typeof opts.strictNullHandling === 'boolean' ? opts.strictNullHandling : defaults.strictNullHandling
    ***REMOVED***;
***REMOVED***;

module.exports = function (object, opts) ***REMOVED***
    var obj = object;
    var options = normalizeStringifyOptions(opts);

    var objKeys;
    var filter;

    if (typeof options.filter === 'function') ***REMOVED***
        filter = options.filter;
        obj = filter('', obj);
    ***REMOVED*** else if (isArray(options.filter)) ***REMOVED***
        filter = options.filter;
        objKeys = filter;
    ***REMOVED***

    var keys = [];

    if (typeof obj !== 'object' || obj === null) ***REMOVED***
        return '';
    ***REMOVED***

    var arrayFormat;
    if (opts && opts.arrayFormat in arrayPrefixGenerators) ***REMOVED***
        arrayFormat = opts.arrayFormat;
    ***REMOVED*** else if (opts && 'indices' in opts) ***REMOVED***
        arrayFormat = opts.indices ? 'indices' : 'repeat';
    ***REMOVED*** else ***REMOVED***
        arrayFormat = 'indices';
    ***REMOVED***

    var generateArrayPrefix = arrayPrefixGenerators[arrayFormat];

    if (!objKeys) ***REMOVED***
        objKeys = Object.keys(obj);
    ***REMOVED***

    if (options.sort) ***REMOVED***
        objKeys.sort(options.sort);
    ***REMOVED***

    for (var i = 0; i < objKeys.length; ++i) ***REMOVED***
        var key = objKeys[i];

        if (options.skipNulls && obj[key] === null) ***REMOVED***
            continue;
        ***REMOVED***
        pushToArray(keys, stringify(
            obj[key],
            key,
            generateArrayPrefix,
            options.strictNullHandling,
            options.skipNulls,
            options.encode ? options.encoder : null,
            options.filter,
            options.sort,
            options.allowDots,
            options.serializeDate,
            options.formatter,
            options.encodeValuesOnly,
            options.charset
        ));
    ***REMOVED***

    var joined = keys.join(options.delimiter);
    var prefix = options.addQueryPrefix === true ? '?' : '';

    if (options.charsetSentinel) ***REMOVED***
        if (options.charset === 'iso-8859-1') ***REMOVED***
            // encodeURIComponent('&#10003;'), the "numeric entity" representation of a checkmark
            prefix += 'utf8=%26%2310003%3B&';
        ***REMOVED*** else ***REMOVED***
            // encodeURIComponent('✓')
            prefix += 'utf8=%E2%9C%93&';
        ***REMOVED***
    ***REMOVED***

    return joined.length > 0 ? prefix + joined : '';
***REMOVED***;
