// twitch/es/API/Helix/Poll/HelixPollApi.mjs
import ***REMOVED*** __assign, __awaiter, __decorate, __extends, __generator ***REMOVED*** from "tslib";
import ***REMOVED*** TwitchApiCallType ***REMOVED*** from 'twitch-api-call';
import ***REMOVED*** extractUserId, rtfm ***REMOVED*** from 'twitch-common';
import ***REMOVED*** BaseApi ***REMOVED*** from "../../BaseApi.mjs";
import ***REMOVED*** HelixPaginatedRequest ***REMOVED*** from "../HelixPaginatedRequest.mjs";
import ***REMOVED*** createPaginatedResult ***REMOVED*** from "../HelixPaginatedResult.mjs";
import ***REMOVED*** makePaginationQuery ***REMOVED*** from "../HelixPagination.mjs";
import ***REMOVED*** HelixPoll ***REMOVED*** from "./HelixPoll.mjs";
/**
 * The Helix API methods that deal with polls.
 *
 * Can be accessed using `client.helix.polls` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const ***REMOVED*** data: polls ***REMOVED*** = await api.helix.polls.getPolls('61369223');
 * ```
 */
var HelixPollApi = /** @class */ (function (_super) ***REMOVED***
    __extends(HelixPollApi, _super);
    function HelixPollApi() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    /**
     * Retrieves a list of polls for the given broadcaster.
     *
     * @param broadcaster The broadcaster to retrieve polls for.
     * @param pagination
     *
     * @expandParams
     */
    HelixPollApi.prototype.getPolls = function (broadcaster, pagination) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'polls',
                            scope: 'channel:read:polls',
                            query: __assign(***REMOVED*** broadcaster_id: extractUserId(broadcaster) ***REMOVED***, makePaginationQuery(pagination))
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, createPaginatedResult(result, HelixPoll, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a paginator for polls for the given broadcaster.
     *
     * @param broadcaster The broadcaster to retrieve polls for.
     */
    HelixPollApi.prototype.getPollsPaginated = function (broadcaster) ***REMOVED***
        var _this = this;
        return new HelixPaginatedRequest(***REMOVED***
            url: 'polls',
            scope: 'channel:read:polls',
            query: ***REMOVED***
                broadcaster_id: extractUserId(broadcaster)
            ***REMOVED***
    ***REMOVED*** this._client, function (data) ***REMOVED*** return new HelixPoll(data, _this._client); ***REMOVED***, 20);
    ***REMOVED***;
    /**
     * Retrieves polls by IDs.
     *
     * @param broadcaster The broadcaster to retrieve the polls for.
     * @param ids The IDs of the polls.
     */
    HelixPollApi.prototype.getPollsByIds = function (broadcaster, ids) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        if (!ids.length) ***REMOVED***
                            return [2 /*return*/, []];
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                type: TwitchApiCallType.Helix,
                                url: 'polls',
                                scope: 'channel:read:polls',
                                query: ***REMOVED***
                                    broadcaster_id: extractUserId(broadcaster),
                                    id: ids
                                ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result.data.map(function (data) ***REMOVED*** return new HelixPoll(data, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a poll by ID.
     *
     * @param broadcaster The broadcaster to retrieve the poll for.
     * @param id The ID of the poll.
     */
    HelixPollApi.prototype.getPollById = function (broadcaster, id) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var polls;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.getPollsByIds(broadcaster, [id])];
                    case 1:
                        polls = _a.sent();
                        return [2 /*return*/, polls.length ? polls[0] : null];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Creates a new poll.
     *
     * @param broadcaster The broadcaster to create the poll for.
     * @param data
     *
     * @expandParams
     */
    HelixPollApi.prototype.createPoll = function (broadcaster, data) ***REMOVED***
        var _a, _b;
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            return __generator(this, function (_c) ***REMOVED***
                switch (_c.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'polls',
                            method: 'POST',
                            scope: 'channel:manage:polls',
                            jsonBody: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster),
                                title: data.title,
                                choices: data.choices.map(function (title) ***REMOVED*** return (***REMOVED*** title: title ***REMOVED***); ***REMOVED***),
                                duration: data.duration,
                                bits_voting_enabled: data.bitsPerVote != null,
                                bits_per_vote: (_a = data.bitsPerVote) !== null && _a !== void 0 ? _a : 0,
                                channel_points_voting_enabled: data.channelPointsPerVote != null,
                                channel_points_per_vote: (_b = data.channelPointsPerVote) !== null && _b !== void 0 ? _b : 0
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _c.sent();
                        return [2 /*return*/, new HelixPoll(result.data[0], this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Ends a poll.
     *
     * @param broadcaster The broadcaster to end the poll for.
     * @param id The ID of the poll to end.
     * @param showResult Whether to allow the result to be viewed publicly.
     */
    HelixPollApi.prototype.endPoll = function (broadcaster, id, showResult) ***REMOVED***
        if (showResult === void 0) ***REMOVED*** showResult = true; ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var result;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED***
                            type: TwitchApiCallType.Helix,
                            url: 'polls',
                            method: 'PATCH',
                            scope: 'channel:manage:polls',
                            jsonBody: ***REMOVED***
                                broadcaster_id: extractUserId(broadcaster),
                                id: id,
                                status: showResult ? 'TERMINATED' : 'ARCHIVED'
                            ***REMOVED***
                        ***REMOVED***)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, new HelixPoll(result.data[0], this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    HelixPollApi = __decorate([
        rtfm('twitch', 'HelixPollApi')
    ], HelixPollApi);
    return HelixPollApi;
***REMOVED***(BaseApi));
export ***REMOVED*** HelixPollApi ***REMOVED***;
