// @d-fischer/typed-event-emitter/lib/EventEmitter.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.EventEmitter = void 0;
var tslib_1 = require("tslib");
var Listener_1 = require("./Listener");
var EventEmitter = /** @class */ (function () ***REMOVED***
    function EventEmitter() ***REMOVED***
        this._eventListeners = new Map();
        this._internalEventListeners = new Map();
    ***REMOVED***
    EventEmitter.prototype.on = function (event, listener) ***REMOVED***
        return this._addListener(false, event, listener);
    ***REMOVED***;
    EventEmitter.prototype.addListener = function (event, listener) ***REMOVED***
        return this._addListener(false, event, listener);
    ***REMOVED***;
    EventEmitter.prototype.removeListener = function (idOrEvent, listener) ***REMOVED***
        this._removeListener(false, idOrEvent, listener);
    ***REMOVED***;
    EventEmitter.prototype.registerEvent = function () ***REMOVED***
        var _this = this;
        var eventBinder = function (handler) ***REMOVED*** return _this.addListener(eventBinder, handler); ***REMOVED***;
        return eventBinder;
    ***REMOVED***;
    EventEmitter.prototype.emit = function (event) ***REMOVED***
        var e_1, _a, e_2, _b;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) ***REMOVED***
            args[_i - 1] = arguments[_i];
        ***REMOVED***
        if (this._eventListeners.has(event)) ***REMOVED***
            try ***REMOVED***
                for (var _c = tslib_1.__values(this._eventListeners.get(event)), _d = _c.next(); !_d.done; _d = _c.next()) ***REMOVED***
                    var listener = _d.value;
                    listener.apply(void 0, tslib_1.__spread(args));
                ***REMOVED***
            ***REMOVED***
            catch (e_1_1) ***REMOVED*** e_1 = ***REMOVED*** error: e_1_1 ***REMOVED***; ***REMOVED***
            finally ***REMOVED***
                try ***REMOVED***
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                ***REMOVED***
                finally ***REMOVED*** if (e_1) throw e_1.error; ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        if (this._internalEventListeners.has(event)) ***REMOVED***
            try ***REMOVED***
                for (var _e = tslib_1.__values(this._internalEventListeners.get(event)), _f = _e.next(); !_f.done; _f = _e.next()) ***REMOVED***
                    var listener = _f.value;
                    listener.apply(void 0, tslib_1.__spread(args));
                ***REMOVED***
            ***REMOVED***
            catch (e_2_1) ***REMOVED*** e_2 = ***REMOVED*** error: e_2_1 ***REMOVED***; ***REMOVED***
            finally ***REMOVED***
                try ***REMOVED***
                    if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                ***REMOVED***
                finally ***REMOVED*** if (e_2) throw e_2.error; ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    EventEmitter.prototype.registerInternalEvent = function () ***REMOVED***
        var _this = this;
        var eventBinder = function (handler) ***REMOVED*** return _this.addInternalListener(eventBinder, handler); ***REMOVED***;
        return eventBinder;
    ***REMOVED***;
    EventEmitter.prototype.addInternalListener = function (event, listener) ***REMOVED***
        return this._addListener(true, event, listener);
    ***REMOVED***;
    EventEmitter.prototype.removeInternalListener = function (idOrEvent, listener) ***REMOVED***
        this._removeListener(true, idOrEvent, listener);
    ***REMOVED***;
    EventEmitter.prototype._addListener = function (internal, event, listener) ***REMOVED***
        var listenerMap = internal ? this._eventListeners : this._internalEventListeners;
        if (listenerMap.has(event)) ***REMOVED***
            listenerMap.get(event).push(listener);
        ***REMOVED***
        else ***REMOVED***
            listenerMap.set(event, [listener]);
        ***REMOVED***
        return new Listener_1.Listener(this, event, listener, internal);
    ***REMOVED***;
    EventEmitter.prototype._removeListener = function (internal, idOrEvent, listener) ***REMOVED***
        var listenerMap = internal ? this._eventListeners : this._internalEventListeners;
        if (!idOrEvent) ***REMOVED***
            listenerMap.clear();
        ***REMOVED***
        else if (typeof idOrEvent === 'object') ***REMOVED***
            var id = idOrEvent;
            this._removeListener(id._internal, id.event, id.listener);
        ***REMOVED***
        else ***REMOVED***
            var event = idOrEvent;
            if (listenerMap.has(event)) ***REMOVED***
                if (listener) ***REMOVED***
                    var listeners = listenerMap.get(event);
                    var idx = 0;
                    while ((idx = listeners.indexOf(listener)) !== -1) ***REMOVED***
                        listeners.splice(idx, 1);
                    ***REMOVED***
                ***REMOVED***
                else ***REMOVED***
                    listenerMap.delete(event);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    return EventEmitter;
***REMOVED***());
exports.EventEmitter = EventEmitter;
