// twitch-chat-client/lib/index.js
var __tsu = ***REMOVED***
    cache: [],
    defineExport: function (name, def) ***REMOVED***
        Object.defineProperty(exports, name, def);
        this.cache.push(***REMOVED*** name: name, def: def ***REMOVED***);
***REMOVED***
    redefineExports: function () ***REMOVED***
        this.cache.forEach(function (exp) ***REMOVED***
            Object.defineProperty(exports, exp.name, exp.def);
        ***REMOVED***);
    ***REMOVED***
***REMOVED***;
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
exports.parseTwitchMessage = exports.toUserName = exports.toChannelName = exports.LogLevel = exports.ChatUser = exports.GlobalUserState = exports.ClearMsg = exports.Whisper = exports.UserState = exports.UserNotice = exports.RoomState = exports.Reconnect = exports.HostTarget = exports.ClearChat = exports.PrivateMessage = exports.ChatClient = void 0;
/* eslint-disable filenames/match-exported */
var shared_utils_1 = require("@d-fischer/shared-utils");
var ChatClient_1 = require("./ChatClient");
__tsu.defineExport("ChatClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ChatClient_1.ChatClient; ***REMOVED*** ***REMOVED***);
/** @deprecated Use the named export `ChatClient` instead. */
var DeprecatedChatClient = shared_utils_1.deprecateClass(ChatClient_1.ChatClient, "[twitch-chat-client] The default export has been deprecated. Use the named export instead:\n\n\timport ***REMOVED*** ChatClient ***REMOVED*** from 'twitch-chat-client';");
/** @deprecated Use the named export `ChatClient` instead. */
exports.default = DeprecatedChatClient;
var TwitchPrivateMessage_1 = require("./StandardCommands/TwitchPrivateMessage");
__tsu.defineExport("PrivateMessage", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return TwitchPrivateMessage_1.TwitchPrivateMessage; ***REMOVED*** ***REMOVED***);
var ClearChat_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/ClearChat");
__tsu.defineExport("ClearChat", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ClearChat_1.ClearChat; ***REMOVED*** ***REMOVED***);
var HostTarget_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/HostTarget");
__tsu.defineExport("HostTarget", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return HostTarget_1.HostTarget; ***REMOVED*** ***REMOVED***);
var Reconnect_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/Reconnect");
__tsu.defineExport("Reconnect", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Reconnect_1.Reconnect; ***REMOVED*** ***REMOVED***);
var RoomState_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/RoomState");
__tsu.defineExport("RoomState", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return RoomState_1.RoomState; ***REMOVED*** ***REMOVED***);
var UserNotice_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/UserNotice");
__tsu.defineExport("UserNotice", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return UserNotice_1.UserNotice; ***REMOVED*** ***REMOVED***);
var UserState_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/UserState");
__tsu.defineExport("UserState", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return UserState_1.UserState; ***REMOVED*** ***REMOVED***);
var Whisper_1 = require("./Capabilities/TwitchCommandsCapability/MessageTypes/Whisper");
__tsu.defineExport("Whisper", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return Whisper_1.Whisper; ***REMOVED*** ***REMOVED***);
var ClearMsg_1 = require("./Capabilities/TwitchTagsCapability/MessageTypes/ClearMsg");
__tsu.defineExport("ClearMsg", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ClearMsg_1.ClearMsg; ***REMOVED*** ***REMOVED***);
var GlobalUserState_1 = require("./Capabilities/TwitchTagsCapability/MessageTypes/GlobalUserState");
__tsu.defineExport("GlobalUserState", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return GlobalUserState_1.GlobalUserState; ***REMOVED*** ***REMOVED***);
var ChatUser_1 = require("./ChatUser");
__tsu.defineExport("ChatUser", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return ChatUser_1.ChatUser; ***REMOVED*** ***REMOVED***);
var logger_1 = require("@d-fischer/logger");
__tsu.defineExport("LogLevel", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return logger_1.LogLevel; ***REMOVED*** ***REMOVED***);
var UserTools_1 = require("./Toolkit/UserTools");
__tsu.defineExport("toChannelName", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return UserTools_1.toChannelName; ***REMOVED*** ***REMOVED***);
__tsu.defineExport("toUserName", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return UserTools_1.toUserName; ***REMOVED*** ***REMOVED***);
var MessageTools_1 = require("./Toolkit/MessageTools");
__tsu.defineExport("parseTwitchMessage", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return MessageTools_1.parseTwitchMessage; ***REMOVED*** ***REMOVED***);
module.exports = exports = DeprecatedChatClient;
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
__tsu.redefineExports();
/** @deprecated Use the named export `ChatClient` instead. */
exports.default = DeprecatedChatClient;
