// twitch/es/API/Kraken/User/UserApi.mjs
import ***REMOVED*** __assign, __awaiter, __decorate, __extends, __generator, __read ***REMOVED*** from "tslib";
import ***REMOVED*** Cacheable, Cached, ClearsCache ***REMOVED*** from '@d-fischer/cache-decorators';
import ***REMOVED*** entriesToObject, indexBy, mapObject ***REMOVED*** from '@d-fischer/shared-utils';
import ***REMOVED*** HttpStatusCodeError ***REMOVED*** from 'twitch-api-call';
import ***REMOVED*** extractUserId, HellFreezesOverError, rtfm ***REMOVED*** from 'twitch-common';
import ***REMOVED*** NoSubscriptionProgramError ***REMOVED*** from "../../../Errors/NoSubscriptionProgramError.mjs";
import ***REMOVED*** BaseApi ***REMOVED*** from "../../BaseApi.mjs";
import ***REMOVED*** EmoteSetList ***REMOVED*** from "../Channel/EmoteSetList.mjs";
import ***REMOVED*** PrivilegedUser ***REMOVED*** from "./PrivilegedUser.mjs";
import ***REMOVED*** User ***REMOVED*** from "./User.mjs";
import ***REMOVED*** UserBlock ***REMOVED*** from "./UserBlock.mjs";
import ***REMOVED*** UserChatInfo ***REMOVED*** from "./UserChatInfo.mjs";
import ***REMOVED*** UserFollow ***REMOVED*** from "./UserFollow.mjs";
import ***REMOVED*** UserSubscription ***REMOVED*** from "./UserSubscription.mjs";
/**
 * The API methods that deal with users.
 *
 * Can be accessed using `client.kraken.users` on an ***REMOVED***@ApiClient***REMOVED*** instance.
 *
 * ## Example
 * ```ts
 * const api = new ApiClient(new StaticAuthProvider(clientId, accessToken));
 * const user = await api.kraken.users.getUser('125328655');
 * ```
 */
var UserApi = /** @class */ (function (_super) ***REMOVED***
    __extends(UserApi, _super);
    function UserApi() ***REMOVED***
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._userByNameCache = new Map();
        return _this;
    ***REMOVED***
    /**
     * Retrieves the user data of the currently authenticated user.
     */
    UserApi.prototype.getMe = function () ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var _a;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        _a = PrivilegedUser.bind;
                        return [4 /*yield*/, this._client.callApi(***REMOVED*** url: 'user', scope: 'user_read' ***REMOVED***)];
                    case 1: return [2 /*return*/, new (_a.apply(PrivilegedUser, [void 0, _b.sent(), this._client]))()];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the user data for the given user ID.
     *
     * @param userId The user ID you want to look up.
     */
    UserApi.prototype.getUser = function (userId) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userData;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this._client.callApi(***REMOVED*** url: "users/" + extractUserId(userId) ***REMOVED***)];
                    case 1:
                        userData = _a.sent();
                        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                        if (!userData) ***REMOVED***
                            throw new HellFreezesOverError('Could not get authenticated user');
                        ***REMOVED***
                        return [2 /*return*/, new User(userData, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the user data for the given user name.
     *
     * @param userName The user name you want to look up.
     */
    UserApi.prototype.getUserByName = function (userName) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var users, user;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        // not using the decorator's cache here as users-by-name is slightly more complex to cache
                        this._cleanUserCache();
                        if (this._userByNameCache.has(userName)) ***REMOVED***
                            return [2 /*return*/, this._userByNameCache.get(userName).value];
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: 'users',
                                query: ***REMOVED*** login: userName ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        users = (_a.sent()).users;
                        if (users.length === 0) ***REMOVED***
                            return [2 /*return*/, null];
                        ***REMOVED***
                        user = new User(users[0], this._client);
                        this._userByNameCache.set(userName, ***REMOVED***
                            value: user,
                            expires: Date.now() + 3600 * 1000
                        ***REMOVED***);
                        return [2 /*return*/, user];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the user data for the given user names.
     *
     * @param userNames The user names you want to look up.
     */
    UserApi.prototype.getUsersByNames = function (userNames) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var cachedEntries, cachedObject, cachedUsers, toFetch, usersData, usersArr, users;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        this._cleanUserCache();
                        userNames = userNames.map(function (name) ***REMOVED*** return name.toLowerCase(); ***REMOVED***);
                        cachedEntries = Array.from(this._userByNameCache.entries()).filter(function (_a) ***REMOVED***
                            var _b = __read(_a, 1), key = _b[0];
                            return userNames.includes(key);
                        ***REMOVED***);
                        cachedObject = entriesToObject(cachedEntries);
                        cachedUsers = mapObject(cachedObject, function (entry) ***REMOVED*** return entry.value; ***REMOVED***);
                        toFetch = userNames.filter(function (name) ***REMOVED*** return !(name in cachedUsers); ***REMOVED***);
                        if (!toFetch.length) ***REMOVED***
                            return [2 /*return*/, cachedUsers];
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: 'users',
                                query: ***REMOVED*** login: toFetch.join(',') ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        usersData = _a.sent();
                        usersArr = usersData.users.map(function (data) ***REMOVED*** return new User(data, _this._client); ***REMOVED***);
                        usersArr.forEach(function (user) ***REMOVED***
                            return _this._userByNameCache.set(user.name, ***REMOVED***
                                value: user,
                                expires: Date.now() + 3600 * 1000
                            ***REMOVED***);
                        ***REMOVED***);
                        users = indexBy(usersArr, 'name');
                        return [2 /*return*/, __assign(__assign(***REMOVED******REMOVED***, cachedUsers), users)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves information about the user's chat appearance and privileges.
     *
     * @param user The user you want to get chat info for.
     */
    UserApi.prototype.getChatInfo = function (user) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, data;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        return [4 /*yield*/, this._client.callApi(***REMOVED*** url: "users/" + userId + "/chat" ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, new UserChatInfo(data, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the emotes a given user can use.
     *
     * @param user The user you want to get emotes for.
     */
    UserApi.prototype.getUserEmotes = function (user) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, data;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/emotes",
                                scope: 'user_subscriptions'
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, new EmoteSetList(data.emoticon_sets)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves the subscription data for a given user to a given channel.
     *
     * @param user The user to retrieve the subscription data of.
     * @param toChannel The channel you want to retrieve the subscription data to.
     */
    UserApi.prototype.getSubscriptionData = function (user, toChannel) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, channelId, _a, e_1;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        channelId = extractUserId(toChannel);
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        _a = UserSubscription.bind;
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/subscriptions/" + channelId,
                                scope: 'user_subscriptions'
                            ***REMOVED***)];
                    case 2: return [2 /*return*/, new (_a.apply(UserSubscription, [void 0, _b.sent(), this._client]))()];
                    case 3:
                        e_1 = _b.sent();
                        if (e_1 instanceof HttpStatusCodeError) ***REMOVED***
                            if (e_1.statusCode === 404) ***REMOVED***
                                return [2 /*return*/, null];
                            ***REMOVED***
                            else if (e_1.statusCode === 422) ***REMOVED***
                                throw new NoSubscriptionProgramError(channelId);
                            ***REMOVED***
                        ***REMOVED***
                        throw e_1;
                    case 4: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Get a list of channels a given user follows.
     *
     * @param user The user you want to retrieve the follows of.
     * @param page The result page you want to retrieve.
     * @param limit The number of results you want to retrieve.
     * @param orderBy The field to order by.
     * @param orderDirection The direction to order in - ascending or descending.
     */
    UserApi.prototype.getFollowedChannels = function (user, page, limit, orderBy, orderDirection) ***REMOVED***
        if (limit === void 0) ***REMOVED*** limit = 25; ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, query, data;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        query = ***REMOVED******REMOVED***;
                        if (page) ***REMOVED***
                            query.offset = ((page - 1) * limit).toString();
                        ***REMOVED***
                        query.limit = limit.toString();
                        if (orderBy) ***REMOVED***
                            query.sortby = orderBy;
                        ***REMOVED***
                        if (orderDirection) ***REMOVED***
                            query.direction = orderDirection;
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/follows/channels",
                                query: query
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.follows.map(function (follow) ***REMOVED*** return new UserFollow(follow, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Get follow data for a given user to a given channel.
     *
     * @param user The user you want to retrieve follow data of.
     * @param channel The channel you want to retrieve follow data to.
     */
    UserApi.prototype.getFollowedChannel = function (user, channel) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, channelId, data, e_2;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        channelId = extractUserId(channel);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/follows/channels/" + channelId
                            ***REMOVED***)];
                    case 2:
                        data = _a.sent();
                        return [2 /*return*/, new UserFollow(data, this._client)];
                    case 3:
                        e_2 = _a.sent();
                        if (e_2 instanceof HttpStatusCodeError) ***REMOVED***
                            if (e_2.statusCode === 404) ***REMOVED***
                                return [2 /*return*/, null];
                            ***REMOVED***
                        ***REMOVED***
                        throw e_2;
                    case 4: return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Follows a given channel with a given user.
     *
     * @param user The user you want to follow with.
     * @param channel The channel to follow.
     * @param notifications Whether the user will receive notifications.
     */
    UserApi.prototype.followChannel = function (user, channel, notifications) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, channelId, data;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        channelId = extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/follows/channels/" + channelId,
                                method: 'PUT',
                                scope: 'user_follows_edit',
                                body: ***REMOVED*** notifications: Boolean(notifications).toString() ***REMOVED***
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, new UserFollow(data, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Unfollows a given channel with a given user.
     *
     * @param user The user you want to unfollow with.
     * @param channel The channel to unfollow.
     */
    UserApi.prototype.unfollowChannel = function (user, channel) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, channelId;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        channelId = extractUserId(channel);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/follows/channels/" + channelId,
                                scope: 'user_follows_edit',
                                method: 'DELETE'
                            ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Retrieves a list of users a given user has blocked.
     *
     * @param user The user you want to retrieve the block list of.
     * @param page The result page you want to retrieve.
     * @param limit The number of results you want to retrieve.
     */
    UserApi.prototype.getBlockedUsers = function (user, page, limit) ***REMOVED***
        if (limit === void 0) ***REMOVED*** limit = 25; ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, query, data;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        query = ***REMOVED*** limit: limit.toString() ***REMOVED***;
                        if (page) ***REMOVED***
                            query.offset = ((page - 1) * limit).toString();
                        ***REMOVED***
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/blocks",
                                query: query,
                                scope: 'user_blocks_read'
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, data.blocks.map(function (block) ***REMOVED*** return new UserBlock(block, _this._client); ***REMOVED***)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Blocks a given user with another given user.
     *
     * @param user The user you want to block with.
     * @param userToBlock The user to block.
     */
    UserApi.prototype.blockUser = function (user, userToBlock) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, userIdToBlock, data;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        userIdToBlock = extractUserId(userToBlock);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/blocks/" + userIdToBlock,
                                method: 'PUT',
                                scope: 'user_blocks_edit'
                            ***REMOVED***)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, new UserBlock(data, this._client)];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    /**
     * Unblocks a given user with another given user.
     *
     * @param user The user you want to unblock with.
     * @param userToUnblock The user to unblock.
     */
    UserApi.prototype.unblockUser = function (user, userToUnblock) ***REMOVED***
        return __awaiter(this, void 0, Promise, function () ***REMOVED***
            var userId, userIdToUnblock;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        userId = extractUserId(user);
                        userIdToUnblock = extractUserId(userToUnblock);
                        return [4 /*yield*/, this._client.callApi(***REMOVED***
                                url: "users/" + userId + "/blocks/" + userIdToUnblock,
                                method: 'DELETE',
                                scope: 'user_blocks_edit'
                            ***REMOVED***)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***);
    ***REMOVED***;
    UserApi.prototype._cleanUserCache = function () ***REMOVED***
        var _this = this;
        var now = Date.now();
        this._userByNameCache.forEach(function (val, key) ***REMOVED***
            if (val.expires < now) ***REMOVED***
                _this._userByNameCache.delete(key);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***;
    __decorate([
        Cached(3600)
    ], UserApi.prototype, "getMe", null);
    __decorate([
        Cached(3600)
    ], UserApi.prototype, "getUser", null);
    __decorate([
        Cached(3600)
    ], UserApi.prototype, "getChatInfo", null);
    __decorate([
        Cached(3600)
    ], UserApi.prototype, "getUserEmotes", null);
    __decorate([
        Cached(3600)
    ], UserApi.prototype, "getSubscriptionData", null);
    __decorate([
        Cached(300)
    ], UserApi.prototype, "getFollowedChannels", null);
    __decorate([
        Cached(300)
    ], UserApi.prototype, "getFollowedChannel", null);
    __decorate([
        ClearsCache('getFollowedChannels', 1),
        ClearsCache('getFollowedChannel', 2)
    ], UserApi.prototype, "followChannel", null);
    __decorate([
        ClearsCache('getFollowedChannels', 1),
        ClearsCache('getFollowedChannel', 2)
    ], UserApi.prototype, "unfollowChannel", null);
    __decorate([
        Cached(3600)
    ], UserApi.prototype, "getBlockedUsers", null);
    __decorate([
        ClearsCache('getBlockedUsers', 1)
    ], UserApi.prototype, "blockUser", null);
    __decorate([
        ClearsCache('getBlockedUsers', 1)
    ], UserApi.prototype, "unblockUser", null);
    UserApi = __decorate([
        Cacheable,
        rtfm('twitch', 'UserApi')
    ], UserApi);
    return UserApi;
***REMOVED***(BaseApi));
export ***REMOVED*** UserApi ***REMOVED***;
